<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Ensembl Metadata: Bio/EnsEMBL/MetaData/AnnotationAnalyzer.pm Source File</title>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('AnnotationAnalyzer_8pm.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">AnnotationAnalyzer.pm</div>  </div>
</div>
<div class="contents">
<a href="AnnotationAnalyzer_8pm.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 
<a name="l00002"></a>00002 =head1 LICENSE
<a name="l00003"></a>00003 
<a name="l00004"></a>00004 Copyright [2009-2016] EMBL-European Bioinformatics Institute
<a name="l00005"></a>00005 
<a name="l00006"></a>00006 Licensed under the Apache License, Version 2.0 (the <span class="stringliteral">&quot;License&quot;</span>);
<a name="l00007"></a>00007 you may not use <span class="keyword">this</span> file except in compliance with the License.
<a name="l00008"></a><a class="code" href="namespaceBio_1_1EnsEMBL_1_1MetaData.html">00008</a> You may obtain a copy of the License at
<a name="l00009"></a>00009 
<a name="l00010"></a>00010      http:<span class="comment">//www.apache.org/licenses/LICENSE-2.0</span>
<a name="l00011"></a>00011 
<a name="l00012"></a>00012 Unless required by applicable law or agreed to in writing, software
<a name="l00013"></a>00013 distributed under the License is distributed on an <span class="stringliteral">&quot;AS IS&quot;</span> BASIS,
<a name="l00014"></a>00014 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
<a name="l00015"></a>00015 See the License <span class="keywordflow">for</span> the specific language governing permissions and
<a name="l00016"></a>00016 limitations under the License.
<a name="l00017"></a>00017 
<a name="l00018"></a>00018 =head1 CONTACT
<a name="l00019"></a>00019 
<a name="l00020"></a>00020   Please email comments or questions to the <span class="keyword">public</span> Ensembl
<a name="l00021"></a><a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html">00021</a>   developers list at &lt;dev@ensembl.org&gt;.
<a name="l00022"></a>00022 
<a name="l00023"></a>00023   Questions may also be sent to the Ensembl help desk at
<a name="l00024"></a>00024   &lt;helpdesk@ensembl.org&gt;.
<a name="l00025"></a>00025 
<a name="l00026"></a>00026 =head1 NAME
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html">Bio::EnsEMBL::MetaData::AnnotationAnalyzer</a>
<a name="l00029"></a>00029 
<a name="l00030"></a>00030 =head1 SYNOPSIS
<a name="l00031"></a>00031 
<a name="l00032"></a>00032 my $anal = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html">Bio::EnsEMBL::MetaData::AnnotationAnalyzer</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a5f4168a9b1eaea0d3834ced9f9949cc8">new</a>();
<a name="l00033"></a>00033 my $details = $anal-&gt;analyze_annotation($dba);
<a name="l00034"></a>00034 
<a name="l00035"></a>00035 =head1 DESCRIPTION
<a name="l00036"></a>00036 
<a name="l00037"></a>00037 Utility <span class="keyword">class </span>for counting xrefs, variations etc. associated with a genome
<a name="l00038"></a>00038 
<a name="l00039"></a>00039 =head1 AUTHOR
<a name="l00040"></a>00040 
<a name="l00041"></a>00041 Dan Staines
<a name="l00042"></a>00042 
<a name="l00043"></a>00043 =cut
<a name="l00044"></a>00044 
<a name="l00045"></a>00045 use strict;
<a name="l00046"></a>00046 use warnings;
<a name="l00047"></a>00047 
<a name="l00048"></a>00048 <span class="keyword">package </span>Bio::EnsEMBL::MetaData::AnnotationAnalyzer;
<a name="l00049"></a>00049 use Bio::EnsEMBL::Utils::Exception qw/<span class="keywordflow">throw</span>/;
<a name="l00050"></a>00050 
<a name="l00051"></a>00051 use Log::Log4perl qw(get_logger);
<a name="l00052"></a>00052 use Data::Dumper;
<a name="l00053"></a>00053 use Config::IniFiles;
<a name="l00054"></a>00054 use LWP::UserAgent;
<a name="l00055"></a>00055 my $ua = LWP::UserAgent-&gt;new();
<a name="l00056"></a>00056 
<a name="l00057"></a>00057 my $url_template =
<a name="l00058"></a>00058 <span class="stringliteral">&#39;https://raw.github.com/EnsemblGenomes/eg-web-DIVISION/master/conf/ini-files/SPECIES.ini&#39;</span>;
<a name="l00059"></a>00059 
<a name="l00060"></a>00060 =head1 SUBROUTINES/METHODS
<a name="l00061"></a>00061 
<a name="l00062"></a>00062 =head2 <span class="keyword">new</span>
<a name="l00063"></a>00063   Description: Return a <span class="keyword">new</span> instance of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html">AnnotationAnalyzer</a>
<a name="l00064"></a>00064   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html">Bio::EnsEMBL::MetaData::AnnotationAnalyzer</a>
<a name="l00065"></a>00065   Exceptions : none
<a name="l00066"></a>00066   Caller     : general
<a name="l00067"></a>00067   Status     : Stable
<a name="l00068"></a>00068 =cut
<a name="l00069"></a>00069 
<a name="l00070"></a>00070 sub <span class="keyword">new</span> {
<a name="l00071"></a>00071   my $caller = shift;
<a name="l00072"></a>00072   my $class  = ref($caller) || $caller;
<a name="l00073"></a>00073   my $self   = bless( {}, $class );
<a name="l00074"></a>00074   $self-&gt;{logger} = get_logger();
<a name="l00075"></a>00075   <span class="keywordflow">return</span> $self;
<a name="l00076"></a>00076 }
<a name="l00077"></a>00077 
<a name="l00078"></a>00078 =head2 analyze_annotation
<a name="l00079"></a>00079   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00080"></a>00080   Description: Analyzes annotation content of the supplied core DB
<a name="l00081"></a>00081   Returntype :  Hash ref with the following keys:
<a name="l00082"></a>00082                     nProteinCoding - number of protein coding genes
<a name="l00083"></a>00083                     nProteinCodingUniProtKBSwissProt - number of protein coding genes with at least one SwissPROT entry
<a name="l00084"></a>00084                     nProteinCodingUniProtKBTrEMBL - number of protein coding genes with at least one TrEMBL entry
<a name="l00085"></a>00085                     nProteinCodingGO - number of protein coding genes with at least one GO term
<a name="l00086"></a>00086                     nProteinCodingInterPro - number of protein coding genes with at least one InterPro match
<a name="l00087"></a>00087                     nUniProtKBSwissProt - number of distinct UniProtKB/TrEMBL entries matching at least one translation
<a name="l00088"></a>00088                     nUniProtKBTrEMBL - number of distinct UniProtKB/TrEMBL entries matching at least one translation
<a name="l00089"></a>00089                     nGO - number of distinct GO terms matching at least one translation
<a name="l00090"></a>00090                     nInterPro - number of distinct InterPro entries matching at least one feature
<a name="l00091"></a>00091                     nInterProDomains - number of distinct protein features matching an InterPro domains
<a name="l00092"></a>00092   Exceptions : none
<a name="l00093"></a>00093   Caller     : general
<a name="l00094"></a>00094   Status     : Stable
<a name="l00095"></a>00095 =cut
<a name="l00096"></a>00096 
<a name="l00097"></a>00097 sub analyze_annotation {
<a name="l00098"></a>00098   my ( $self, $dba ) = @_;
<a name="l00099"></a>00099   $self-&gt;{logger}-&gt;debug( <span class="stringliteral">&quot;Analysing annotation for &quot;</span> . $dba-&gt;species() );
<a name="l00100"></a>00100   <span class="keywordflow">return</span> {
<a name="l00101"></a>00101       nProteinCoding =&gt; $self-&gt;count_by_biotype( $dba, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
<a name="l00102"></a>00102       nProteinCodingGO =&gt; $self-&gt;count_by_xref( $dba, <span class="stringliteral">&#39;GO&#39;</span>, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
<a name="l00103"></a>00103       nProteinCodingUniProtKB =&gt;
<a name="l00104"></a>00104         $self-&gt;count_by_xref( $dba, [ <span class="stringliteral">&#39;Uniprot/SWISSPROT&#39;</span>, <span class="stringliteral">&#39;Uniprot/SPTREMBL&#39;</span> ],
<a name="l00105"></a>00105                               <span class="stringliteral">&#39;protein_coding&#39;</span> ),
<a name="l00106"></a>00106       nProteinCodingUniProtKBSwissProt =&gt;
<a name="l00107"></a>00107         $self-&gt;count_by_xref( $dba, <span class="stringliteral">&#39;Uniprot/SWISSPROT&#39;</span>, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
<a name="l00108"></a>00108       nProteinCodingUniProtKBTrEMBL =&gt;
<a name="l00109"></a>00109         $self-&gt;count_by_xref( $dba, <span class="stringliteral">&#39;Uniprot/SPTREMBL&#39;</span>, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
<a name="l00110"></a>00110       nProteinCodingInterPro =&gt; $self-&gt;count_by_interpro($dba),
<a name="l00111"></a>00111       nGO =&gt; $self-&gt;count_by_xref( $dba, <span class="stringliteral">&#39;GO&#39;</span>, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
<a name="l00112"></a>00112       nUniProtKBSwissProt =&gt; $self-&gt;count_xrefs( $dba, <span class="stringliteral">&#39;Uniprot/SWISSPROT&#39;</span> ),
<a name="l00113"></a>00113       nUniProtKBTrEMBL    =&gt; $self-&gt;count_xrefs( $dba, <span class="stringliteral">&#39;Uniprot/SPTREMBL&#39;</span> ),
<a name="l00114"></a>00114       nInterPro           =&gt; $self-&gt;count_interpro($dba),
<a name="l00115"></a>00115       nInterProDomains =&gt; $self-&gt;count_interpro_domains($dba) };
<a name="l00116"></a>00116 }
<a name="l00117"></a>00117 
<a name="l00118"></a>00118 =head2 analyze_features
<a name="l00119"></a>00119   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00120"></a>00120   Description: Analyzes features found in the supplied core/otherfeatures database (simple_feature,repeat_feature,protein_align_feature,dna_align_features)
<a name="l00121"></a>00121   Returntype :  Hash ref (keys are feature tables, values are hashrefs of count by analysis name)
<a name="l00122"></a>00122   Exceptions : none
<a name="l00123"></a>00123   Caller     : general
<a name="l00124"></a>00124   Status     : Stable
<a name="l00125"></a>00125 =cut
<a name="l00126"></a>00126 
<a name="l00127"></a>00127 sub analyze_features {
<a name="l00128"></a>00128   my ( $self, $dba ) = @_;
<a name="l00129"></a>00129   <span class="keywordflow">return</span> { simpleFeatures =&gt; $self-&gt;count_features( $dba, <span class="stringliteral">&#39;simple_feature&#39;</span> ),
<a name="l00130"></a>00130            repeatFeatures =&gt; $self-&gt;count_features( $dba, <span class="stringliteral">&#39;repeat_feature&#39;</span> ) };
<a name="l00131"></a>00131 }
<a name="l00132"></a>00132 
<a name="l00133"></a>00133 =head2 analyze_variation
<a name="l00134"></a>00134   Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
<a name="l00135"></a>00135   Description: Analyzes variation found in the supplied variation database
<a name="l00136"></a>00136   Returntype :  Hash ref with the following keys
<a name="l00137"></a>00137                     variations - count of variations per source
<a name="l00138"></a>00138                     structuralVariations - count of structural variations per source
<a name="l00139"></a>00139                     genotypes - count of genotypes per sample
<a name="l00140"></a>00140                     phenotypes - count of variation annotations per phenotype
<a name="l00141"></a>00141   Exceptions : none
<a name="l00142"></a>00142   Caller     : general
<a name="l00143"></a>00143   Status     : Stable
<a name="l00144"></a>00144 =cut
<a name="l00145"></a>00145 
<a name="l00146"></a>00146 sub analyze_variation {
<a name="l00147"></a>00147   my ( $self, $dba ) = @_;
<a name="l00148"></a>00148   <span class="keywordflow">return</span> { variations           =&gt; $self-&gt;count_variations($dba),
<a name="l00149"></a>00149            structuralVariations =&gt; $self-&gt;count_structural_variations($dba),
<a name="l00150"></a>00150            genotypes            =&gt; $self-&gt;count_genotypes($dba),
<a name="l00151"></a>00151            phenotypes           =&gt; $self-&gt;count_phenotypes($dba) };
<a name="l00152"></a>00152 }
<a name="l00153"></a>00153 
<a name="l00154"></a>00154 =head2 analyze_compara
<a name="l00155"></a>00155   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00156"></a>00156   Arg        : Bio::EnsEMBL::Compara::DBSQL::DBAdaptor
<a name="l00157"></a>00157   Description: Analyzes compara methods found in the supplied compara database involving the supplied core species
<a name="l00158"></a>00158   Returntype :  Hash ref (keys are method names, values are arrays of partner species)
<a name="l00159"></a>00159   Exceptions : none
<a name="l00160"></a>00160   Caller     : general
<a name="l00161"></a>00161   Status     : Stable
<a name="l00162"></a>00162 =cut
<a name="l00163"></a>00163 
<a name="l00164"></a>00164 sub analyze_compara {
<a name="l00165"></a>00165   my ( $self, $dba, $core ) = @_;
<a name="l00166"></a>00166   my $compara = {};
<a name="l00167"></a>00167   eval {
<a name="l00168"></a>00168     my $gdb =
<a name="l00169"></a>00169       $dba-&gt;get_GenomeDBAdaptor()-&gt;fetch_by_registry_name( $core-&gt;species() );
<a name="l00170"></a>00170     <span class="keywordflow">for</span> my $mlss (
<a name="l00171"></a>00171        @{ $dba-&gt;get_MethodLinkSpeciesSetAdaptor()-&gt;fetch_all_by_GenomeDB($gdb) }
<a name="l00172"></a>00172       )
<a name="l00173"></a>00173     {
<a name="l00174"></a>00174       my $t = $mlss-&gt;method()-&gt;type();
<a name="l00175"></a>00175       next
<a name="l00176"></a>00176         <span class="keywordflow">if</span> ( $t eq <span class="stringliteral">&#39;FAMILY&#39;</span> ||
<a name="l00177"></a>00177              $t eq <span class="stringliteral">&#39;ENSEMBL_ORTHOLOGUES&#39;</span> ||
<a name="l00178"></a>00178              $t eq <span class="stringliteral">&#39;ENSEMBL_PARALOGUES&#39;</span> );
<a name="l00179"></a>00179       <span class="keywordflow">for</span> my $gdb2 ( grep { $gdb-&gt;dbID() ne $_-&gt;dbID() }
<a name="l00180"></a>00180                      @{ $mlss-&gt;species_set_obj-&gt;genome_dbs() } )
<a name="l00181"></a>00181       {
<a name="l00182"></a>00182         push( @{ $compara-&gt;{$t} }, $gdb2-&gt;name() );
<a name="l00183"></a>00183       }
<a name="l00184"></a>00184     }
<a name="l00185"></a>00185   };
<a name="l00186"></a>00186   <span class="keywordflow">if</span> ($@) {
<a name="l00187"></a>00187     warn <span class="stringliteral">&quot;No compara entry found for &quot;</span> . $dba-&gt;species();
<a name="l00188"></a>00188   }
<a name="l00189"></a>00189   <span class="keywordflow">return</span> $compara;
<a name="l00190"></a>00190 } ## end sub analyze_compara
<a name="l00191"></a>00191 
<a name="l00192"></a>00192 =head2 analyze_alignments
<a name="l00193"></a>00193   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00194"></a>00194   Description: Analyzes alignments <span class="keywordflow">for</span> sequences to the assembly 
<a name="l00195"></a>00195   Returntype : Hash ref - keys are source_type (e.g. RNA seq), values are arrayrefs of hashrefs containing track details (source_name,description,source_url)
<a name="l00196"></a>00196   Exceptions : none
<a name="l00197"></a>00197   Caller     : general
<a name="l00198"></a>00198   Status     : Stable
<a name="l00199"></a>00199 =cut
<a name="l00200"></a>00200 
<a name="l00201"></a>00201 sub analyze_alignments {
<a name="l00202"></a>00202   my ( $self, $dba ) = @_;
<a name="l00203"></a>00203   my $ali = {};
<a name="l00204"></a>00204   my $pf = $self-&gt;count_features( $dba, <span class="stringliteral">&#39;protein_align_feature&#39;</span> );
<a name="l00205"></a>00205   <span class="keywordflow">if</span> ( scalar( keys %$pf ) &gt; 0 ) {
<a name="l00206"></a>00206     $ali-&gt;{proteinAlignFeatures} = $pf;
<a name="l00207"></a>00207   }
<a name="l00208"></a>00208   my $df = $self-&gt;count_features( $dba, <span class="stringliteral">&#39;dna_align_feature&#39;</span> );
<a name="l00209"></a>00209   <span class="keywordflow">if</span> ( scalar( keys %$df ) &gt; 0 ) {
<a name="l00210"></a>00210     $ali-&gt;{dnaAlignFeatures} = $df;
<a name="l00211"></a>00211   }
<a name="l00212"></a>00212   <span class="keywordflow">return</span> $ali;
<a name="l00213"></a>00213 }
<a name="l00214"></a>00214 
<a name="l00215"></a>00215 =head2 analyze_tracks
<a name="l00216"></a>00216   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00217"></a>00217   Description: Analyzes tracks attached to the supplied core database
<a name="l00218"></a>00218   Returntype : Hash ref - keys are source_type (e.g. RNA seq), values are arrayrefs of hashrefs containing track details (source_name,description,source_url)
<a name="l00219"></a>00219   Exceptions : none
<a name="l00220"></a>00220   Caller     : general
<a name="l00221"></a>00221   Status     : Stable   
<a name="l00222"></a>00222 =cut
<a name="l00223"></a>00223 
<a name="l00224"></a>00224 sub analyze_tracks {
<a name="l00225"></a>00225   my ( $self, $species, $division ) = @_;
<a name="l00226"></a>00226   $species = ucfirst($species);
<a name="l00227"></a>00227   ( $division = lc $division ) =~ s/ensembl<span class="comment">//;</span>
<a name="l00228"></a>00228   # <span class="keyword">get</span> the ini file from git
<a name="l00229"></a>00229   ( my $ini_url = $url_template ) =~ s/SPECIES/$species/;
<a name="l00230"></a>00230   $ini_url =~ s/DIVISION/$division/;
<a name="l00231"></a>00231 
<a name="l00232"></a>00232   my $req = HTTP::Request-&gt;new( GET =&gt; $ini_url );
<a name="l00233"></a>00233 <span class="preprocessor">  # Pass request to the user agent and get a response back</span>
<a name="l00234"></a>00234 <span class="preprocessor"></span>  my $res = $ua-&gt;request($req);
<a name="l00235"></a>00235   my $ini;
<a name="l00236"></a>00236 <span class="preprocessor">  # Check the outcome of the response</span>
<a name="l00237"></a>00237 <span class="preprocessor"></span>  <span class="keywordflow">if</span> ( $res-&gt;is_success ) {
<a name="l00238"></a>00238     $ini = $res-&gt;content;
<a name="l00239"></a>00239   }
<a name="l00240"></a>00240   <span class="keywordflow">else</span> {
<a name="l00241"></a>00241     $self-&gt;{logger}
<a name="l00242"></a>00242       -&gt;debug( <span class="stringliteral">&quot;Could not retrieve $ini_url: &quot;</span> . $res-&gt;status_line );
<a name="l00243"></a>00243   }
<a name="l00244"></a>00244 
<a name="l00245"></a>00245 <span class="preprocessor"># parse out and look at:</span>
<a name="l00246"></a>00246 <span class="preprocessor"></span><span class="preprocessor"># [ENSEMBL_INTERNAL_BAM_SOURCES]</span>
<a name="l00247"></a>00247 <span class="preprocessor"></span><span class="preprocessor"># 1_Puccinia_triticina_SRR035315 = dna_align_est</span>
<a name="l00248"></a>00248 <span class="preprocessor"></span><span class="preprocessor"># walk through keys and use keys to find sections like:</span>
<a name="l00249"></a>00249 <span class="preprocessor"></span><span class="preprocessor"># [1_Puccinia_triticina_SRR035315]</span>
<a name="l00250"></a>00250 <span class="preprocessor"></span><span class="preprocessor"># source_name        = Transcriptomics sequences of fresh spores. Run id: SRR035315</span>
<a name="l00251"></a>00251 <span class="preprocessor"></span><span class="preprocessor"># description        = RNA-Seq study.</span>
<a name="l00252"></a>00252 <span class="preprocessor"></span><span class="preprocessor"># source_url         = http://ftp.sra.ebi.ac.uk/vol1/ERZ000/ERZ000002/SRR035315.bam</span>
<a name="l00253"></a>00253 <span class="preprocessor"></span><span class="preprocessor"># source_type        = rnaseq</span>
<a name="l00254"></a>00254 <span class="preprocessor"></span><span class="preprocessor"># display            = normal</span>
<a name="l00255"></a>00255 <span class="preprocessor"></span><span class="preprocessor"># then store some or all of this in my output e.g. {bam}{source_type}[{source_name,description,source_url}]</span>
<a name="l00256"></a>00256 <span class="preprocessor"></span>  my $bams = {};
<a name="l00257"></a>00257   <span class="keywordflow">if</span> ( defined $ini ) {
<a name="l00258"></a>00258     my $cfg = Config::IniFiles-&gt;new( -file =&gt; \$ini );
<a name="l00259"></a>00259     <span class="keywordflow">if</span> ( defined $cfg ) {
<a name="l00260"></a>00260       <span class="keywordflow">for</span> my $bam ( $cfg-&gt;Parameters(<span class="stringliteral">&quot;ENSEMBL_INTERNAL_BAM_SOURCES&quot;</span>) ) {
<a name="l00261"></a>00261         push @{ $bams-&gt;{ $cfg-&gt;val( $bam, <span class="stringliteral">&#39;source_type&#39;</span> ) } }, {
<a name="l00262"></a>00262             <span class="keywordtype">id</span>          =&gt; $bam,
<a name="l00263"></a>00263             source_name =&gt; $cfg-&gt;val( $bam, <span class="stringliteral">&#39;source_name&#39;</span> ),
<a name="l00264"></a>00264             source_url  =&gt; $cfg-&gt;val( $bam, <span class="stringliteral">&#39;source_url&#39;</span> ),
<a name="l00265"></a>00265             description =&gt; $cfg-&gt;val( $bam, <span class="stringliteral">&#39;description&#39;</span> ) };
<a name="l00266"></a>00266       }
<a name="l00267"></a>00267     }
<a name="l00268"></a>00268   }
<a name="l00269"></a>00269   <span class="keywordflow">return</span> $bams;
<a name="l00270"></a>00270 } ## end sub analyze_tracks
<a name="l00271"></a>00271 
<a name="l00272"></a>00272 =head2 count_by_biotype
<a name="l00273"></a>00273   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00274"></a>00274   Arg        : String (biotype)
<a name="l00275"></a>00275   Description: Returns count of genes with specified biotype
<a name="l00276"></a>00276   Returntype : Integer
<a name="l00277"></a>00277   Exceptions : none
<a name="l00278"></a>00278   Caller     : internal
<a name="l00279"></a>00279   Status     : Stable
<a name="l00280"></a>00280 =cut
<a name="l00281"></a>00281 
<a name="l00282"></a>00282 sub count_by_biotype {
<a name="l00283"></a>00283   my ( $self, $dba, $biotype ) = @_;
<a name="l00284"></a>00284   <span class="keywordflow">return</span> $dba-&gt;get_GeneAdaptor()-&gt;count_all_by_biotype($biotype);
<a name="l00285"></a>00285 }
<a name="l00286"></a>00286 
<a name="l00287"></a>00287 =head2 count_by_xref
<a name="l00288"></a>00288   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00289"></a>00289   Arg        : Arrayref of dbnames 
<a name="l00290"></a>00290   Arg        : (optional) String (biotype)
<a name="l00291"></a>00291   Description: Returns count of genes with specified cross-reference
<a name="l00292"></a>00292   Returntype : Integer
<a name="l00293"></a>00293   Exceptions : none
<a name="l00294"></a>00294   Caller     : <span class="keyword">internal</span>
<a name="l00295"></a>00295   Status     : Stable
<a name="l00296"></a>00296 =cut
<a name="l00297"></a>00297 
<a name="l00298"></a>00298 my $xref_gene_sql = q/select count(distinct(gene_id)) 
<a name="l00299"></a>00299 from gene g
<a name="l00300"></a>00300 join seq_region s using (seq_region_id) 
<a name="l00301"></a>00301 join coord_system c using (coord_system_id) 
<a name="l00302"></a>00302 join transcript tr using (gene_id)
<a name="l00303"></a>00303 join translation t using (transcript_id)
<a name="l00304"></a>00304 join object_xref ox on (ox.ensembl_id=t.translation_id and ox.ensembl_object_type=&#39;Translation&#39;)
<a name="l00305"></a>00305 join xref x using (xref_id)
<a name="l00306"></a>00306 join external_db d using (external_db_id)
<a name="l00307"></a>00307 where species_id=? and d.db_name in (NAMES)/;
<a name="l00308"></a>00308 
<a name="l00309"></a>00309 my $biotype_clause = q/ and g.biotype=?/;
<a name="l00310"></a>00310 
<a name="l00311"></a>00311 sub count_by_xref {
<a name="l00312"></a>00312   my ( $self, $dba, $db_names, $biotype ) = @_;
<a name="l00313"></a>00313   $self-&gt;{logger}-&gt;debug( <span class="stringliteral">&quot;Counting genes by &quot;</span> .
<a name="l00314"></a>00314                       join( <span class="stringliteral">&quot;,&quot;</span>, $db_names ) . <span class="stringliteral">&quot; xref for &quot;</span> . $dba-&gt;species() );
<a name="l00315"></a>00315   $db_names = [$db_names] <span class="keywordflow">if</span> ( ref($db_names) ne <span class="stringliteral">&#39;ARRAY&#39;</span> );
<a name="l00316"></a>00316   my $sql = $xref_gene_sql;
<a name="l00317"></a>00317   my $db_name = join <span class="charliteral">&#39;,&#39;</span>, map { <span class="stringliteral">&quot;\&quot;$_\&quot;&quot;</span> } @$db_names;
<a name="l00318"></a>00318   $sql =~ s/NAMES/$db_name/;
<a name="l00319"></a>00319   my $params = [ $dba-&gt;species_id() ];
<a name="l00320"></a>00320   <span class="keywordflow">if</span> ( defined $biotype ) {
<a name="l00321"></a>00321     $sql .= $biotype_clause;
<a name="l00322"></a>00322     push @$params, $biotype;
<a name="l00323"></a>00323   }
<a name="l00324"></a>00324   $self-&gt;{logger}
<a name="l00325"></a>00325     -&gt;debug( <span class="stringliteral">&quot;Executing $sql with params: [&quot;</span> . join( <span class="stringliteral">&quot;,&quot;</span>, @$params ) );
<a name="l00326"></a>00326   <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
<a name="l00327"></a>00327     -&gt;execute_single_result( -SQL =&gt; $sql, -PARAMS =&gt; $params );
<a name="l00328"></a>00328 }
<a name="l00329"></a>00329 
<a name="l00330"></a>00330 =head2 count_xrefs
<a name="l00331"></a>00331   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00332"></a>00332   Arg        : One or more dbnames
<a name="l00333"></a>00333   Description: Returns count of numbers of cross-references to supplied database(s)
<a name="l00334"></a>00334   Returntype : Integer
<a name="l00335"></a>00335   Exceptions : none
<a name="l00336"></a>00336   Caller     : internal
<a name="l00337"></a>00337   Status     : Stable
<a name="l00338"></a>00338 =cut
<a name="l00339"></a>00339 
<a name="l00340"></a>00340 my $gene_xref_count_sql = q/
<a name="l00341"></a>00341 select count(distinct(dbprimary_acc)) 
<a name="l00342"></a>00342 from gene g
<a name="l00343"></a>00343 join seq_region s using (seq_region_id) 
<a name="l00344"></a>00344 join coord_system c using (coord_system_id) 
<a name="l00345"></a>00345 join object_xref ox on (ox.ensembl_id=g.gene_id and ox.ensembl_object_type=&#39;Gene&#39;)
<a name="l00346"></a>00346 join xref x using (xref_id)
<a name="l00347"></a>00347 join external_db d using (external_db_id)
<a name="l00348"></a>00348 where species_id=? and d.db_name=?/;
<a name="l00349"></a>00349 
<a name="l00350"></a>00350 my $transcript_xref_count_sql = q/
<a name="l00351"></a>00351 select count(distinct(dbprimary_acc)) 
<a name="l00352"></a>00352 from transcript tr
<a name="l00353"></a>00353 join seq_region s using (seq_region_id) 
<a name="l00354"></a>00354 join coord_system c using (coord_system_id) 
<a name="l00355"></a>00355 join object_xref ox on (ox.ensembl_id=tr.transcript_id and ox.ensembl_object_type=&#39;Transcript&#39;)
<a name="l00356"></a>00356 join xref x using (xref_id)
<a name="l00357"></a>00357 join external_db d using (external_db_id)
<a name="l00358"></a>00358 where species_id=? and d.db_name=?/;
<a name="l00359"></a>00359 
<a name="l00360"></a>00360 my $translation_xref_count_sql = q/
<a name="l00361"></a>00361 select count(distinct(dbprimary_acc)) 
<a name="l00362"></a>00362 from transcript tr
<a name="l00363"></a>00363 join translation t using (transcript_id)
<a name="l00364"></a>00364 join seq_region s using (seq_region_id) 
<a name="l00365"></a>00365 join coord_system c using (coord_system_id) 
<a name="l00366"></a>00366 join object_xref ox on (ox.ensembl_id=t.translation_id and ox.ensembl_object_type=&#39;Translation&#39;)
<a name="l00367"></a>00367 join xref x using (xref_id)
<a name="l00368"></a>00368 join external_db d using (external_db_id)
<a name="l00369"></a>00369 where species_id=? and d.db_name=?/;
<a name="l00370"></a>00370 
<a name="l00371"></a>00371 sub count_xrefs {
<a name="l00372"></a>00372   my $self = shift;
<a name="l00373"></a>00373   my $dba  = shift;
<a name="l00374"></a>00374   my $tot  = 0;
<a name="l00375"></a>00375   <span class="keywordflow">for</span> my $db_name (@_) {
<a name="l00376"></a>00376     $tot +=
<a name="l00377"></a>00377       $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
<a name="l00378"></a>00378                                      -SQL    =&gt; $gene_xref_count_sql,
<a name="l00379"></a>00379                                      -PARAMS =&gt; [ $dba-&gt;species_id(), $db_name ]
<a name="l00380"></a>00380       );
<a name="l00381"></a>00381     $tot +=
<a name="l00382"></a>00382       $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
<a name="l00383"></a>00383                                      -SQL    =&gt; $transcript_xref_count_sql,
<a name="l00384"></a>00384                                      -PARAMS =&gt; [ $dba-&gt;species_id(), $db_name ]
<a name="l00385"></a>00385       );
<a name="l00386"></a>00386     $tot +=
<a name="l00387"></a>00387       $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
<a name="l00388"></a>00388                                      -SQL    =&gt; $translation_xref_count_sql,
<a name="l00389"></a>00389                                      -PARAMS =&gt; [ $dba-&gt;species_id(), $db_name ]
<a name="l00390"></a>00390       );
<a name="l00391"></a>00391   }
<a name="l00392"></a>00392   <span class="keywordflow">return</span> $tot;
<a name="l00393"></a>00393 }
<a name="l00394"></a>00394 
<a name="l00395"></a>00395 =head2 count_by_interpro
<a name="l00396"></a>00396   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00397"></a>00397   Description: Returns count of genes with InterPro domains
<a name="l00398"></a>00398   Returntype : Integer
<a name="l00399"></a>00399   Exceptions : none
<a name="l00400"></a>00400   Caller     : <span class="keyword">internal</span>
<a name="l00401"></a>00401   Status     : Stable
<a name="l00402"></a>00402 =cut
<a name="l00403"></a>00403 
<a name="l00404"></a>00404 my $count_by_interpro_base = q/ from interpro
<a name="l00405"></a>00405 join protein_feature on (<span class="keywordtype">id</span>=hit_name)
<a name="l00406"></a>00406 join translation using (translation_id)
<a name="l00407"></a>00407 join transcript using (transcript_id)
<a name="l00408"></a>00408 join seq_region using (seq_region_id)
<a name="l00409"></a>00409 join coord_system using (coord_system_id)
<a name="l00410"></a>00410 where species_id=?
<a name="l00411"></a>00411 /;
<a name="l00412"></a>00412 
<a name="l00413"></a>00413 my $count_by_interpro =
<a name="l00414"></a>00414   &#39;select count(distinct(interpro_ac)) &#39; . $count_by_interpro_base;
<a name="l00415"></a>00415 
<a name="l00416"></a>00416 sub count_by_interpro {
<a name="l00417"></a>00417   my ( $self, $dba ) = @_;
<a name="l00418"></a>00418   <span class="keywordflow">return</span>
<a name="l00419"></a>00419     $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
<a name="l00420"></a>00420                                                -SQL    =&gt; $count_by_interpro,
<a name="l00421"></a>00421                                                -PARAMS =&gt; [ $dba-&gt;species_id() ]
<a name="l00422"></a>00422     );
<a name="l00423"></a>00423 }
<a name="l00424"></a>00424 
<a name="l00425"></a>00425 =head2 count_interpro
<a name="l00426"></a>00426   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00427"></a>00427   Description: Returns count of InterPro cross-references
<a name="l00428"></a>00428   Returntype : Integer
<a name="l00429"></a>00429   Exceptions : none
<a name="l00430"></a>00430   Caller     : <span class="keyword">internal</span>
<a name="l00431"></a>00431   Status     : Stable
<a name="l00432"></a>00432 =cut
<a name="l00433"></a>00433 
<a name="l00434"></a>00434 my $count_interpro =
<a name="l00435"></a>00435   q/select count(distinct(translation_id)) / . $count_by_interpro_base;
<a name="l00436"></a>00436 
<a name="l00437"></a>00437 sub count_interpro {
<a name="l00438"></a>00438   my ( $self, $dba ) = @_;
<a name="l00439"></a>00439   <span class="keywordflow">return</span>
<a name="l00440"></a>00440     $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
<a name="l00441"></a>00441                                                -SQL    =&gt; $count_interpro,
<a name="l00442"></a>00442                                                -PARAMS =&gt; [ $dba-&gt;species_id() ]
<a name="l00443"></a>00443     );
<a name="l00444"></a>00444 }
<a name="l00445"></a>00445 
<a name="l00446"></a>00446 =head2 count_interpro_domains
<a name="l00447"></a>00447   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00448"></a>00448   Description: Returns count of InterPro domains
<a name="l00449"></a>00449   Returntype : Integer
<a name="l00450"></a>00450   Exceptions : none
<a name="l00451"></a>00451   Caller     : <span class="keyword">internal</span>
<a name="l00452"></a>00452   Status     : Stable
<a name="l00453"></a>00453 =cut
<a name="l00454"></a>00454 
<a name="l00455"></a>00455 my $count_interpro_domains =
<a name="l00456"></a>00456   q/select count(distinct(protein_feature_id)) / . $count_by_interpro_base;
<a name="l00457"></a>00457 
<a name="l00458"></a>00458 sub count_interpro_domains {
<a name="l00459"></a>00459   my ( $self, $dba ) = @_;
<a name="l00460"></a>00460   <span class="keywordflow">return</span>
<a name="l00461"></a>00461     $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
<a name="l00462"></a>00462                                                -SQL =&gt; $count_interpro_domains,
<a name="l00463"></a>00463                                                -PARAMS =&gt; [ $dba-&gt;species_id() ]
<a name="l00464"></a>00464     );
<a name="l00465"></a>00465 }
<a name="l00466"></a>00466 
<a name="l00467"></a>00467 =head2 count_toplevel
<a name="l00468"></a>00468   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00469"></a>00469   Description: Returns count of top-level regions
<a name="l00470"></a>00470   Returntype : Integer
<a name="l00471"></a>00471   Exceptions : none
<a name="l00472"></a>00472   Caller     : <span class="keyword">internal</span>
<a name="l00473"></a>00473   Status     : Stable
<a name="l00474"></a>00474 =cut
<a name="l00475"></a>00475 
<a name="l00476"></a>00476 my $count_toplevel = q/select count(*) from seq_region 
<a name="l00477"></a>00477 join seq_region_attrib using (seq_region_id) 
<a name="l00478"></a>00478 join coord_system using (coord_system_id) 
<a name="l00479"></a>00479 join attrib_type using (attrib_type_id)
<a name="l00480"></a>00480 where species_id=? and code=&#39;toplevel&#39;/;
<a name="l00481"></a>00481 
<a name="l00482"></a>00482 sub count_toplevel {
<a name="l00483"></a>00483   my ( $self, $dba ) = @_;
<a name="l00484"></a>00484   <span class="keywordflow">return</span>
<a name="l00485"></a>00485     $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
<a name="l00486"></a>00486                                                -SQL    =&gt; $count_toplevel,
<a name="l00487"></a>00487                                                -PARAMS =&gt; [ $dba-&gt;species_id() ]
<a name="l00488"></a>00488     );
<a name="l00489"></a>00489 }
<a name="l00490"></a>00490 
<a name="l00491"></a>00491 =head2 count_features
<a name="l00492"></a>00492   Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00493"></a>00493   Arg        : String - feature type (e.g. RepeatFeature)
<a name="l00494"></a>00494   Description: Returns count of specified feature by analysis
<a name="l00495"></a>00495   Returntype : Hashref
<a name="l00496"></a>00496   Exceptions : none
<a name="l00497"></a>00497   Caller     : internal
<a name="l00498"></a>00498   Status     : Stable
<a name="l00499"></a>00499 =cut
<a name="l00500"></a>00500 
<a name="l00501"></a>00501 my $count_features = q/select logic_name,count(*) 
<a name="l00502"></a>00502 from TABLE join analysis using (analysis_id) 
<a name="l00503"></a>00503 join seq_region using (seq_region_id) 
<a name="l00504"></a>00504 join coord_system using (coord_system_id) 
<a name="l00505"></a>00505 where species_id=? group by logic_name/;
<a name="l00506"></a>00506 
<a name="l00507"></a>00507 sub count_features {
<a name="l00508"></a>00508   my ( $self, $dba, $table ) = @_;
<a name="l00509"></a>00509   my $sql = $count_features;
<a name="l00510"></a>00510   $sql =~ s/TABLE/$table/;
<a name="l00511"></a>00511   <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
<a name="l00512"></a>00512     -&gt;execute_into_hash( -SQL =&gt; $sql, -PARAMS =&gt; [ $dba-&gt;species_id() ] );
<a name="l00513"></a>00513 }
<a name="l00514"></a>00514 
<a name="l00515"></a>00515 =head2 count_variations
<a name="l00516"></a>00516   Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
<a name="l00517"></a>00517   Description: Returns count of variations by source
<a name="l00518"></a>00518   Returntype : Hashref
<a name="l00519"></a>00519   Exceptions : none
<a name="l00520"></a>00520   Caller     : <span class="keyword">internal</span>
<a name="l00521"></a>00521   Status     : Stable
<a name="l00522"></a>00522 =cut
<a name="l00523"></a>00523 
<a name="l00524"></a>00524 my $count_variation = q/select ifnull(s.name,<span class="stringliteral">&quot;unknown&quot;</span>),count(*) 
<a name="l00525"></a>00525 from variation v 
<a name="l00526"></a>00526 join source s using (source_id) 
<a name="l00527"></a>00527 group by s.name/;
<a name="l00528"></a>00528 
<a name="l00529"></a>00529 sub count_variations {
<a name="l00530"></a>00530   my ( $self, $dba ) = @_;
<a name="l00531"></a>00531   <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
<a name="l00532"></a>00532     -&gt;execute_into_hash( -SQL =&gt; $count_variation, -PARAMS =&gt; [] );
<a name="l00533"></a>00533 }
<a name="l00534"></a>00534 
<a name="l00535"></a>00535 =head2 count_structural_variations
<a name="l00536"></a>00536   Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
<a name="l00537"></a>00537   Description: Returns count of structural variations by source
<a name="l00538"></a>00538   Returntype : Hashref
<a name="l00539"></a>00539   Exceptions : none
<a name="l00540"></a>00540   Caller     : <span class="keyword">internal</span>
<a name="l00541"></a>00541   Status     : Stable
<a name="l00542"></a>00542 =cut
<a name="l00543"></a>00543 
<a name="l00544"></a>00544 my $count_structural_variation = q/select ifnull(s.name,<span class="stringliteral">&quot;unknown&quot;</span>),count(*) 
<a name="l00545"></a>00545 from structural_variation v 
<a name="l00546"></a>00546 join source s using (source_id) 
<a name="l00547"></a>00547 group by s.name/;
<a name="l00548"></a>00548 
<a name="l00549"></a>00549 sub count_structural_variations {
<a name="l00550"></a>00550   my ( $self, $dba ) = @_;
<a name="l00551"></a>00551   <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
<a name="l00552"></a>00552     -&gt;execute_into_hash( -SQL =&gt; $count_structural_variation, -PARAMS =&gt; [] );
<a name="l00553"></a>00553 }
<a name="l00554"></a>00554 
<a name="l00555"></a>00555 =head2 count_genotypes
<a name="l00556"></a>00556   Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
<a name="l00557"></a>00557   Description: Returns count of genotypes by source
<a name="l00558"></a>00558   Returntype : Hashref
<a name="l00559"></a>00559   Exceptions : none
<a name="l00560"></a>00560   Caller     : <span class="keyword">internal</span>
<a name="l00561"></a>00561   Status     : Stable
<a name="l00562"></a>00562 =cut
<a name="l00563"></a>00563 
<a name="l00564"></a>00564 my $count_genotypes = q/
<a name="l00565"></a>00565 select ifnull(name,<span class="stringliteral">&quot;unknown&quot;</span>),count(*) 
<a name="l00566"></a>00566 from tmp_individual_genotype_single_bp 
<a name="l00567"></a>00567 join individual using (individual_id) group by name/;
<a name="l00568"></a>00568 
<a name="l00569"></a>00569 sub count_genotypes {
<a name="l00570"></a>00570   my ( $self, $dba ) = @_;
<a name="l00571"></a>00571   my $cnt = {};
<a name="l00572"></a>00572   eval {
<a name="l00573"></a>00573     $cnt =
<a name="l00574"></a>00574       $dba-&gt;dbc()-&gt;sql_helper()
<a name="l00575"></a>00575       -&gt;execute_into_hash( -SQL =&gt; $count_genotypes, -PARAMS =&gt; [] );
<a name="l00576"></a>00576   };
<a name="l00577"></a>00577   <span class="keywordflow">return</span> $cnt;
<a name="l00578"></a>00578 }
<a name="l00579"></a>00579 
<a name="l00580"></a>00580 =head2 count_phenotypes
<a name="l00581"></a>00581   Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
<a name="l00582"></a>00582   Description: Returns count of phenotypes by name
<a name="l00583"></a>00583   Returntype : Integer
<a name="l00584"></a>00584   Exceptions : none
<a name="l00585"></a>00585   Caller     : <span class="keyword">internal</span>
<a name="l00586"></a>00586   Status     : Stable
<a name="l00587"></a>00587 =cut
<a name="l00588"></a>00588 
<a name="l00589"></a>00589 my $count_phenotypes = q/
<a name="l00590"></a>00590 select ifnull(p.name,<span class="stringliteral">&quot;unknown&quot;</span>),count(*) 
<a name="l00591"></a>00591 from phenotype p 
<a name="l00592"></a>00592 join phenotype_feature pf using (phenotype_id) 
<a name="l00593"></a>00593 join variation v on (object_id=v.name and type=&#39;Variation&#39;) 
<a name="l00594"></a>00594 group by p.name;
<a name="l00595"></a>00595 /;
<a name="l00596"></a>00596 
<a name="l00597"></a>00597 sub count_phenotypes {
<a name="l00598"></a>00598   my ( $self, $dba ) = @_;
<a name="l00599"></a>00599   <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
<a name="l00600"></a>00600     -&gt;execute_into_hash( -SQL =&gt; $count_phenotypes, -PARAMS =&gt; [] );
<a name="l00601"></a>00601 }
<a name="l00602"></a>00602 
<a name="l00603"></a>00603 1;
</pre></div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="AnnotationAnalyzer_8pm.html">AnnotationAnalyzer.pm</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
