<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Ensembl Metadata: Bio::EnsEMBL::MetaData::MetaDataProcessor Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="inherits.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="annotated.html"><span>Full&#160;Class&#160;List</span></a></li>
      <li><a href="functions.html"><span>Methods&#160;from&#160;all&#160;packages</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#pub-methods">Available Methods</a>  </div>
  <div class="headertitle">
<div class="title">Bio::EnsEMBL::MetaData::MetaDataProcessor Class Reference</div>  </div>
</div>
<div class="contents">
<!-- doxytag: class="Bio::EnsEMBL::MetaData::MetaDataProcessor" -->
<p><a href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor-members.html">List of all members.</a></p>
<hr/><a name="details" id="details"></a><h2>Class Summary</h2>
<div class="textblock"><h2><a class="anchor" id="Synopsis"></a>
Synopsis</h2>
<div class="fragment"><pre class="fragment">my $processor = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html">Bio::EnsEMBL::MetaData::MetaDataProcessor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a32c63ed55ff1ef0c84687294a4380ecd">new</a>();
my $info = $processor-&gt;process_genome($dba);
</pre></div> <h2><a class="anchor" id="Description"></a>
Description</h2>
 <pre>
Object for generating GenomeInfo objects from DBAdaptor objects.

</pre> <h2><a class="anchor" id="See-also"></a>
See-also</h2>
<p><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1BaseInfo.html">Bio::EnsEMBL::MetaData::BaseInfo</a> <a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeOrganismInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeOrganismInfoAdaptor</a> </p>

<p>Definition at line <a class="el" href="MetaDataProcessor_8pm_source.html#l00029">29</a> of file <a class="el" href="MetaDataProcessor_8pm_source.html">MetaDataProcessor.pm</a>.</p>
</div><table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Available Methods</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Long&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#abf2e21bfe55cbfb5b2d3f28b10560659">get_dbsize</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a32c63ed55ff1ef0c84687294a4380ecd">new</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a67a558d1578d2343cd2393807fe6394c">process_compara</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public <br class="typebreak"/>
<a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a0edad57751ae8fbe56ab226aafb33c18">process_genome</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Arrayref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a9d37c6b5f65e890a49169ce66acebfaa">process_metadata</a> ()</td></tr>
</table>
<hr/><h2>Method Documentation</h2>
<a class="anchor" id="abf2e21bfe55cbfb5b2d3f28b10560659"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::MetaDataProcessor::get_dbsize" ref="abf2e21bfe55cbfb5b2d3f28b10560659" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Long Bio::EnsEMBL::MetaData::MetaDataProcessor::get_dbsize </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : DBAdaptor
  Description: Calculate size of supplied database
  Returntype : Long
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-get_dbsize" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-get_dbsize-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-get_dbsize-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-get_dbsize-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#abf2e21bfe55cbfb5b2d3f28b10560659">get_dbsize</a> {
  my ($dba) = @_;
  <span class="keywordflow">return</span>
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
    -SQL =&gt;
<span class="stringliteral">&quot;select SUM(data_length + index_length) from information_schema.tables where table_schema=?&quot;</span>,
    -PARAMS =&gt; [ $dba-&gt;dbc()-&gt;dbname() ] );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a32c63ed55ff1ef0c84687294a4380ecd"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::MetaDataProcessor::new" ref="a32c63ed55ff1ef0c84687294a4380ecd" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::MetaData::MetaDataProcessor::new </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-new" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-new-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-new-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-new-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <span class="keyword">new</span> {
  my ( $caller, @args ) = @_;
  my $class = ref($caller) || $caller;
  my $self = bless( {}, $class );
  ( $self-&gt;{contigs},      $self-&gt;{annotation_analyzer},
    $self-&gt;{variation},    $self-&gt;{compara},
    $self-&gt;{info_adaptor}, $self-&gt;{force_update},
    $self-&gt;{release},      $self-&gt;{eg_release} )
    = rearrange( [ <span class="stringliteral">&#39;CONTIGS&#39;</span>,      <span class="stringliteral">&#39;ANNOTATION_ANALYZER&#39;</span>,
                   <span class="stringliteral">&#39;VARIATION&#39;</span>,    <span class="stringliteral">&#39;COMPARA&#39;</span>,
                   <span class="stringliteral">&#39;INFO_ADAPTOR&#39;</span>, <span class="stringliteral">&#39;FORCE_UPDATE&#39;</span> ],
                 @args );
  $self-&gt;{logger} = get_logger();
  <span class="keywordflow">return</span> $self;
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a67a558d1578d2343cd2393807fe6394c"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::MetaDataProcessor::process_compara" ref="a67a558d1578d2343cd2393807fe6394c" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a> Bio::EnsEMBL::MetaData::MetaDataProcessor::process_compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::Compara::DBSQL::DBAdaptor
  Arg        : Arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
  Description: Process supplied compara
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-process_compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-process_compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-process_compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-process_compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a67a558d1578d2343cd2393807fe6394c">process_compara</a> {
  my ( $self, $compara, $genomes ) = @_;
  <span class="keywordflow">if</span> ( !defined $genomes ) {
    $genomes = {};
  }
  my $comparas = [];
  eval {
    $self-&gt;{logger}
      -&gt;info( <span class="stringliteral">&quot;Processing compara database &quot;</span> . $compara-&gt;dbc()-&gt;dbname() );

    ( my $division = $compara-&gt;dbc()-&gt;dbname() ) =~
      s/ensembl_compara_([a-z_]+)_[0-9]+_[0-9]+/$1/;

    $division = $DIVISION_NAMES-&gt;{$division} || $division;

    my $adaptor = $compara-&gt;get_MethodLinkSpeciesSetAdaptor();

    <span class="keywordflow">for</span> my $method (
      qw/PROTEIN_TREES BLASTZ_NET LASTZ_NET TRANSLATED_BLAT TRANSLATED_BLAT_NET SYNTENY/
      )
    {

      $self-&gt;{logger}
        -&gt;info( <span class="stringliteral">&quot;Processing method type $method from compara database &quot;</span> .
                $compara-&gt;dbc()-&gt;dbname() );

<span class="preprocessor">      # group by species_set</span>
<span class="preprocessor"></span>      my $mlss_by_ss = {};
      <span class="keywordflow">for</span> my $mlss ( @{ $adaptor-&gt;fetch_all_by_method_link_type($method) } ) {
        push @{ $mlss_by_ss-&gt;{ $mlss-&gt;species_set_obj()-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1BaseInfo.html#addea948253febfab2a41e517b3dec92e">dbID</a>() } }, $mlss;
      }

      <span class="keywordflow">for</span> my $mlss_list ( values %$mlss_by_ss ) {

        my $dbs = {};
        my $ss_name;
        <span class="keywordflow">for</span> my $mlss ( @{$mlss_list} ) {

          $ss_name ||= $mlss-&gt;species_set_obj()-&gt;get_tagvalue(<span class="stringliteral">&#39;name&#39;</span>);
          <span class="keywordflow">for</span> my $gdb ( @{ $mlss-&gt;species_set_obj()-&gt;genome_dbs() } ) {
            $dbs-&gt;{ $gdb-&gt;name() } = $gdb;
          }
          <span class="keywordflow">if</span> ( !defined $ss_name ) {
            $ss_name = $mlss-&gt;name();
          }
          <span class="keywordflow">if</span> ( defined $ss_name ) {
            last;
          }
        }

        $self-&gt;{logger}-&gt;info(
<span class="stringliteral">&quot;Processing species set $ss_name for method $method from compara database &quot;</span>
            . $compara-&gt;dbc()-&gt;dbname() );

        my $compara_info =
          <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html#af724cfb183b4763d5f823acf84ae68a0">new</a>(
                                           -DBNAME =&gt; $compara-&gt;dbc()-&gt;dbname(),
                                           -DIVISION =&gt; $division,
                                           -METHOD   =&gt; $method,
                                           -SET_NAME =&gt; $ss_name,
                                           -GENOMES  =&gt; [] );

        <span class="keywordflow">for</span> my $gdb ( values %{$dbs} ) {

          $self-&gt;{logger}-&gt;info( <span class="stringliteral">&quot;Processing species &quot;</span> . $gdb-&gt;name() .
<span class="stringliteral">&quot; from species set $ss_name for method $method from compara database &quot;</span>
            . $compara-&gt;dbc()-&gt;dbname() );

          my $genomeInfo = $genomes-&gt;{ $gdb-&gt;name() };

<span class="preprocessor">          # have we got one in the database already?</span>
<span class="preprocessor"></span>          <span class="keywordflow">if</span> ( !defined $genomeInfo &amp;&amp; defined $self-&gt;{info_adaptor} ) {
            $self-&gt;{logger}-&gt;debug(<span class="stringliteral">&quot;Checking in the database&quot;</span>);
            $genomeInfo = $self-&gt;{info_adaptor}-&gt;fetch_by_name( $gdb-&gt;name() );

            <span class="keywordflow">if</span> ( !defined $genomeInfo ) {
              my $current_release = $self-&gt;{info_adaptor}-&gt;data_release();
              <span class="keywordflow">if</span> ( defined $current_release-&gt;ensembl_genomes_version() ) {
<span class="preprocessor">                # try the ensembl release</span>
<span class="preprocessor"></span>                my $ensembl_release =
                  $self-&gt;{info_adaptor}-&gt;db()-&gt;get_DataReleaseInfoAdaptor()
                  -&gt;fetch_by_ensembl_release(
                                          $current_release-&gt;ensembl_version() );
                $self-&gt;{info_adaptor}-&gt;data_release($ensembl_release);
                $genomeInfo =
                  $self-&gt;{info_adaptor}-&gt;fetch_by_name( $gdb-&gt;name() );
                $self-&gt;{info_adaptor}-&gt;data_release($current_release);
              }
            }

            <span class="keywordflow">if</span> ( !defined $genomeInfo ) {
              croak <span class="stringliteral">&quot;Could not find genome info object for &quot;</span> . $gdb-&gt;name();
            }
            $self-&gt;{logger}-&gt;debug(<span class="stringliteral">&quot;Got one from the database&quot;</span>);
            $genomes-&gt;{ $gdb-&gt;name() } = $genomeInfo;
          } ## end <span class="keywordflow">if</span> ( !defined $genomeInfo...)

<span class="preprocessor">          # last attempt, create one</span>
<span class="preprocessor"></span>          <span class="keywordflow">if</span> ( !defined $genomeInfo ) {
            $self-&gt;{logger}-&gt;info( <span class="stringliteral">&quot;Creating info object for &quot;</span> . $gdb-&gt;name() );

<span class="preprocessor">            # get core dba</span>
<span class="preprocessor"></span>            my $dba;
            eval {
              $dba =
                Bio::EnsEMBL::Registry-&gt;get_DBAdaptor( $gdb-&gt;name(), <span class="stringliteral">&#39;core&#39;</span> );
            };
            <span class="keywordflow">if</span> ( defined $dba ) {
              $genomeInfo = $self-&gt;process_genome( { core =&gt; $dba } );
            }
            <span class="keywordflow">else</span> {
              croak <span class="stringliteral">&quot;Could not find DBAdaptor for &quot;</span> . $gdb-&gt;name();
            }
            $genomeInfo-&gt;base_count(0);
            <span class="keywordflow">if</span> ( !defined $genomeInfo ) {
              croak <span class="stringliteral">&quot;Could not create a genome info object for &quot;</span> . $gdb-&gt;name();
            }
            $genomes-&gt;{ $gdb-&gt;name() } = $genomeInfo;
          }
          croak <span class="stringliteral">&quot;Could not find genome info for &quot;</span> . $gdb-&gt;name()
            unless defined $genomeInfo;

          push @{ $compara_info-&gt;genomes() }, $genomeInfo;

          <span class="keywordflow">if</span> ( !defined $genomeInfo-&gt;compara() ) {
            $genomeInfo-&gt;compara( [$compara_info] );
          }
          <span class="keywordflow">else</span> {
            push @{ $genomeInfo-&gt;compara() }, $compara_info;
          }
        } ## end <span class="keywordflow">for</span> my $gdb ( values %{...})
        $self-&gt;{logger}
          -&gt;info(<span class="stringliteral">&quot;Adding compara info to list of analyses to store&quot;</span>);
        push @$comparas, $compara_info <span class="keywordflow">if</span> defined $compara_info;
      } ## end <span class="keywordflow">for</span> my $mlss_list ( values...)
    } ## end <span class="keywordflow">for</span> my $method ( ...)

    $self-&gt;{logger}-&gt;info(
         <span class="stringliteral">&quot;Completed processing compara database &quot;</span> . $compara-&gt;dbc()-&gt;dbname() );
  };    ## end eval
  <span class="keywordflow">if</span> ($@) {
    $self-&gt;{logger}-&gt;warn( <span class="stringliteral">&quot;Could not process compara: &quot;</span> . $@ );
  }
  <span class="keywordflow">return</span> $comparas;
} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a67a558d1578d2343cd2393807fe6394c">process_compara</a>
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a0edad57751ae8fbe56ab226aafb33c18"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::MetaDataProcessor::process_genome" ref="a0edad57751ae8fbe56ab226aafb33c18" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public <a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a> Bio::EnsEMBL::MetaData::MetaDataProcessor::process_genome </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Hashref of DBAdaptors
  Description: Process supplied genome
  Returntype : <a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-process_genome" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-process_genome-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-process_genome-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-process_genome-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a0edad57751ae8fbe56ab226aafb33c18">process_genome</a> {
  my ( $self, $dbas ) = @_;
  my $dba = $dbas-&gt;{core};
  <span class="keywordflow">if</span> ( !defined $dba ) {
    confess <span class="stringliteral">&quot;DBA not defined for processing&quot;</span>;
  }
  $dba-&gt;dbc()-&gt;reconnect_when_lost(1);

<span class="preprocessor">  # get metadata container</span>
<span class="preprocessor"></span>  my $meta   = $dba-&gt;get_MetaContainer();
  my $dbname = $dba-&gt;dbc()-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#aa984387c2ded214125ccf83191ae2226">dbname</a>();
  my $size   = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#abf2e21bfe55cbfb5b2d3f28b10560659">get_dbsize</a>($dba);
  my $tableN =
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
        -SQL =&gt;
          <span class="stringliteral">&quot;select count(*) from information_schema.tables where table_schema=?&quot;</span>,
        -PARAMS =&gt; [$dbname] );

  my $strain      = $meta-&gt;single_value_by_key(<span class="stringliteral">&#39;species.strain&#39;</span>);
  my $serotype    = $meta-&gt;single_value_by_key(<span class="stringliteral">&#39;species.serotype&#39;</span>);
  my $name        = $meta-&gt;get_display_name();
  my $taxonomy_id = $meta-&gt;get_taxonomy_id();
  my $species_taxonomy_id =
    $meta-&gt;single_value_by_key(<span class="stringliteral">&#39;species.species_taxonomy_id&#39;</span>) || $taxonomy_id;
  my $assembly_accession = $meta-&gt;single_value_by_key(<span class="stringliteral">&#39;assembly.accession&#39;</span>);
  my $assembly_name      = $meta-&gt;single_value_by_key(<span class="stringliteral">&#39;assembly.name&#39;</span>);
  my $genebuild          = $meta-&gt;single_value_by_key(<span class="stringliteral">&#39;genebuild.start_date&#39;</span>);

<span class="preprocessor">  # get highest assembly level</span>
<span class="preprocessor"></span>  my ($assembly_level) =
    @{
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_simple(
         -SQL =&gt;
           <span class="stringliteral">&#39;select name from coord_system where species_id=? order by rank asc&#39;</span>,
         -PARAMS =&gt; [ $dba-&gt;species_id() ] ) };
  my $division  = <span class="stringliteral">&#39;Ensembl&#39;</span>;
  my @divisions = sort @{ $meta-&gt;list_value_by_key(<span class="stringliteral">&#39;species.division&#39;</span>) };
  <span class="keywordflow">if</span> ( scalar @divisions &gt; 0 ) {
    $division = $divisions[-1];
  }

  my $md =
    <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#a8a7161c7738910c10ba3e8cd9d33b7bf">new</a>(
                                   -name         =&gt; $dba-&gt;species(),
                                   -species_id   =&gt; $dba-&gt;species_id(),
                                   -division     =&gt; $division,
                                   -dbname       =&gt; $dbname,
                                   -data_release =&gt; $self-&gt;{data_release},
                                   -strain       =&gt; $strain,
                                   -serotype     =&gt; $serotype,
                                   -display_name =&gt; $name,
                                   -taxonomy_id  =&gt; $taxonomy_id,
                                   -species_taxonomy_id =&gt; $species_taxonomy_id,
                                   -assembly_accession  =&gt; $assembly_accession,
                                   -assembly_name       =&gt; $assembly_name,
                                   -genebuild           =&gt; $genebuild,
                                   -assembly_level      =&gt; $assembly_level );

<span class="preprocessor">  # get list of seq names</span>
<span class="preprocessor"></span>  my $seqs_arr = [];

  <span class="keywordflow">if</span> ( defined $self-&gt;{contigs} ) {
    my $seqs = {};

<span class="preprocessor">    # 1. get complete list of seq_regions as a hash vs. ENA synonyms</span>
<span class="preprocessor"></span>    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_no_return(
      -SQL =&gt; q/select distinct s.name, ss.synonym 
      from coord_system c  
      join seq_region s using (coord_system_id)  
      left join seq_region_synonym ss on 
        (ss.seq_region_id=s.seq_region_id and ss.external_db_id in 
            (select external_db_id from external_db where db_name=<span class="stringliteral">&#39;INSDC&#39;</span>)) 
      where c.species_id=? and attrib like <span class="stringliteral">&#39;%default_version%&#39;</span>/,
      -PARAMS   =&gt; [ $dba-&gt;species_id() ],
      -CALLBACK =&gt; sub {
        my ( $name, $acc ) = @{ shift @_ };
        $seqs-&gt;{$name} = $acc;
        <span class="keywordflow">return</span>;
      } );

<span class="preprocessor">    # 2. add accessions where the name is flagged as being in ENA</span>
<span class="preprocessor"></span>    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_no_return(
      -SQL =&gt; q/
      select s.name 
      from coord_system c  
      join seq_region s using (coord_system_id)       
      join seq_region_attrib sa <span class="keyword">using</span> (seq_region_id)  
      where sa.value=<span class="stringliteral">&#39;ENA&#39;</span> and c.species_id=? and attrib like <span class="stringliteral">&#39;%default_version%&#39;</span>/,
      -PARAMS   =&gt; [ $dba-&gt;species_id() ],
      -CALLBACK =&gt; sub {
        my ($acc) = @{ shift @_ };
        $seqs-&gt;{$acc} = $acc;
        <span class="keywordflow">return</span>;
      } );

    <span class="keywordflow">while</span> ( my ( $key, $acc ) = each %$seqs ) {
      push @$seqs_arr, { name =&gt; $key, acc =&gt; $acc };
    }
  } ## end <span class="keywordflow">if</span> ( defined $self-&gt;{contigs...})

  $md-&gt;assembly()-&gt;sequences($seqs_arr);

<span class="preprocessor">  # get toplevel base count</span>
<span class="preprocessor"></span>  my $base_counts =
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_simple(
    -SQL =&gt;
q/select value from genome_statistics where statistic=<span class="stringliteral">&#39;ref_length&#39;</span> and species_id=?/,
    -PARAMS =&gt; [ $dba-&gt;species_id() ] );

  <span class="keywordflow">if</span> ( scalar @$base_counts == 0 ) {
    $base_counts = $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_simple(
      -SQL =&gt; q/select sum(s.length) from seq_region s 
join coord_system c using (coord_system_id) 
join seq_region_attrib sa <span class="keyword">using</span> (seq_region_id) 
join attrib_type a <span class="keyword">using</span> (attrib_type_id) 
where a.code=<span class="stringliteral">&#39;toplevel&#39;</span> and species_id=?/,
      -PARAMS =&gt; [ $dba-&gt;species_id() ] );
  }

  $md-&gt;assembly()-&gt;base_count( $base_counts-&gt;[0] );

<span class="preprocessor">  # get associated PMIDs</span>
<span class="preprocessor"></span>  $md-&gt;organism()-&gt;publications(
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_simple(
      -SQL =&gt; q/select distinct dbprimary_acc from 
      xref
      join external_db <span class="keyword">using</span> (external_db_id)
      join seq_region_attrib sa on (xref.xref_id=sa.value)
      join attrib_type using (attrib_type_id)
      join seq_region <span class="keyword">using</span> (seq_region_id)
      join coord_system <span class="keyword">using</span> (coord_system_id)
      where species_id=? and code=<span class="stringliteral">&#39;xref_id&#39;</span> and db_name in (<span class="stringliteral">&#39;PUBMED&#39;</span>)/,
      -PARAMS =&gt; [ $dba-&gt;species_id() ] ) );

<span class="preprocessor">  # add aliases</span>
<span class="preprocessor"></span>  $md-&gt;organism()-&gt;aliases(
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_simple(
      -SQL =&gt; q/select distinct meta_value from meta
      where species_id=? and meta_key=<span class="stringliteral">&#39;species.alias&#39;</span>/,
      -PARAMS =&gt; [ $dba-&gt;species_id() ] ) );

  <span class="keywordflow">if</span> ( defined $self-&gt;{annotation_analyzer} ) {

<span class="preprocessor">    # core annotation</span>
<span class="preprocessor"></span>    $self-&gt;{logger}
      -&gt;info( <span class="stringliteral">&quot;Processing &quot;</span> . $dba-&gt;species() . <span class="stringliteral">&quot; core annotation&quot;</span> );
    $md-&gt;annotations( $self-&gt;{annotation_analyzer}-&gt;analyze_annotation($dba) );

<span class="preprocessor">    # features</span>
<span class="preprocessor"></span>    my $core_ali  = $self-&gt;{annotation_analyzer}-&gt;analyze_alignments($dba);
    my $other_ali = {};
    $md-&gt;features( $self-&gt;{annotation_analyzer}-&gt;analyze_features($dba) );
    my $other_features = $dbas-&gt;{otherfeatures};
    <span class="keywordflow">if</span> ( defined $other_features ) {
      $self-&gt;{logger}
        -&gt;info( <span class="stringliteral">&quot;Processing &quot;</span> . $dba-&gt;species() . <span class="stringliteral">&quot; otherfeatures annotation&quot;</span> );
      my %features = ( %{ $md-&gt;features() },
                       %{$self-&gt;{annotation_analyzer}
                           -&gt;analyze_features($other_features) } );
      $other_ali =
        $self-&gt;{annotation_analyzer}-&gt;analyze_alignments($other_features);
      $size += <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#abf2e21bfe55cbfb5b2d3f28b10560659">get_dbsize</a>($other_features);
      $md-&gt;features( \%features );
      $md-&gt;add_database( $other_features-&gt;dbc()-&gt;dbname() );
    }
    my $variation = $dbas-&gt;{variation};

<span class="preprocessor">    # variation</span>
<span class="preprocessor"></span>    <span class="keywordflow">if</span> ( defined $variation ) {
      $self-&gt;{logger}
        -&gt;info( <span class="stringliteral">&quot;Processing &quot;</span> . $dba-&gt;species() . <span class="stringliteral">&quot; variation annotation&quot;</span> );
      $md-&gt;variations(
                  $self-&gt;{annotation_analyzer}-&gt;analyze_variation($variation) );
      $size += <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#abf2e21bfe55cbfb5b2d3f28b10560659">get_dbsize</a>($variation);
      $md-&gt;add_database( $variation-&gt;dbc()-&gt;dbname() );
    }

    my $funcgen = $dbas-&gt;{funcgen};
    <span class="keywordflow">if</span> ( defined $funcgen ) {
      $self-&gt;{logger}
        -&gt;info( <span class="stringliteral">&quot;Processing &quot;</span> . $dba-&gt;species() . <span class="stringliteral">&quot; funcgen annotation&quot;</span> );
      $md-&gt;add_database( $funcgen-&gt;dbc()-&gt;dbname() );
    }

<span class="preprocessor">    # BAM</span>
<span class="preprocessor"></span>    $self-&gt;{logger}
      -&gt;info( <span class="stringliteral">&quot;Processing &quot;</span> . $dba-&gt;species() . <span class="stringliteral">&quot; read aligments&quot;</span> );
    my $read_ali =
      $self-&gt;{annotation_analyzer}
      -&gt;analyze_tracks( $md-&gt;name(), $md-&gt;division() );
    my %all_ali = ( %{$core_ali}, %{$other_ali} );

<span class="preprocessor">    # add bam tracks by count - use source name</span>
<span class="preprocessor"></span>    <span class="keywordflow">for</span> my $bam ( @{ $read_ali-&gt;{bam} } ) {
      $all_ali{bam}{ $bam-&gt;{<span class="keywordtype">id</span>} }++;
    }
    $md-&gt;other_alignments( \%all_ali );
    $md-&gt;db_size($size);

  } ## end <span class="keywordflow">if</span> ( defined $self-&gt;{annotation_analyzer...})
  $dba-&gt;dbc()-&gt;disconnect_if_idle();
  <span class="keywordflow">return</span> $md;
} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a0edad57751ae8fbe56ab226aafb33c18">process_genome</a>
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a9d37c6b5f65e890a49169ce66acebfaa"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::MetaDataProcessor::process_metadata" ref="a9d37c6b5f65e890a49169ce66acebfaa" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Arrayref Bio::EnsEMBL::MetaData::MetaDataProcessor::process_metadata </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Arrayref of DBAdaptors
  Description: Process supplied DBAdaptors
  Returntype : Arrayref of <a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-process_metadata" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-process_metadata-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-process_metadata-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-process_metadata-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a9d37c6b5f65e890a49169ce66acebfaa">process_metadata</a> {
  my ( $self, $dbas ) = @_;

<span class="preprocessor">  # 1. create hash of DBAs</span>
<span class="preprocessor"></span>  my $dba_hash = {};
  my $comparas = [];
  <span class="keywordflow">for</span> my $dba ( grep { $_-&gt;dbc()-&gt;dbname() !~ /ancestral/ } @{$dbas} ) {
    my $type;
    my $species = $dba-&gt;species();
    <span class="keywordflow">for</span> my $t (qw(core otherfeatures variation funcgen)) {
      <span class="keywordflow">if</span> ( $dba-&gt;dbc()-&gt;dbname() =~ m/$t/ ) {
        $type = $t;
        last;
      }
    }
    <span class="keywordflow">if</span> ( defined $type ) {
      $dba_hash-&gt;{$species}{$type} = $dba <span class="keywordflow">if</span> defined $type;
    }
    elsif ( $dba-&gt;dbc()-&gt;dbname() =~ m/_compara_/ ) {
      push @$comparas, $dba;
    }
  }

<span class="preprocessor">  # 2. iterate through each genome</span>
<span class="preprocessor"></span>  my $genome_infos = {};
  my $n            = 0;
  my $total        = scalar( keys %$dba_hash );
  <span class="keywordflow">while</span> ( my ( $genome, $dbas ) = each %$dba_hash ) {
    $self-&gt;{logger}-&gt;info( <span class="stringliteral">&quot;Processing &quot;</span> . $genome . <span class="stringliteral">&quot; (&quot;</span> . ++$n . <span class="stringliteral">&quot;/$total)&quot;</span> );
    $genome_infos-&gt;{$genome} = $self-&gt;process_genome($dbas);
  }

<span class="preprocessor">  # 3. apply compara</span>
<span class="preprocessor"></span>  <span class="keywordflow">for</span> my $compara (@$comparas) {
    $self-&gt;process_compara( $compara, $genome_infos );
  }

  <span class="keywordflow">return</span> [ values(%$genome_infos) ];
} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a9d37c6b5f65e890a49169ce66acebfaa">process_metadata</a>
</pre></div> </div> 
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>Bio/EnsEMBL/MetaData/<a class="el" href="MetaDataProcessor_8pm_source.html">MetaDataProcessor.pm</a></li>
</ul>
</div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="namespaceBio.html">Bio</a>      </li>
      <li class="navelem"><a class="el" href="namespaceBio_1_1EnsEMBL.html">EnsEMBL</a>      </li>
      <li class="navelem"><a class="el" href="namespaceBio_1_1EnsEMBL_1_1MetaData.html">MetaData</a>      </li>
      <li class="navelem"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html">MetaDataProcessor</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
