<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Ensembl Metadata: Bio::EnsEMBL::MetaData::AnnotationAnalyzer Class Reference</title>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="inherits.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="annotated.html"><span>Full&#160;Class&#160;List</span></a></li>
      <li><a href="functions.html"><span>Methods&#160;from&#160;all&#160;packages</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="summary">
<a href="#pub-methods">Available Methods</a>  </div>
  <div class="headertitle">
<div class="title">Bio::EnsEMBL::MetaData::AnnotationAnalyzer Class Reference</div>  </div>
</div>
<div class="contents">
<!-- doxytag: class="Bio::EnsEMBL::MetaData::AnnotationAnalyzer" -->
<p><a href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer-members.html">List of all members.</a></p>
<hr/><a name="details" id="details"></a><h2>Class Summary</h2>
<div class="textblock"><h2><a class="anchor" id="Synopsis"></a>
Synopsis</h2>
<div class="fragment"><pre class="fragment">my $anal = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html">Bio::EnsEMBL::MetaData::AnnotationAnalyzer</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a5f4168a9b1eaea0d3834ced9f9949cc8">new</a>();
my $details = $anal-&gt;analyze_annotation($dba);
</pre></div> <h2><a class="anchor" id="Description"></a>
Description</h2>
 <pre>
Utility class for counting xrefs, variations etc. associated with a genome

</pre> 
<p>Definition at line <a class="el" href="AnnotationAnalyzer_8pm_source.html#l00021">21</a> of file <a class="el" href="AnnotationAnalyzer_8pm_source.html">AnnotationAnalyzer.pm</a>.</p>
</div><table class="memberdecls">
<tr><td colspan="2"><h2><a name="pub-methods"></a>
Available Methods</h2></td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hash&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a042490ef3b46395a294a5a9c35e11120">analyze_alignments</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hash&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a7edeceef17b3ff1c761dc20f8f64b281">analyze_annotation</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hash&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#aa0cd3200a6539ad8d1b85a5686f37525">analyze_compara</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hash&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a0716a89a2ac3527d9379b7f7d38dd738">analyze_features</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hash&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#aebfb6fd82021d7730d548cccea941d68">analyze_tracks</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hash&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a18c347e3e223473fdadf22b94ac3db97">analyze_variation</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a28ba9929a4f20f16d83198d866e54d58">count_by_biotype</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a47985aa409cda7c9bf24a4852d8da311">count_by_interpro</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a282e53d51f66d3bddf0a484e13b2c253">count_by_xref</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hashref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a968cb097a84aa3997572cf6830432db5">count_features</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hashref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a3a34c2301e88a8bf222224e123e4f192">count_genotypes</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#af1c956c14c3e1340fe9700de173e81c6">count_interpro</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a79f6ba911fda918bc0f3bf705e728507">count_interpro_domains</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a26a752422284398e530716f9513b93a6">count_phenotypes</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hashref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a5c616639af76865a3d01da32ac9c8685">count_structural_variations</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#af30433391a1972fe05b0dd44f11ed72b">count_toplevel</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Hashref&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a4599cef2d5fae4032837d7f812b394c0">count_variations</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public Int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#aeed885c6354d1bfb7686e0913752f285">count_xrefs</a> ()</td></tr>
<tr><td class="memItemLeft" align="right" valign="top">public&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a5f4168a9b1eaea0d3834ced9f9949cc8">new</a> ()</td></tr>
</table>
<hr/><h2>Method Documentation</h2>
<a class="anchor" id="a042490ef3b46395a294a5a9c35e11120"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_alignments" ref="a042490ef3b46395a294a5a9c35e11120" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hash Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_alignments </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Description: Analyzes alignments for sequences to the assembly 
  Returntype : Hash ref - keys are source_type (e.g. RNA seq), values are arrayrefs of hashrefs containing track details (source_name,description,source_url)
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-analyze_alignments" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-analyze_alignments-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-analyze_alignments-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-analyze_alignments-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a042490ef3b46395a294a5a9c35e11120">analyze_alignments</a> {
  my ( $self, $dba ) = @_;
  my $ali = {};
  my $pf = $self-&gt;count_features( $dba, <span class="stringliteral">&#39;protein_align_feature&#39;</span> );
  <span class="keywordflow">if</span> ( scalar( keys %$pf ) &gt; 0 ) {
    $ali-&gt;{proteinAlignFeatures} = $pf;
  }
  my $df = $self-&gt;count_features( $dba, <span class="stringliteral">&#39;dna_align_feature&#39;</span> );
  <span class="keywordflow">if</span> ( scalar( keys %$df ) &gt; 0 ) {
    $ali-&gt;{dnaAlignFeatures} = $df;
  }
  <span class="keywordflow">return</span> $ali;
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a7edeceef17b3ff1c761dc20f8f64b281"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_annotation" ref="a7edeceef17b3ff1c761dc20f8f64b281" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hash Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_annotation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Description: Analyzes annotation content of the supplied core DB
  Returntype : 	Hash ref with the following keys:
					nProteinCoding - number of protein coding genes
					nProteinCodingUniProtKBSwissProt - number of protein coding genes with at least one SwissPROT entry
					nProteinCodingUniProtKBTrEMBL - number of protein coding genes with at least one TrEMBL entry
					nProteinCodingGO - number of protein coding genes with at least one GO term
					nProteinCodingInterPro - number of protein coding genes with at least one InterPro match
					nUniProtKBSwissProt - number of distinct UniProtKB/TrEMBL entries matching at least one translation
					nUniProtKBTrEMBL - number of distinct UniProtKB/TrEMBL entries matching at least one translation
					nGO - number of distinct GO terms matching at least one translation
					nInterPro - number of distinct InterPro entries matching at least one feature
					nInterProDomains - number of distinct protein features matching an InterPro domains
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-analyze_annotation" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-analyze_annotation-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-analyze_annotation-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-analyze_annotation-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a7edeceef17b3ff1c761dc20f8f64b281">analyze_annotation</a> {
  my ( $self, $dba ) = @_;
  $self-&gt;{logger}-&gt;debug( <span class="stringliteral">&quot;Analysing annotation for &quot;</span> . $dba-&gt;species() );
  <span class="keywordflow">return</span> {
      nProteinCoding =&gt; $self-&gt;count_by_biotype( $dba, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
      nProteinCodingGO =&gt; $self-&gt;count_by_xref( $dba, <span class="stringliteral">&#39;GO&#39;</span>, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
      nProteinCodingUniProtKB =&gt;
        $self-&gt;count_by_xref( $dba, [ <span class="stringliteral">&#39;Uniprot/SWISSPROT&#39;</span>, <span class="stringliteral">&#39;Uniprot/SPTREMBL&#39;</span> ],
                              <span class="stringliteral">&#39;protein_coding&#39;</span> ),
      nProteinCodingUniProtKBSwissProt =&gt;
        $self-&gt;count_by_xref( $dba, <span class="stringliteral">&#39;Uniprot/SWISSPROT&#39;</span>, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
      nProteinCodingUniProtKBTrEMBL =&gt;
        $self-&gt;count_by_xref( $dba, <span class="stringliteral">&#39;Uniprot/SPTREMBL&#39;</span>, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
      nProteinCodingInterPro =&gt; $self-&gt;count_by_interpro($dba),
      nGO =&gt; $self-&gt;count_by_xref( $dba, <span class="stringliteral">&#39;GO&#39;</span>, <span class="stringliteral">&#39;protein_coding&#39;</span> ),
      nUniProtKBSwissProt =&gt; $self-&gt;count_xrefs( $dba, <span class="stringliteral">&#39;Uniprot/SWISSPROT&#39;</span> ),
      nUniProtKBTrEMBL    =&gt; $self-&gt;count_xrefs( $dba, <span class="stringliteral">&#39;Uniprot/SPTREMBL&#39;</span> ),
      nInterPro           =&gt; $self-&gt;count_interpro($dba),
      nInterProDomains =&gt; $self-&gt;count_interpro_domains($dba) };
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="aa0cd3200a6539ad8d1b85a5686f37525"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_compara" ref="aa0cd3200a6539ad8d1b85a5686f37525" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hash Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_compara </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Arg        : Bio::EnsEMBL::Compara::DBSQL::DBAdaptor
  Description: Analyzes compara methods found in the supplied compara database involving the supplied core species
  Returntype : 	Hash ref (keys are method names, values are arrays of partner species)
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-analyze_compara" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-analyze_compara-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-analyze_compara-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-analyze_compara-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#aa0cd3200a6539ad8d1b85a5686f37525">analyze_compara</a> {
  my ( $self, $dba, $core ) = @_;
  my $compara = {};
  eval {
    my $gdb =
      $dba-&gt;get_GenomeDBAdaptor()-&gt;fetch_by_registry_name( $core-&gt;species() );
    <span class="keywordflow">for</span> my $mlss (
       @{ $dba-&gt;get_MethodLinkSpeciesSetAdaptor()-&gt;fetch_all_by_GenomeDB($gdb) }
      )
    {
      my $t = $mlss-&gt;method()-&gt;type();
      next
        <span class="keywordflow">if</span> ( $t eq <span class="stringliteral">&#39;FAMILY&#39;</span> ||
             $t eq <span class="stringliteral">&#39;ENSEMBL_ORTHOLOGUES&#39;</span> ||
             $t eq <span class="stringliteral">&#39;ENSEMBL_PARALOGUES&#39;</span> );
      <span class="keywordflow">for</span> my $gdb2 ( grep { $gdb-&gt;dbID() ne $_-&gt;dbID() }
                     @{ $mlss-&gt;species_set_obj-&gt;genome_dbs() } )
      {
        push( @{ $compara-&gt;{$t} }, $gdb2-&gt;name() );
      }
    }
  };
  <span class="keywordflow">if</span> ($@) {
    warn <span class="stringliteral">&quot;No compara entry found for &quot;</span> . $dba-&gt;species();
  }
  <span class="keywordflow">return</span> $compara;
} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#aa0cd3200a6539ad8d1b85a5686f37525">analyze_compara</a>
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a0716a89a2ac3527d9379b7f7d38dd738"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_features" ref="a0716a89a2ac3527d9379b7f7d38dd738" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hash Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_features </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Description: Analyzes features found in the supplied core/otherfeatures database (simple_feature,repeat_feature,protein_align_feature,dna_align_features)
  Returntype : 	Hash ref (keys are feature tables, values are hashrefs of count by analysis name)
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-analyze_features" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-analyze_features-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-analyze_features-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-analyze_features-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a0716a89a2ac3527d9379b7f7d38dd738">analyze_features</a> {
  my ( $self, $dba ) = @_;
  <span class="keywordflow">return</span> { simpleFeatures =&gt; $self-&gt;count_features( $dba, <span class="stringliteral">&#39;simple_feature&#39;</span> ),
           repeatFeatures =&gt; $self-&gt;count_features( $dba, <span class="stringliteral">&#39;repeat_feature&#39;</span> ) };
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="aebfb6fd82021d7730d548cccea941d68"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_tracks" ref="aebfb6fd82021d7730d548cccea941d68" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hash Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_tracks </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Description: Analyzes tracks attached to the supplied core database
  Returntype : Hash ref - keys are source_type (e.g. RNA seq), values are arrayrefs of hashrefs containing track details (source_name,description,source_url)
  Exceptions : none
  Caller     : general
  Status     : Stable	
 </pre><div id="codesection-analyze_tracks" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-analyze_tracks-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-analyze_tracks-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-analyze_tracks-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#aebfb6fd82021d7730d548cccea941d68">analyze_tracks</a> {
  my ( $self, $species, $division ) = @_;
  $species = ucfirst($species);
  ( $division = lc $division ) =~ s/ensembl<span class="comment">//;</span>
  # <span class="keyword">get</span> the ini file from git
  ( my $ini_url = $url_template ) =~ s/SPECIES/$species/;
  $ini_url =~ s/DIVISION/$division/;

  my $req = HTTP::Request-&gt;new( GET =&gt; $ini_url );
<span class="preprocessor">  # Pass request to the user agent and get a response back</span>
<span class="preprocessor"></span>  my $res = $ua-&gt;request($req);
  my $ini;
<span class="preprocessor">  # Check the outcome of the response</span>
<span class="preprocessor"></span>  <span class="keywordflow">if</span> ( $res-&gt;is_success ) {
    $ini = $res-&gt;content;
  }
  <span class="keywordflow">else</span> {
    $self-&gt;{logger}
      -&gt;debug( <span class="stringliteral">&quot;Could not retrieve $ini_url: &quot;</span> . $res-&gt;status_line );
  }

<span class="preprocessor"># parse out and look at:</span>
<span class="preprocessor"></span><span class="preprocessor"># [ENSEMBL_INTERNAL_BAM_SOURCES]</span>
<span class="preprocessor"></span><span class="preprocessor"># 1_Puccinia_triticina_SRR035315 = dna_align_est</span>
<span class="preprocessor"></span><span class="preprocessor"># walk through keys and use keys to find sections like:</span>
<span class="preprocessor"></span><span class="preprocessor"># [1_Puccinia_triticina_SRR035315]</span>
<span class="preprocessor"></span><span class="preprocessor"># source_name        = Transcriptomics sequences of fresh spores. Run id: SRR035315</span>
<span class="preprocessor"></span><span class="preprocessor"># description        = RNA-Seq study.</span>
<span class="preprocessor"></span><span class="preprocessor"># source_url         = http://ftp.sra.ebi.ac.uk/vol1/ERZ000/ERZ000002/SRR035315.bam</span>
<span class="preprocessor"></span><span class="preprocessor"># source_type        = rnaseq</span>
<span class="preprocessor"></span><span class="preprocessor"># display            = normal</span>
<span class="preprocessor"></span><span class="preprocessor"># then store some or all of this in my output e.g. {bam}{source_type}[{source_name,description,source_url}]</span>
<span class="preprocessor"></span>  my $bams = {};
  <span class="keywordflow">if</span> ( defined $ini ) {
    my $cfg = Config::IniFiles-&gt;new( -file =&gt; \$ini );
    <span class="keywordflow">if</span> ( defined $cfg ) {
      <span class="keywordflow">for</span> my $bam ( $cfg-&gt;Parameters(<span class="stringliteral">&quot;ENSEMBL_INTERNAL_BAM_SOURCES&quot;</span>) ) {
        push @{ $bams-&gt;{ $cfg-&gt;val( $bam, <span class="stringliteral">&#39;source_type&#39;</span> ) } }, {
            <span class="keywordtype">id</span>          =&gt; $bam,
            source_name =&gt; $cfg-&gt;val( $bam, <span class="stringliteral">&#39;source_name&#39;</span> ),
            source_url  =&gt; $cfg-&gt;val( $bam, <span class="stringliteral">&#39;source_url&#39;</span> ),
            description =&gt; $cfg-&gt;val( $bam, <span class="stringliteral">&#39;description&#39;</span> ) };
      }
    }
  }
  <span class="keywordflow">return</span> $bams;
} ## end sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#aebfb6fd82021d7730d548cccea941d68">analyze_tracks</a>
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a18c347e3e223473fdadf22b94ac3db97"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_variation" ref="a18c347e3e223473fdadf22b94ac3db97" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hash Bio::EnsEMBL::MetaData::AnnotationAnalyzer::analyze_variation </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
  Description: Analyzes variation found in the supplied variation database
  Returntype : 	Hash ref with the following keys
					variations - count of variations per source
					structuralVariations - count of structural variations per source
					genotypes - count of genotypes per sample
					phenotypes - count of variation annotations per phenotype
  Exceptions : none
  Caller     : general
  Status     : Stable
 </pre><div id="codesection-analyze_variation" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-analyze_variation-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-analyze_variation-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-analyze_variation-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a18c347e3e223473fdadf22b94ac3db97">analyze_variation</a> {
  my ( $self, $dba ) = @_;
  <span class="keywordflow">return</span> { variations           =&gt; $self-&gt;count_variations($dba),
           structuralVariations =&gt; $self-&gt;count_structural_variations($dba),
           genotypes            =&gt; $self-&gt;count_genotypes($dba),
           phenotypes           =&gt; $self-&gt;count_phenotypes($dba) };
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a28ba9929a4f20f16d83198d866e54d58"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_by_biotype" ref="a28ba9929a4f20f16d83198d866e54d58" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_by_biotype </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Arg        : String (biotype)
  Description: Returns count of genes with specified biotype
  Returntype : Integer
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_by_biotype" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_by_biotype-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_by_biotype-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_by_biotype-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a28ba9929a4f20f16d83198d866e54d58">count_by_biotype</a> {
  my ( $self, $dba, $biotype ) = @_;
  <span class="keywordflow">return</span> $dba-&gt;get_GeneAdaptor()-&gt;count_all_by_biotype($biotype);
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a47985aa409cda7c9bf24a4852d8da311"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_by_interpro" ref="a47985aa409cda7c9bf24a4852d8da311" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_by_interpro </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Description: Returns count of genes with InterPro domains
  Returntype : Integer
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_by_interpro" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_by_interpro-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_by_interpro-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_by_interpro-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a47985aa409cda7c9bf24a4852d8da311">count_by_interpro</a> {
  my ( $self, $dba ) = @_;
  <span class="keywordflow">return</span>
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
                                               -SQL    =&gt; $count_by_interpro,
                                               -PARAMS =&gt; [ $dba-&gt;species_id() ]
    );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a282e53d51f66d3bddf0a484e13b2c253"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_by_xref" ref="a282e53d51f66d3bddf0a484e13b2c253" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_by_xref </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Arg        : Arrayref of dbnames 
  Arg        : (optional) String (biotype)
  Description: Returns count of genes with specified cross-reference
  Returntype : Integer
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_by_xref" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_by_xref-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_by_xref-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_by_xref-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a282e53d51f66d3bddf0a484e13b2c253">count_by_xref</a> {
  my ( $self, $dba, $db_names, $biotype ) = @_;
  $self-&gt;{logger}-&gt;debug( <span class="stringliteral">&quot;Counting genes by &quot;</span> .
                      join( <span class="stringliteral">&quot;,&quot;</span>, $db_names ) . <span class="stringliteral">&quot; xref for &quot;</span> . $dba-&gt;species() );
  $db_names = [$db_names] <span class="keywordflow">if</span> ( ref($db_names) ne <span class="stringliteral">&#39;ARRAY&#39;</span> );
  my $sql = $xref_gene_sql;
  my $db_name = join <span class="charliteral">&#39;,&#39;</span>, map { <span class="stringliteral">&quot;\&quot;$_\&quot;&quot;</span> } @$db_names;
  $sql =~ s/NAMES/$db_name/;
  my $params = [ $dba-&gt;species_id() ];
  <span class="keywordflow">if</span> ( defined $biotype ) {
    $sql .= $biotype_clause;
    push @$params, $biotype;
  }
  $self-&gt;{logger}
    -&gt;debug( <span class="stringliteral">&quot;Executing $sql with params: [&quot;</span> . join( <span class="stringliteral">&quot;,&quot;</span>, @$params ) );
  <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
    -&gt;execute_single_result( -SQL =&gt; $sql, -PARAMS =&gt; $params );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a968cb097a84aa3997572cf6830432db5"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_features" ref="a968cb097a84aa3997572cf6830432db5" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hashref Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_features </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Arg        : String - feature type (e.g. RepeatFeature)
  Description: Returns count of specified feature by analysis
  Returntype : Hashref
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_features" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_features-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_features-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_features-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a968cb097a84aa3997572cf6830432db5">count_features</a> {
  my ( $self, $dba, $table ) = @_;
  my $sql = $count_features;
  $sql =~ s/TABLE/$table/;
  <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
    -&gt;execute_into_hash( -SQL =&gt; $sql, -PARAMS =&gt; [ $dba-&gt;species_id() ] );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a3a34c2301e88a8bf222224e123e4f192"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_genotypes" ref="a3a34c2301e88a8bf222224e123e4f192" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hashref Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_genotypes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
  Description: Returns count of genotypes by source
  Returntype : Hashref
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_genotypes" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_genotypes-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_genotypes-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_genotypes-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a3a34c2301e88a8bf222224e123e4f192">count_genotypes</a> {
  my ( $self, $dba ) = @_;
  my $cnt = {};
  eval {
    $cnt =
      $dba-&gt;dbc()-&gt;sql_helper()
      -&gt;execute_into_hash( -SQL =&gt; $count_genotypes, -PARAMS =&gt; [] );
  };
  <span class="keywordflow">return</span> $cnt;
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="af1c956c14c3e1340fe9700de173e81c6"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_interpro" ref="af1c956c14c3e1340fe9700de173e81c6" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_interpro </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Description: Returns count of InterPro cross-references
  Returntype : Integer
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_interpro" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_interpro-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_interpro-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_interpro-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#af1c956c14c3e1340fe9700de173e81c6">count_interpro</a> {
  my ( $self, $dba ) = @_;
  <span class="keywordflow">return</span>
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
                                               -SQL    =&gt; $count_interpro,
                                               -PARAMS =&gt; [ $dba-&gt;species_id() ]
    );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a79f6ba911fda918bc0f3bf705e728507"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_interpro_domains" ref="a79f6ba911fda918bc0f3bf705e728507" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_interpro_domains </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Description: Returns count of InterPro domains
  Returntype : Integer
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_interpro_domains" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_interpro_domains-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_interpro_domains-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_interpro_domains-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a79f6ba911fda918bc0f3bf705e728507">count_interpro_domains</a> {
  my ( $self, $dba ) = @_;
  <span class="keywordflow">return</span>
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
                                               -SQL =&gt; $count_interpro_domains,
                                               -PARAMS =&gt; [ $dba-&gt;species_id() ]
    );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a26a752422284398e530716f9513b93a6"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_phenotypes" ref="a26a752422284398e530716f9513b93a6" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_phenotypes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
  Description: Returns count of phenotypes by name
  Returntype : Integer
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_phenotypes" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_phenotypes-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_phenotypes-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_phenotypes-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a26a752422284398e530716f9513b93a6">count_phenotypes</a> {
  my ( $self, $dba ) = @_;
  <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
    -&gt;execute_into_hash( -SQL =&gt; $count_phenotypes, -PARAMS =&gt; [] );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a5c616639af76865a3d01da32ac9c8685"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_structural_variations" ref="a5c616639af76865a3d01da32ac9c8685" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hashref Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_structural_variations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
  Description: Returns count of structural variations by source
  Returntype : Hashref
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_structural_variations" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_structural_variations-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_structural_variations-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_structural_variations-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a5c616639af76865a3d01da32ac9c8685">count_structural_variations</a> {
  my ( $self, $dba ) = @_;
  <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
    -&gt;execute_into_hash( -SQL =&gt; $count_structural_variation, -PARAMS =&gt; [] );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="af30433391a1972fe05b0dd44f11ed72b"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_toplevel" ref="af30433391a1972fe05b0dd44f11ed72b" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_toplevel </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Description: Returns count of top-level regions
  Returntype : Integer
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_toplevel" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_toplevel-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_toplevel-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_toplevel-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#af30433391a1972fe05b0dd44f11ed72b">count_toplevel</a> {
  my ( $self, $dba ) = @_;
  <span class="keywordflow">return</span>
    $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
                                               -SQL    =&gt; $count_toplevel,
                                               -PARAMS =&gt; [ $dba-&gt;species_id() ]
    );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a4599cef2d5fae4032837d7f812b394c0"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_variations" ref="a4599cef2d5fae4032837d7f812b394c0" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Hashref Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_variations </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::Variation::DBSQL::DBAdaptor
  Description: Returns count of variations by source
  Returntype : Hashref
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_variations" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_variations-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_variations-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_variations-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a4599cef2d5fae4032837d7f812b394c0">count_variations</a> {
  my ( $self, $dba ) = @_;
  <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;sql_helper()
    -&gt;execute_into_hash( -SQL =&gt; $count_variation, -PARAMS =&gt; [] );
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="aeed885c6354d1bfb7686e0913752f285"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_xrefs" ref="aeed885c6354d1bfb7686e0913752f285" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Int Bio::EnsEMBL::MetaData::AnnotationAnalyzer::count_xrefs </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<pre>  Arg        : Bio::EnsEMBL::DBSQL::DBAdaptor
  Arg        : One or more dbnames
  Description: Returns count of numbers of cross-references to supplied database(s)
  Returntype : Integer
  Exceptions : none
  Caller     : internal
  Status     : Stable
 </pre><div id="codesection-count_xrefs" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-count_xrefs-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-count_xrefs-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-count_xrefs-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#aeed885c6354d1bfb7686e0913752f285">count_xrefs</a> {
  my $self = shift;
  my $dba  = shift;
  my $tot  = 0;
  <span class="keywordflow">for</span> my $db_name (@_) {
    $tot +=
      $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
                                     -SQL    =&gt; $gene_xref_count_sql,
                                     -PARAMS =&gt; [ $dba-&gt;species_id(), $db_name ]
      );
    $tot +=
      $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
                                     -SQL    =&gt; $transcript_xref_count_sql,
                                     -PARAMS =&gt; [ $dba-&gt;species_id(), $db_name ]
      );
    $tot +=
      $dba-&gt;dbc()-&gt;sql_helper()-&gt;execute_single_result(
                                     -SQL    =&gt; $translation_xref_count_sql,
                                     -PARAMS =&gt; [ $dba-&gt;species_id(), $db_name ]
      );
  }
  <span class="keywordflow">return</span> $tot;
}
</pre></div> </div> 
</div>
</div>
<a class="anchor" id="a5f4168a9b1eaea0d3834ced9f9949cc8"></a><!-- doxytag: member="Bio::EnsEMBL::MetaData::AnnotationAnalyzer::new" ref="a5f4168a9b1eaea0d3834ced9f9949cc8" args="()" -->
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">public Bio::EnsEMBL::MetaData::AnnotationAnalyzer::new </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div>
<div class="memdoc">
<p>Undocumented method</p>
<div id="codesection-new" class="dynheader closed" style="cursor:pointer;" onclick="return toggleVisibility(this)">  
	<img id='codesection-new-trigger' src='closed.png' style='display:inline'><b>Code:</b>
</div>
<div id='codesection-new-summary' class='dyncontent' style='display:block;font-size:small;'>click to view</div>
<div id='codesection-new-content' class='dyncontent' style='display: none;'> 
 <div class="fragment"><pre class="fragment">sub <span class="keyword">new</span> {
  my $caller = shift;
  my $class  = ref($caller) || $caller;
  my $self   = bless( {}, $class );
  $self-&gt;{logger} = get_logger();
  <span class="keywordflow">return</span> $self;
}
</pre></div> </div> 
</div>
</div>
<hr/>The documentation for this class was generated from the following file:<ul>
<li>Bio/EnsEMBL/MetaData/<a class="el" href="AnnotationAnalyzer_8pm_source.html">AnnotationAnalyzer.pm</a></li>
</ul>
</div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="namespaceBio.html">Bio</a>      </li>
      <li class="navelem"><a class="el" href="namespaceBio_1_1EnsEMBL.html">EnsEMBL</a>      </li>
      <li class="navelem"><a class="el" href="namespaceBio_1_1EnsEMBL_1_1MetaData.html">MetaData</a>      </li>
      <li class="navelem"><a class="el" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html">AnnotationAnalyzer</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
