<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Ensembl Metadata: Bio/EnsEMBL/MetaData/DBSQL/GenomeInfoAdaptor.pm Source File</title>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('GenomeInfoAdaptor_8pm.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">GenomeInfoAdaptor.pm</div>  </div>
</div>
<div class="contents">
<a href="GenomeInfoAdaptor_8pm.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 
<a name="l00002"></a>00002 =head1 LICENSE
<a name="l00003"></a>00003 
<a name="l00004"></a>00004 Copyright [1999-2014] EMBL-European Bioinformatics Institute
<a name="l00005"></a>00005 
<a name="l00006"></a>00006 Licensed under the Apache License, Version 2.0 (the <span class="stringliteral">&quot;License&quot;</span>);
<a name="l00007"></a>00007 you may not use <span class="keyword">this</span> file except in compliance with the License.
<a name="l00008"></a>00008 You may obtain a copy of the License at
<a name="l00009"></a>00009 
<a name="l00010"></a>00010      http:<span class="comment">//www.apache.org/licenses/LICENSE-2.0</span>
<a name="l00011"></a>00011 
<a name="l00012"></a>00012 Unless required by applicable law or agreed to in writing, software
<a name="l00013"></a>00013 distributed under the License is distributed on an <span class="stringliteral">&quot;AS IS&quot;</span> BASIS,
<a name="l00014"></a>00014 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
<a name="l00015"></a>00015 See the License <span class="keywordflow">for</span> the specific language governing permissions and
<a name="l00016"></a>00016 limitations under the License.
<a name="l00017"></a>00017 
<a name="l00018"></a>00018 =cut
<a name="l00019"></a>00019 
<a name="l00020"></a>00020 =pod
<a name="l00021"></a>00021 
<a name="l00022"></a>00022 =head1 NAME
<a name="l00023"></a>00023 
<a name="l00024"></a>00024 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>
<a name="l00025"></a>00025 
<a name="l00026"></a>00026 =head1 SYNOPSIS
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <span class="preprocessor"># metadata_db is an instance of MetaDataDBAdaptor</span>
<a name="l00029"></a>00029 <span class="preprocessor"></span>my $adaptor = $metadata_db-&gt;get_GenomeInfoAdaptor();
<a name="l00030"></a>00030 my $md = $gdba-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a76c2418354c61a9d2fd07dd2c881df74">fetch_by_name</a>(<span class="stringliteral">&quot;homo_sapiens&quot;</span>);
<a name="l00031"></a>00031 
<a name="l00032"></a>00032 =head1 DESCRIPTION
<a name="l00033"></a>00033 
<a name="l00034"></a>00034 Adaptor <span class="keywordflow">for</span> storing and retrieving GenomeInfo objects from MySQL ensembl_metadata database
<a name="l00035"></a>00035 
<a name="l00036"></a>00036 To start working with an adaptor:
<a name="l00037"></a>00037 
<a name="l00038"></a>00038 <span class="preprocessor"># getting an adaptor</span>
<a name="l00039"></a>00039 <span class="preprocessor"></span><span class="preprocessor">## adaptor for latest public Ensembl release</span>
<a name="l00040"></a>00040 <span class="preprocessor"></span>my $gdba = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a208b2615ce95c153741debe91c6aca29">build_ensembl_adaptor</a>();
<a name="l00041"></a>00041 <span class="preprocessor">## adaptor for specified public Ensembl release</span>
<a name="l00042"></a>00042 <span class="preprocessor"></span>my $gdba = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a208b2615ce95c153741debe91c6aca29">build_ensembl_adaptor</a>(83);
<a name="l00043"></a>00043 <span class="preprocessor">## adaptor for latest public EG release</span>
<a name="l00044"></a>00044 <span class="preprocessor"></span>my $gdba = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa892bb1501e0bd0bbea154ed46dde3bc">build_ensembl_genomes_adaptor</a>();
<a name="l00045"></a>00045 <span class="preprocessor">## adaptor for specified public EG release</span>
<a name="l00046"></a>00046 <span class="preprocessor"></span>my $gdba = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;build_ensembl_genome_adaptor(30);
<a name="l00047"></a>00047 <span class="preprocessor">## manually specify a given database</span>
<a name="l00048"></a>00048 <span class="preprocessor"></span>my $gdba = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;new(-USER=&gt;<span class="stringliteral">&#39;anonymous&#39;</span>,
<a name="l00049"></a>00049 -PORT=&gt;4157,
<a name="l00050"></a>00050 -HOST=&gt;<span class="stringliteral">&#39;mysql-eg-publicsql.ebi.ac.uk&#39;</span>,
<a name="l00051"></a>00051 -DBNAME=&gt;<span class="stringliteral">&#39;ensembl_metadata&#39;</span>);
<a name="l00052"></a>00052 
<a name="l00053"></a>00053 To find genomes, use the fetch methods. These will work with the release <span class="keyword">set</span> on the adaptor
<a name="l00054"></a>00054 which is the latest Ensembl release by <span class="keywordflow">default</span>.
<a name="l00055"></a>00055 
<a name="l00056"></a>00056 # find a genome by name
<a name="l00057"></a>00057 my $genome = $gdba-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a76c2418354c61a9d2fd07dd2c881df74">fetch_by_name</a>(<span class="stringliteral">&#39;homo_sapiens&#39;</span>);
<a name="l00058"></a>00058 
<a name="l00059"></a>00059 <span class="preprocessor"># find and iterate over all genomes</span>
<a name="l00060"></a>00060 <span class="preprocessor"></span><span class="keywordflow">for</span> my $genome (@{$gdba-&gt;fetch_all()}) {
<a name="l00061"></a>00061     print $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#a099657bcf23824f8c7342ea680b04a1e">name</a>().<span class="stringliteral">&quot;\n&quot;</span>;
<a name="l00062"></a>00062 }
<a name="l00063"></a>00063 
<a name="l00064"></a>00064 <span class="preprocessor"># find and iterate over all genomes with variation</span>
<a name="l00065"></a>00065 <span class="preprocessor"></span><span class="keywordflow">for</span> my $genome (@{$gdba-&gt;fetch_all_with_variation()}) {
<a name="l00066"></a>00066     print $genome-&gt;name().<span class="stringliteral">&quot;\n&quot;</span>;
<a name="l00067"></a>00067 }
<a name="l00068"></a>00068 
<a name="l00069"></a>00069 <span class="preprocessor"># to change the release</span>
<a name="l00070"></a>00070 <span class="preprocessor"></span>$gdba-&gt;set_ensembl_release(82);
<a name="l00071"></a>00071 $gdba-&gt;set_ensembl_genomes_release(82);
<a name="l00072"></a>00072 
<a name="l00073"></a>00073 my $genome = $gdba-&gt;fetch_by_name(<span class="stringliteral">&#39;arabidopsis_thaliana&#39;</span>);
<a name="l00074"></a>00074 
<a name="l00075"></a>00075 <span class="preprocessor"># find and iterate over all genomes from plants</span>
<a name="l00076"></a>00076 <span class="preprocessor"></span><span class="keywordflow">for</span> my $genome (@{$gdba-&gt;fetch_all_by_division(<span class="stringliteral">&#39;EnsemblPlants&#39;</span>)}) {
<a name="l00077"></a>00077     print $genome-&gt;name().<span class="stringliteral">&quot;\n&quot;</span>;
<a name="l00078"></a>00078 }
<a name="l00079"></a>00079 
<a name="l00080"></a>00080 <span class="preprocessor"># find all comparas for the division of interest</span>
<a name="l00081"></a>00081 <span class="preprocessor"></span>my $comparas = $gdba-&gt;fetch_all_compara_by_division(<span class="stringliteral">&#39;EnsemblPlants&#39;</span>);
<a name="l00082"></a><a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">00082</a> 
<a name="l00083"></a>00083 <span class="preprocessor"># find the peptide compara</span>
<a name="l00084"></a>00084 <span class="preprocessor"></span>my ($compara) = grep {$_-&gt;is_peptide_compara()} @$comparas;
<a name="l00085"></a>00085 print $compara-&gt;division().<span class="stringliteral">&quot; &quot;</span>.$compara-&gt;method().<span class="stringliteral">&quot;(&quot;</span>.$compara-&gt;dbname().<span class="stringliteral">&quot;)\n&quot;</span>;
<a name="l00086"></a>00086 
<a name="l00087"></a>00087 <span class="preprocessor"># print out all the genomes in this compara</span>
<a name="l00088"></a>00088 <span class="preprocessor"></span><span class="keywordflow">for</span> my $genome (@{$compara-&gt;genomes()}) {
<a name="l00089"></a>00089     print $genome-&gt;name().<span class="stringliteral">&quot;\n&quot;</span>;
<a name="l00090"></a>00090 }
<a name="l00091"></a>00091 
<a name="l00092"></a>00092 =head1 Author
<a name="l00093"></a>00093 
<a name="l00094"></a>00094 Dan Staines
<a name="l00095"></a>00095 
<a name="l00096"></a>00096 =cut
<a name="l00097"></a>00097 
<a name="l00098"></a>00098 <span class="keyword">package </span>Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor;
<a name="l00099"></a>00099 
<a name="l00100"></a>00100 use strict;
<a name="l00101"></a>00101 use warnings;
<a name="l00102"></a>00102 
<a name="l00103"></a>00103 use base qw/<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1BaseInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::BaseInfoAdaptor</a>/;
<a name="l00104"></a>00104 
<a name="l00105"></a>00105 use Carp qw(cluck croak);
<a name="l00106"></a>00106 use Bio::EnsEMBL::Utils::Argument qw( rearrange );
<a name="l00107"></a>00107 use <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>;
<a name="l00108"></a>00108 use Data::Dumper;
<a name="l00109"></a>00109 use Scalar::Util qw(looks_like_number refaddr);
<a name="l00110"></a>00110 use <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1PublicMySQLServer.html">Bio::EnsEMBL::Utils::PublicMySQLServer</a> qw/e_args eg_args/;
<a name="l00111"></a>00111 use Bio::EnsEMBL::Utils::Exception qw/<span class="keywordflow">throw</span>/;
<a name="l00112"></a>00112 use <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1MetaDataDBAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::MetaDataDBAdaptor</a>;
<a name="l00113"></a>00113 
<a name="l00114"></a>00114 =head1 METHODS
<a name="l00115"></a>00115 
<a name="l00116"></a>00116 =head2 build_ensembl_adaptor
<a name="l00117"></a>00117   Arg        : (optional) Ensembl release number - <span class="keywordflow">default</span> is latest
<a name="l00118"></a>00118   Description: Build an adaptor <span class="keywordflow">for</span> the <span class="keyword">public</span> Ensembl metadata database
<a name="l00119"></a>00119   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>
<a name="l00120"></a>00120   Exceptions : none
<a name="l00121"></a>00121   Caller     : general
<a name="l00122"></a>00122   Status     : Stable
<a name="l00123"></a>00123 =cut
<a name="l00124"></a>00124 
<a name="l00125"></a>00125 sub build_ensembl_adaptor {
<a name="l00126"></a>00126   my ($release) = @_;
<a name="l00127"></a>00127   my $args = e_args();
<a name="l00128"></a>00128   $args-&gt;{-DBNAME} = <span class="stringliteral">&#39;ensembl_metadata&#39;</span>;
<a name="l00129"></a>00129   my $adaptor =
<a name="l00130"></a>00130     <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;new(
<a name="l00131"></a>00131                 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1MetaDataDBAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::MetaDataDBAdaptor</a>-&gt;<span class="keyword">new</span>(%$args) );
<a name="l00132"></a>00132   <span class="keywordflow">if</span> ( defined $release ) {
<a name="l00133"></a>00133     $adaptor-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a0be22371fc9e36b55deab2cb5a3d2cc1">set_ensembl_release</a>($release);
<a name="l00134"></a>00134   }
<a name="l00135"></a>00135   <span class="keywordflow">return</span> $adaptor;
<a name="l00136"></a>00136 }
<a name="l00137"></a>00137 
<a name="l00138"></a>00138 =head2 build_ensembl_genomes_adaptor
<a name="l00139"></a>00139   Arg        : (optional) EG release number  - <span class="keywordflow">default</span> is latest
<a name="l00140"></a>00140   Description: Build an adaptor <span class="keywordflow">for</span> the <span class="keyword">public</span> Ensembl Genomes metadata database
<a name="l00141"></a>00141   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>
<a name="l00142"></a>00142   Exceptions : none
<a name="l00143"></a>00143   Caller     : general
<a name="l00144"></a>00144   Status     : Stable
<a name="l00145"></a>00145 =cut
<a name="l00146"></a>00146 
<a name="l00147"></a>00147 sub build_ensembl_genomes_adaptor {
<a name="l00148"></a>00148   my ($release) = @_;
<a name="l00149"></a>00149   my $args = eg_args();
<a name="l00150"></a>00150   $args-&gt;{-DBNAME} = <span class="stringliteral">&#39;ensembl_metadata&#39;</span>;
<a name="l00151"></a>00151   my $adaptor =
<a name="l00152"></a>00152     <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;new(
<a name="l00153"></a>00153                 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1MetaDataDBAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::MetaDataDBAdaptor</a>-&gt;<span class="keyword">new</span>(%$args) );
<a name="l00154"></a>00154   $adaptor-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a878b804631d3b5868ea531e5566ea498">set_ensembl_genomes_release</a>($release);
<a name="l00155"></a>00155   <span class="keywordflow">return</span> $adaptor;
<a name="l00156"></a>00156 }
<a name="l00157"></a>00157 
<a name="l00158"></a>00158 =head2 set_release
<a name="l00159"></a>00159   Arg        : Ensembl release number
<a name="l00160"></a>00160   Description: Set release to use when querying 
<a name="l00161"></a>00161   Returntype : None
<a name="l00162"></a>00162   Exceptions : none
<a name="l00163"></a>00163   Caller     : general
<a name="l00164"></a>00164   Status     : Stable
<a name="l00165"></a>00165 =cut
<a name="l00166"></a>00166 
<a name="l00167"></a>00167 sub set_ensembl_release {
<a name="l00168"></a>00168   my ( $self, $release ) = @_;
<a name="l00169"></a>00169   my $release_info =
<a name="l00170"></a>00170     $self-&gt;db()-&gt;get_DataReleaseInfoAdaptor()
<a name="l00171"></a>00171     -&gt;fetch_by_ensembl_release($release);
<a name="l00172"></a>00172   <span class="keywordflow">if</span> ( !defined $release_info ) {
<a name="l00173"></a>00173     <span class="keywordflow">throw</span> <span class="stringliteral">&quot;Could not find Ensembl release $release&quot;</span>;
<a name="l00174"></a>00174   }
<a name="l00175"></a>00175   <span class="keywordflow">else</span> {
<a name="l00176"></a>00176     $self-&gt;data_release($release_info);
<a name="l00177"></a>00177   }
<a name="l00178"></a>00178   <span class="keywordflow">return</span>;
<a name="l00179"></a>00179 }
<a name="l00180"></a>00180 
<a name="l00181"></a>00181 =head2 set_ensembl_genomes_release
<a name="l00182"></a>00182   Arg        : Ensembl Genomes release number
<a name="l00183"></a>00183   Description: Set release to use when querying 
<a name="l00184"></a>00184   Returntype : None
<a name="l00185"></a>00185   Exceptions : none
<a name="l00186"></a>00186   Caller     : general
<a name="l00187"></a>00187   Status     : Stable
<a name="l00188"></a>00188 =cut
<a name="l00189"></a>00189 
<a name="l00190"></a>00190 sub set_ensembl_genomes_release {
<a name="l00191"></a>00191   my ( $self, $release ) = @_;
<a name="l00192"></a>00192   my $release_info;
<a name="l00193"></a>00193   <span class="keywordflow">if</span> ( !defined $release ) {
<a name="l00194"></a>00194     $release_info =
<a name="l00195"></a>00195       $self-&gt;db()-&gt;get_DataReleaseInfoAdaptor()
<a name="l00196"></a>00196       -&gt;fetch_current_ensembl_genomes_release($release);
<a name="l00197"></a>00197   }
<a name="l00198"></a>00198   <span class="keywordflow">else</span> {
<a name="l00199"></a>00199     $release_info =
<a name="l00200"></a>00200       $self-&gt;db()-&gt;get_DataReleaseInfoAdaptor()
<a name="l00201"></a>00201       -&gt;fetch_by_ensembl_genomes_release($release);
<a name="l00202"></a>00202   }
<a name="l00203"></a>00203   <span class="keywordflow">if</span> ( !defined $release_info ) {
<a name="l00204"></a>00204     <span class="keywordflow">throw</span> <span class="stringliteral">&quot;Could not find Ensembl Genomes release $release&quot;</span>;
<a name="l00205"></a>00205   }
<a name="l00206"></a>00206   <span class="keywordflow">else</span> {
<a name="l00207"></a>00207     $self-&gt;data_release($release_info);
<a name="l00208"></a>00208   }
<a name="l00209"></a>00209   <span class="keywordflow">return</span>;
<a name="l00210"></a>00210 }
<a name="l00211"></a>00211 
<a name="l00212"></a>00212 =head2 data_release
<a name="l00213"></a>00213   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DataReleaseInfo.html">Bio::EnsEMBL::MetaData::DataReleaseInfo</a>
<a name="l00214"></a>00214   Description: Default release to use when querying 
<a name="l00215"></a>00215   Returntype : None
<a name="l00216"></a>00216   Exceptions : none
<a name="l00217"></a>00217   Caller     : general
<a name="l00218"></a>00218   Status     : Stable
<a name="l00219"></a>00219 =cut
<a name="l00220"></a>00220 
<a name="l00221"></a>00221 sub data_release {
<a name="l00222"></a>00222   my ( $self, $release ) = @_;
<a name="l00223"></a>00223   <span class="keywordflow">if</span> ( defined $release ) {
<a name="l00224"></a>00224 <span class="preprocessor">    # replace release if we&#39;ve been given something different</span>
<a name="l00225"></a>00225 <span class="preprocessor"></span>    <span class="keywordflow">if</span> ( !defined $self-&gt;{data_release} ||
<a name="l00226"></a>00226          refaddr($release) != refaddr( $self-&gt;{data_release} ) )
<a name="l00227"></a>00227     {
<a name="l00228"></a>00228       $self-&gt;{data_release} = $release;
<a name="l00229"></a>00229       $self-&gt;db()-&gt;get_GenomeComparaInfoAdaptor()-&gt;data_release($release);
<a name="l00230"></a>00230     }
<a name="l00231"></a>00231   }
<a name="l00232"></a>00232   <span class="keywordflow">if</span> ( !defined $self-&gt;{data_release} ) {
<a name="l00233"></a>00233 <span class="preprocessor">    # default to current Ensembl release</span>
<a name="l00234"></a>00234 <span class="preprocessor"></span>    $self-&gt;{data_release} =
<a name="l00235"></a>00235       $self-&gt;db()-&gt;get_DataReleaseInfoAdaptor()
<a name="l00236"></a>00236       -&gt;fetch_current_ensembl_release();
<a name="l00237"></a>00237     <span class="keywordflow">if</span> ( defined $self-&gt;{data_release} ) {
<a name="l00238"></a>00238       $self-&gt;db()-&gt;get_GenomeComparaInfoAdaptor()
<a name="l00239"></a>00239         -&gt;data_release( $self-&gt;{data_release} );
<a name="l00240"></a>00240     }
<a name="l00241"></a>00241   }
<a name="l00242"></a>00242   <span class="keywordflow">if</span> ( !defined $self-&gt;{data_release}-&gt;dbID() ) {
<a name="l00243"></a>00243     $self-&gt;db()-&gt;get_DataReleaseInfoAdaptor()-&gt;store( $self-&gt;{data_release} );
<a name="l00244"></a>00244   }
<a name="l00245"></a>00245   <span class="keywordflow">return</span> $self-&gt;{data_release};
<a name="l00246"></a>00246 } ## end sub data_release
<a name="l00247"></a>00247 
<a name="l00248"></a>00248 =head2 store
<a name="l00249"></a>00249   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00250"></a>00250   Description: Stores the supplied <span class="keywordtype">object</span> and all associated child objects (includes other genomes attached by compara <span class="keywordflow">if</span> not already stored)
<a name="l00251"></a>00251   Returntype : None
<a name="l00252"></a>00252   Exceptions : none
<a name="l00253"></a>00253   Caller     : general
<a name="l00254"></a>00254   Status     : Stable
<a name="l00255"></a>00255 =cut
<a name="l00256"></a>00256 
<a name="l00257"></a>00257 sub store {
<a name="l00258"></a>00258   my ( $self, $genome ) = @_;
<a name="l00259"></a>00259   <span class="keywordflow">if</span> ( !defined $genome-&gt;data_release() ) {
<a name="l00260"></a>00260     $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#a090b7b65151c6ce78babe02445961be7">data_release</a>( $self-&gt;data_release() );
<a name="l00261"></a>00261   }
<a name="l00262"></a>00262   <span class="keywordflow">if</span> ( !defined $genome-&gt;data_release()-&gt;dbID() ) {
<a name="l00263"></a>00263     $self-&gt;db()-&gt;get_DataReleaseInfoAdaptor()-&gt;store( $genome-&gt;data_release() );
<a name="l00264"></a>00264   }
<a name="l00265"></a>00265   <span class="keywordflow">if</span> ( !defined $genome-&gt;assembly() ) {
<a name="l00266"></a>00266     <span class="keywordflow">throw</span>(<span class="stringliteral">&quot;Genome must be associated with an assembly&quot;</span>);
<a name="l00267"></a>00267   }
<a name="l00268"></a>00268   <span class="keywordflow">if</span> ( !defined $genome-&gt;assembly()-&gt;dbID() ) {
<a name="l00269"></a>00269     $self-&gt;db()-&gt;get_GenomeAssemblyInfoAdaptor()-&gt;store( $genome-&gt;assembly() );
<a name="l00270"></a>00270   }
<a name="l00271"></a>00271   <span class="keywordflow">if</span> ( !defined $genome-&gt;dbID() ) {
<a name="l00272"></a>00272 <span class="preprocessor">    # find out if genome exists first</span>
<a name="l00273"></a>00273 <span class="preprocessor"></span>    my ($dbID) =
<a name="l00274"></a>00274       @{
<a name="l00275"></a>00275       $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_simple(
<a name="l00276"></a>00276         -SQL =&gt;
<a name="l00277"></a>00277 <span class="stringliteral">&quot;select genome_id from genome where data_release_id=? and assembly_id=?&quot;</span>,
<a name="l00278"></a>00278         -PARAMS =&gt;
<a name="l00279"></a>00279           [ $genome-&gt;data_release()-&gt;dbID(), $genome-&gt;assembly()-&gt;dbID() ] ) };
<a name="l00280"></a>00280 
<a name="l00281"></a>00281     <span class="keywordflow">if</span> ( defined $dbID ) {
<a name="l00282"></a>00282       $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1BaseInfo.html#addea948253febfab2a41e517b3dec92e">dbID</a>($dbID);
<a name="l00283"></a>00283       $genome-&gt;adaptor($self);
<a name="l00284"></a>00284     }
<a name="l00285"></a>00285   }
<a name="l00286"></a>00286 
<a name="l00287"></a>00287   <span class="keywordflow">if</span> ( defined $genome-&gt;dbID() ) {
<a name="l00288"></a>00288     <span class="keywordflow">return</span> $self-&gt;update($genome);
<a name="l00289"></a>00289   }
<a name="l00290"></a>00290   $self-&gt;db()-&gt;get_GenomeAssemblyInfoAdaptor()-&gt;store( $genome-&gt;assembly() );
<a name="l00291"></a>00291   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l00292"></a>00292     -SQL =&gt; q/insert into genome(division_id,
<a name="l00293"></a>00293 genebuild,has_pan_compara,has_variations,has_peptide_compara,
<a name="l00294"></a>00294 has_genome_alignments,has_synteny,has_other_alignments,assembly_id,organism_id,data_release_id)
<a name="l00295"></a>00295         values(?,?,?,?,?,?,?,?,?,?,?)/,
<a name="l00296"></a>00296     -PARAMS =&gt; [ $self-&gt;_get_division_id( $genome-&gt;division() ),
<a name="l00297"></a>00297                  $genome-&gt;genebuild(),
<a name="l00298"></a>00298                  $genome-&gt;has_pan_compara(),
<a name="l00299"></a>00299                  $genome-&gt;has_variations(),
<a name="l00300"></a>00300                  $genome-&gt;has_peptide_compara(),
<a name="l00301"></a>00301                  $genome-&gt;has_genome_alignments(),
<a name="l00302"></a>00302                  $genome-&gt;has_synteny(),
<a name="l00303"></a>00303                  $genome-&gt;has_other_alignments(),
<a name="l00304"></a>00304                  $genome-&gt;assembly()-&gt;dbID(),
<a name="l00305"></a>00305                  $genome-&gt;assembly()-&gt;organism()-&gt;dbID(),
<a name="l00306"></a>00306                  $genome-&gt;data_release()-&gt;dbID() ],
<a name="l00307"></a>00307     -CALLBACK =&gt; sub {
<a name="l00308"></a>00308       my ( $sth, $dbh, $rv ) = @_;
<a name="l00309"></a>00309       $genome-&gt;dbID( $dbh-&gt;{mysql_insertid} );
<a name="l00310"></a>00310     } );
<a name="l00311"></a>00311   $genome-&gt;adaptor($self);
<a name="l00312"></a>00312   $self-&gt;_store_databases($genome);
<a name="l00313"></a>00313   $self-&gt;_store_annotations($genome);
<a name="l00314"></a>00314   $self-&gt;_store_features($genome);
<a name="l00315"></a>00315   $self-&gt;_store_variations($genome);
<a name="l00316"></a>00316   $self-&gt;_store_alignments($genome);
<a name="l00317"></a>00317   $self-&gt;_store_compara($genome);
<a name="l00318"></a>00318   $self-&gt;_store_cached_obj($genome);
<a name="l00319"></a>00319   <span class="keywordflow">return</span>;
<a name="l00320"></a>00320 } ## end sub store
<a name="l00321"></a>00321 
<a name="l00322"></a>00322 =head2 update
<a name="l00323"></a>00323   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00324"></a>00324   Description: Updates the supplied <span class="keywordtype">object</span> and all associated child objects (includes other genomes attached by compara <span class="keywordflow">if</span> not already stored)
<a name="l00325"></a>00325   Returntype : None
<a name="l00326"></a>00326   Exceptions : none
<a name="l00327"></a>00327   Caller     : general
<a name="l00328"></a>00328   Status     : Stable
<a name="l00329"></a>00329 =cut
<a name="l00330"></a>00330 
<a name="l00331"></a>00331 sub update {
<a name="l00332"></a>00332   my ( $self, $genome ) = @_;
<a name="l00333"></a>00333   <span class="keywordflow">if</span> ( !defined $genome-&gt;dbID() ) {
<a name="l00334"></a>00334     croak <span class="stringliteral">&quot;Cannot update an object that has not already been stored&quot;</span>;
<a name="l00335"></a>00335   }
<a name="l00336"></a>00336   $self-&gt;db()-&gt;get_GenomeAssemblyInfoAdaptor()-&gt;update( $genome-&gt;assembly() );
<a name="l00337"></a>00337   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l00338"></a>00338     -SQL =&gt; q/update genome <span class="keyword">set</span> division_id=?,
<a name="l00339"></a>00339 genebuild=?,has_pan_compara=?,has_variations=?,has_peptide_compara=?,
<a name="l00340"></a>00340 has_genome_alignments=?,has_synteny=?,has_other_alignments=?,assembly_id=?,organism_id=?,data_release_id=? where genome_id=?/
<a name="l00341"></a>00341     ,
<a name="l00342"></a>00342     -PARAMS =&gt; [ $self-&gt;_get_division_id( $genome-&gt;division() ),
<a name="l00343"></a>00343                  $genome-&gt;genebuild(),
<a name="l00344"></a>00344                  $genome-&gt;has_pan_compara(),
<a name="l00345"></a>00345                  $genome-&gt;has_variations(),
<a name="l00346"></a>00346                  $genome-&gt;has_peptide_compara(),
<a name="l00347"></a>00347                  $genome-&gt;has_genome_alignments(),
<a name="l00348"></a>00348                  $genome-&gt;has_synteny(),
<a name="l00349"></a>00349                  $genome-&gt;has_other_alignments(),
<a name="l00350"></a>00350                  $genome-&gt;assembly()-&gt;dbID(),
<a name="l00351"></a>00351                  $genome-&gt;assembly()-&gt;organism()-&gt;dbID(),
<a name="l00352"></a>00352                  $genome-&gt;data_release()-&gt;dbID(),
<a name="l00353"></a>00353                  $genome-&gt;dbID() ] );
<a name="l00354"></a>00354   $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#a29d96e7a145a18364d0de3416d988f11">adaptor</a>($self);
<a name="l00355"></a>00355   $self-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#a53d1a1c51ea4c44e449197abbe6f7ffc">_store_databases</a>($genome);
<a name="l00356"></a>00356   $self-&gt;_store_annotations($genome);
<a name="l00357"></a>00357   $self-&gt;_store_features($genome);
<a name="l00358"></a>00358   $self-&gt;_store_variations($genome);
<a name="l00359"></a>00359   $self-&gt;_store_alignments($genome);
<a name="l00360"></a>00360   $self-&gt;_store_compara($genome);
<a name="l00361"></a>00361   <span class="keywordflow">return</span>;
<a name="l00362"></a>00362 } ## end sub update
<a name="l00363"></a>00363 
<a name="l00364"></a>00364 =head2 update_booleans
<a name="l00365"></a>00365   Description: Updates <span class="keywordtype">boolean</span> genome attributes <span class="keywordflow">for</span> all genomes
<a name="l00366"></a>00366   Returntype : None
<a name="l00367"></a>00367   Exceptions : none
<a name="l00368"></a>00368   Caller     : <span class="keyword">internal</span>
<a name="l00369"></a>00369   Status     : Stable
<a name="l00370"></a>00370 =cut
<a name="l00371"></a>00371 
<a name="l00372"></a>00372 sub update_booleans {
<a name="l00373"></a>00373   my ($self) = @_;
<a name="l00374"></a>00374 
<a name="l00375"></a>00375 <span class="preprocessor">  #has_peptide_compara</span>
<a name="l00376"></a>00376 <span class="preprocessor"></span>  $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l00377"></a>00377     -SQL =&gt; q/update genome g join genome_compara_analysis gc <span class="keyword">using</span> (genome_id) 
<a name="l00378"></a>00378       join compara_analysis c <span class="keyword">using</span> (compara_analysis_id) 
<a name="l00379"></a>00379       <span class="keyword">set</span> g.has_peptide_compara=1 where c.division&lt;&gt;<span class="stringliteral">&#39;EnsemblPan&#39;</span> and
<a name="l00380"></a>00380       c.method=<span class="stringliteral">&#39;PROTEIN_TREES&#39;</span>/
<a name="l00381"></a>00381   );
<a name="l00382"></a>00382 
<a name="l00383"></a>00383 <span class="preprocessor">  #has_pan_compara</span>
<a name="l00384"></a>00384 <span class="preprocessor"></span>  $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l00385"></a>00385     -SQL =&gt; q/update genome g join genome_compara_analysis gc <span class="keyword">using</span> (genome_id) 
<a name="l00386"></a>00386       join compara_analysis c <span class="keyword">using</span> (compara_analysis_id) 
<a name="l00387"></a>00387       <span class="keyword">set</span> g.has_pan_compara=1 where c.division=<span class="stringliteral">&#39;EnsemblPan&#39;</span> and
<a name="l00388"></a>00388       c.method=<span class="stringliteral">&#39;PROTEIN_TREES&#39;</span>/
<a name="l00389"></a>00389   );
<a name="l00390"></a>00390 
<a name="l00391"></a>00391 <span class="preprocessor">  #has_genome_alignments</span>
<a name="l00392"></a>00392 <span class="preprocessor"></span>  $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l00393"></a>00393     -SQL =&gt; q/update genome g join genome_compara_analysis gc <span class="keyword">using</span> (genome_id) 
<a name="l00394"></a>00394       join compara_analysis c <span class="keyword">using</span> (compara_analysis_id) 
<a name="l00395"></a>00395       <span class="keyword">set</span> g.has_genome_alignments=1 where c.method in 
<a name="l00396"></a>00396       (<span class="stringliteral">&#39;TRANSLATED_BLAT_NET&#39;</span>,<span class="stringliteral">&#39;LASTZ_NET&#39;</span>,<span class="stringliteral">&#39;TBLAT&#39;</span>,<span class="stringliteral">&#39;ATAC&#39;</span>,<span class="stringliteral">&#39;BLASTZ_NET&#39;</span>)/
<a name="l00397"></a>00397   );
<a name="l00398"></a>00398 
<a name="l00399"></a>00399 <span class="preprocessor">  #has_synteny</span>
<a name="l00400"></a>00400 <span class="preprocessor"></span>  $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l00401"></a>00401     -SQL =&gt; q/update genome g join genome_compara_analysis gc <span class="keyword">using</span> (genome_id) 
<a name="l00402"></a>00402       join compara_analysis c <span class="keyword">using</span> (compara_analysis_id) 
<a name="l00403"></a>00403       <span class="keyword">set</span> g.has_synteny=1 where c.method=<span class="stringliteral">&#39;SYNTENY&#39;</span>/
<a name="l00404"></a>00404   );
<a name="l00405"></a>00405 
<a name="l00406"></a>00406 <span class="preprocessor">  #has_other_alignments</span>
<a name="l00407"></a>00407 <span class="preprocessor"></span>  $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l00408"></a>00408     -SQL =&gt; q/update genome g join genome_alignment a <span class="keyword">using</span> (genome_id) 
<a name="l00409"></a>00409       <span class="keyword">set</span> g.has_other_alignments=1/
<a name="l00410"></a>00410   );
<a name="l00411"></a>00411 
<a name="l00412"></a>00412 <span class="preprocessor">  #has_variations</span>
<a name="l00413"></a>00413 <span class="preprocessor"></span>  $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l00414"></a>00414     -SQL =&gt; q/update genome g join genome_variation a <span class="keyword">using</span> (genome_id)
<a name="l00415"></a>00415       <span class="keyword">set</span> g.has_variations=1/
<a name="l00416"></a>00416   );
<a name="l00417"></a>00417 
<a name="l00418"></a>00418   <span class="keywordflow">return</span>;
<a name="l00419"></a>00419 } ## end sub update_booleans
<a name="l00420"></a>00420 
<a name="l00421"></a>00421 =head2 fetch_databases 
<a name="l00422"></a>00422   Arg        : release (optional)
<a name="l00423"></a>00423   Description: Fetch all genome-associated databases for the specified release
<a name="l00424"></a>00424   Returntype : Arrayref of strings
<a name="l00425"></a>00425   Exceptions : none
<a name="l00426"></a>00426   Caller     : general
<a name="l00427"></a>00427   Status     : Stable
<a name="l00428"></a>00428 =cut
<a name="l00429"></a>00429 
<a name="l00430"></a>00430 sub fetch_databases {
<a name="l00431"></a>00431   my ( $self, $release ) = @_;
<a name="l00432"></a>00432   <span class="keywordflow">if</span> ( !defined $release ) {
<a name="l00433"></a>00433     $release = $self-&gt;data_release();
<a name="l00434"></a>00434   }
<a name="l00435"></a>00435   <span class="keywordflow">return</span> $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_simple(
<a name="l00436"></a>00436     -SQL =&gt; q/select distinct dbname from genome_database
<a name="l00437"></a>00437       join genome <span class="keyword">using</span> (genome_id) where data_release_id=?/,
<a name="l00438"></a>00438     -PARAMS =&gt; [ $release-&gt;dbID() ] );
<a name="l00439"></a>00439 }
<a name="l00440"></a>00440 
<a name="l00441"></a>00441 =head2 fetch_division_databases 
<a name="l00442"></a>00442   Arg        : division
<a name="l00443"></a>00443   Arg        : release (optional)
<a name="l00444"></a>00444   Description: Fetch all genome-associated databases for the specified release
<a name="l00445"></a>00445   Returntype : Arrayref of strings
<a name="l00446"></a>00446   Exceptions : none
<a name="l00447"></a>00447   Caller     : general
<a name="l00448"></a>00448   Status     : Stable
<a name="l00449"></a>00449 =cut
<a name="l00450"></a>00450 
<a name="l00451"></a>00451 sub fetch_division_databases {
<a name="l00452"></a>00452   my ( $self, $division, $release ) = @_;
<a name="l00453"></a>00453   <span class="keywordflow">if</span> ( !defined $release ) {
<a name="l00454"></a>00454     $release = $self-&gt;data_release();
<a name="l00455"></a>00455   }
<a name="l00456"></a>00456   <span class="keywordflow">return</span> $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_simple(
<a name="l00457"></a>00457     -SQL =&gt; q/select distinct gd.dbname from genome_database gd
<a name="l00458"></a>00458       join genome g using (genome_id)
<a name="l00459"></a>00459       join division d <span class="keyword">using</span> (division_id)
<a name="l00460"></a>00460       where data_release_id=? and (d.name=? OR d.short_name=?)/,
<a name="l00461"></a>00461     -PARAMS =&gt; [ $release-&gt;dbID(), $division, $division ] );
<a name="l00462"></a>00462 }
<a name="l00463"></a>00463 
<a name="l00464"></a>00464 =head2 fetch_by_organism 
<a name="l00465"></a>00465   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeOrganismInfo.html">Bio::EnsEMBL::MetaData::GenomeOrganismInfo</a>
<a name="l00466"></a>00466   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00467"></a>00467   Description: Fetch genome info <span class="keywordflow">for</span> specified organism
<a name="l00468"></a>00468   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00469"></a>00469   Exceptions : none
<a name="l00470"></a>00470   Caller     : general
<a name="l00471"></a>00471   Status     : Stable
<a name="l00472"></a>00472 =cut
<a name="l00473"></a>00473 
<a name="l00474"></a>00474 sub fetch_by_organism {
<a name="l00475"></a>00475   my ( $self, $organism, $keen ) = @_;
<a name="l00476"></a>00476   <span class="keywordflow">if</span> ( ref($organism) eq <span class="stringliteral">&#39;Bio::EnsEMBL::MetaData::GenomeOrganismInfo&#39;</span> ) {
<a name="l00477"></a>00477     $organism = $organism-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#ae2ae7289199643ff4352d80941bdb6cd">dbID</a>();
<a name="l00478"></a>00478   }
<a name="l00479"></a>00479   <span class="keywordflow">return</span> $self-&gt;_first_element(
<a name="l00480"></a>00480        $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;organism_id&#39;</span>, $organism }, $keen ) );
<a name="l00481"></a>00481 }
<a name="l00482"></a>00482 
<a name="l00483"></a>00483 =head2 fetch_all_by_organisms
<a name="l00484"></a>00484   Arg        : Array ref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeOrganismInfo.html">Bio::EnsEMBL::MetaData::GenomeOrganismInfo</a>
<a name="l00485"></a>00485   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00486"></a>00486   Description: Fetch genome info <span class="keywordflow">for</span> specified organisms
<a name="l00487"></a>00487   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00488"></a>00488   Exceptions : none
<a name="l00489"></a>00489   Caller     : general
<a name="l00490"></a>00490   Status     : Stable
<a name="l00491"></a>00491 =cut
<a name="l00492"></a>00492 
<a name="l00493"></a>00493 sub fetch_all_by_organisms {
<a name="l00494"></a>00494   my ( $self, $organisms, $keen ) = @_;
<a name="l00495"></a>00495   <span class="keywordflow">if</span> ( !defined $organisms || scalar($organisms) == 0 ) {
<a name="l00496"></a>00496     <span class="keywordflow">return</span> [];
<a name="l00497"></a>00497   }
<a name="l00498"></a>00498   <span class="keywordflow">return</span>
<a name="l00499"></a>00499     $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;organism_id&#39;</span>,
<a name="l00500"></a>00500                                        [ map { $_-&gt;dbID() } @$organisms ] },
<a name="l00501"></a>00501                                      $keen );
<a name="l00502"></a>00502 }
<a name="l00503"></a>00503 
<a name="l00504"></a>00504 =head2 fetch_by_assembly 
<a name="l00505"></a>00505   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeAssemblyInfo.html">Bio::EnsEMBL::MetaData::GenomeAssemblyInfo</a>
<a name="l00506"></a>00506   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00507"></a>00507   Description: Fetch genome info <span class="keywordflow">for</span> specified assembly
<a name="l00508"></a>00508   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00509"></a>00509   Exceptions : none
<a name="l00510"></a>00510   Caller     : general
<a name="l00511"></a>00511   Status     : Stable
<a name="l00512"></a>00512 =cut
<a name="l00513"></a>00513 
<a name="l00514"></a>00514 sub fetch_by_assembly {
<a name="l00515"></a>00515   my ( $self, $assembly, $keen ) = @_;
<a name="l00516"></a>00516   <span class="keywordflow">if</span> ( ref($assembly) eq <span class="stringliteral">&#39;Bio::EnsEMBL::MetaData::GenomeAssemblyInfo&#39;</span> ) {
<a name="l00517"></a>00517     $assembly = $assembly-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#ae2ae7289199643ff4352d80941bdb6cd">dbID</a>();
<a name="l00518"></a>00518   }
<a name="l00519"></a>00519   <span class="keywordflow">return</span> $self-&gt;_first_element(
<a name="l00520"></a>00520        $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;assembly_id&#39;</span>, $assembly }, $keen ) );
<a name="l00521"></a>00521 }
<a name="l00522"></a>00522 
<a name="l00523"></a>00523 =head2 fetch_all_by_assemblies 
<a name="l00524"></a>00524   Arg        : array of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeAssemblyInfo.html">Bio::EnsEMBL::MetaData::GenomeAssemblyInfo</a>
<a name="l00525"></a>00525   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00526"></a>00526   Description: Fetch genome info <span class="keywordflow">for</span> specified assembly
<a name="l00527"></a>00527   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00528"></a>00528   Exceptions : none
<a name="l00529"></a>00529   Caller     : general
<a name="l00530"></a>00530   Status     : Stable
<a name="l00531"></a>00531 =cut
<a name="l00532"></a>00532 
<a name="l00533"></a>00533 sub fetch_all_by_assemblies {
<a name="l00534"></a>00534   my ( $self, $assemblies, $keen ) = @_;
<a name="l00535"></a>00535   <span class="keywordflow">if</span> ( !defined $assemblies || scalar(@$assemblies) == 0 ) {
<a name="l00536"></a>00536     <span class="keywordflow">return</span> [];
<a name="l00537"></a>00537   }
<a name="l00538"></a>00538   <span class="keywordflow">return</span>
<a name="l00539"></a>00539     $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;assembly_id&#39;</span>,
<a name="l00540"></a>00540                                        [ map { $_-&gt;dbID() } @$assemblies ] },
<a name="l00541"></a>00541                                      $keen );
<a name="l00542"></a>00542 }
<a name="l00543"></a>00543 
<a name="l00544"></a>00544 =head2 fetch_all_by_sequence_accession
<a name="l00545"></a>00545   Arg        : INSDC sequence accession e.g. U00096.1 or U00096
<a name="l00546"></a>00546   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00547"></a>00547   Description: Fetch genome info <span class="keywordflow">for</span> specified sequence accession
<a name="l00548"></a>00548   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00549"></a>00549   Exceptions : none
<a name="l00550"></a>00550   Caller     : general
<a name="l00551"></a>00551   Status     : Stable
<a name="l00552"></a>00552 =cut
<a name="l00553"></a>00553 
<a name="l00554"></a>00554 sub fetch_all_by_sequence_accession {
<a name="l00555"></a>00555   my ( $self, $id, $keen ) = @_;
<a name="l00556"></a>00556   my $ass =
<a name="l00557"></a>00557     $self-&gt;_assembly_adaptor()-&gt;fetch_all_by_sequence_accession_versioned($id);
<a name="l00558"></a>00558   my $infos = $self-&gt;fetch_all_by_assemblies( $ass, $keen );
<a name="l00559"></a>00559   <span class="keywordflow">if</span> ( !defined $infos || scalar(@$infos) == 0 ) {
<a name="l00560"></a>00560     $ass =
<a name="l00561"></a>00561       $self-&gt;_assembly_adaptor()
<a name="l00562"></a>00562       -&gt;fetch_all_by_sequence_accession_unversioned($id);
<a name="l00563"></a>00563     $infos = $self-&gt;fetch_all_by_assemblies( $ass, $keen );
<a name="l00564"></a>00564   }
<a name="l00565"></a>00565   <span class="keywordflow">return</span> $infos;
<a name="l00566"></a>00566 }
<a name="l00567"></a>00567 
<a name="l00568"></a>00568 =head2 fetch_all_by_sequence_accession_unversioned
<a name="l00569"></a>00569   Arg        : INSDC sequence accession e.g. U00096
<a name="l00570"></a>00570   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00571"></a>00571   Description: Fetch genome info <span class="keywordflow">for</span> specified sequence accession
<a name="l00572"></a>00572   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00573"></a>00573   Exceptions : none
<a name="l00574"></a>00574   Caller     : general
<a name="l00575"></a>00575   Status     : Stable
<a name="l00576"></a>00576 =cut
<a name="l00577"></a>00577 
<a name="l00578"></a>00578 sub fetch_all_by_sequence_accession_unversioned {
<a name="l00579"></a>00579   my ( $self, $id, $keen ) = @_;
<a name="l00580"></a>00580   my $ass =
<a name="l00581"></a>00581     $self-&gt;_assembly_adaptor()-&gt;fetch_all_sequence_accession_unversioned($id);
<a name="l00582"></a>00582   <span class="keywordflow">return</span> $self-&gt;fetch_all_by_assemblies( $ass, $keen );
<a name="l00583"></a>00583 }
<a name="l00584"></a>00584 
<a name="l00585"></a>00585 =head2 fetch_all_by_sequence_accession_versioned
<a name="l00586"></a>00586   Arg        : INSDC sequence accession e.g. U00096.1
<a name="l00587"></a>00587   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00588"></a>00588   Description: Fetch genome info <span class="keywordflow">for</span> specified sequence accession
<a name="l00589"></a>00589   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00590"></a>00590   Exceptions : none
<a name="l00591"></a>00591   Caller     : general
<a name="l00592"></a>00592   Status     : Stable
<a name="l00593"></a>00593 =cut
<a name="l00594"></a>00594 
<a name="l00595"></a>00595 sub fetch_all_by_sequence_accession_versioned {
<a name="l00596"></a>00596   my ( $self, $id, $keen ) = @_;
<a name="l00597"></a>00597   my $ass =
<a name="l00598"></a>00598     $self-&gt;_assembly_adaptor()-&gt;fetch_all_sequence_accession_versioned($id);
<a name="l00599"></a>00599   <span class="keywordflow">return</span> $self-&gt;fetch_all_by_assemblies( $ass, $keen );
<a name="l00600"></a>00600 }
<a name="l00601"></a>00601 
<a name="l00602"></a>00602 =head2 fetch_by_assembly_accession
<a name="l00603"></a>00603   Arg        : INSDC assembly accession
<a name="l00604"></a>00604   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00605"></a>00605   Description: Fetch genome info <span class="keywordflow">for</span> specified assembly ID (versioned or unversioned)
<a name="l00606"></a>00606   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00607"></a>00607   Exceptions : none
<a name="l00608"></a>00608   Caller     : general
<a name="l00609"></a>00609   Status     : Stable
<a name="l00610"></a>00610 =cut
<a name="l00611"></a>00611 
<a name="l00612"></a>00612 sub fetch_by_assembly_accession {
<a name="l00613"></a>00613   my ( $self, $id, $keen ) = @_;
<a name="l00614"></a>00614   my $ass = $self-&gt;_assembly_adaptor()-&gt;fetch_by_assembly_accession($id);
<a name="l00615"></a>00615   <span class="keywordflow">return</span> $self-&gt;fetch_by_assembly( $ass, $keen );
<a name="l00616"></a>00616 }
<a name="l00617"></a>00617 
<a name="l00618"></a>00618 =head2 fetch_all_by_assembly_set_chain
<a name="l00619"></a>00619   Arg        : INSDC assembly <span class="keyword">set</span> chain (unversioned accession)
<a name="l00620"></a>00620   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00621"></a>00621   Description: Fetch genome info <span class="keywordflow">for</span> specified assembly <span class="keyword">set</span> chain
<a name="l00622"></a>00622   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00623"></a>00623   Exceptions : none
<a name="l00624"></a>00624   Caller     : general
<a name="l00625"></a>00625   Status     : Stable
<a name="l00626"></a>00626 =cut
<a name="l00627"></a>00627 
<a name="l00628"></a>00628 sub fetch_all_by_assembly_set_chain {
<a name="l00629"></a>00629   my ( $self, $id, $keen ) = @_;
<a name="l00630"></a>00630   my $ass = $self-&gt;_assembly_adaptor()-&gt;fetch_all_by_assembly_set_chain($id);
<a name="l00631"></a>00631   <span class="keywordflow">return</span> $self-&gt;fetch_all_by_assemblies( $ass, $keen );
<a name="l00632"></a>00632 }
<a name="l00633"></a>00633 
<a name="l00634"></a>00634 =head2 fetch_all_by_division
<a name="l00635"></a>00635   Arg        : Name of division
<a name="l00636"></a>00636   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00637"></a>00637   Description: Fetch genome infos <span class="keywordflow">for</span> specified division
<a name="l00638"></a>00638   Returntype : Arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00639"></a>00639   Exceptions : none
<a name="l00640"></a>00640   Caller     : general
<a name="l00641"></a>00641   Status     : Stable
<a name="l00642"></a>00642 =cut
<a name="l00643"></a>00643 
<a name="l00644"></a>00644 sub fetch_all_by_division {
<a name="l00645"></a>00645   my ( $self, $division, $keen ) = @_;
<a name="l00646"></a>00646   <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;division&#39;</span>, $division }, $keen );
<a name="l00647"></a>00647 }
<a name="l00648"></a>00648 
<a name="l00649"></a>00649 =head2 fetch_by_display_name
<a name="l00650"></a>00650   Arg        : Production name of genome 
<a name="l00651"></a>00651   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00652"></a>00652   Description: Fetch genome info <span class="keywordflow">for</span> specified species
<a name="l00653"></a>00653   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00654"></a>00654   Exceptions : none
<a name="l00655"></a>00655   Caller     : general
<a name="l00656"></a>00656   Status     : Stable
<a name="l00657"></a>00657 =cut
<a name="l00658"></a>00658 
<a name="l00659"></a>00659 sub fetch_by_display_name {
<a name="l00660"></a>00660   my ( $self, $name, $keen ) = @_;
<a name="l00661"></a>00661   my $org = $self-&gt;_organism_adaptor()-&gt;fetch_by_display_name($name);
<a name="l00662"></a>00662   <span class="keywordflow">return</span> $self-&gt;fetch_by_organism( $org, $keen );
<a name="l00663"></a>00663 }
<a name="l00664"></a>00664 
<a name="l00665"></a>00665 =head2 fetch_by_name
<a name="l00666"></a>00666   Arg        : Display name of genome 
<a name="l00667"></a>00667   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00668"></a>00668   Description: Fetch genome info <span class="keywordflow">for</span> specified species
<a name="l00669"></a>00669   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00670"></a>00670   Exceptions : none
<a name="l00671"></a>00671   Caller     : general
<a name="l00672"></a>00672   Status     : Stable
<a name="l00673"></a>00673 =cut
<a name="l00674"></a>00674 
<a name="l00675"></a>00675 sub fetch_by_name {
<a name="l00676"></a>00676   my ( $self, $name, $keen ) = @_;
<a name="l00677"></a>00677   my $org = $self-&gt;_organism_adaptor()-&gt;fetch_by_name($name);
<a name="l00678"></a>00678   <span class="keywordflow">return</span> $self-&gt;fetch_by_organism( $org, $keen );
<a name="l00679"></a>00679 }
<a name="l00680"></a>00680 
<a name="l00681"></a>00681 =head2 fetch_by_any_name
<a name="l00682"></a>00682   Arg        : Name of genome (display, species, alias etc)
<a name="l00683"></a>00683   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00684"></a>00684   Description: Fetch genome info <span class="keywordflow">for</span> specified species
<a name="l00685"></a>00685   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00686"></a>00686   Exceptions : none
<a name="l00687"></a>00687   Caller     : general
<a name="l00688"></a>00688   Status     : Stable
<a name="l00689"></a>00689 =cut
<a name="l00690"></a>00690 
<a name="l00691"></a>00691 sub fetch_by_any_name {
<a name="l00692"></a>00692   my ( $self, $name, $keen ) = @_;
<a name="l00693"></a>00693   my $org = $self-&gt;_organism_adaptor()-&gt;fetch_by_name($name);
<a name="l00694"></a>00694   <span class="keywordflow">if</span> ( !defined $org ) {
<a name="l00695"></a>00695     $org = $self-&gt;_organism_adaptor()-&gt;fetch_by_alias($name);
<a name="l00696"></a>00696   }
<a name="l00697"></a>00697   <span class="keywordflow">return</span> $self-&gt;fetch_by_organism( $org, $keen );
<a name="l00698"></a>00698 }
<a name="l00699"></a>00699 
<a name="l00700"></a>00700 =head2 fetch_all_by_dbname
<a name="l00701"></a>00701   Arg        : Name of database
<a name="l00702"></a>00702   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00703"></a>00703   Description: Fetch genome info <span class="keywordflow">for</span> specified database
<a name="l00704"></a>00704   Returntype : arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00705"></a>00705   Exceptions : none
<a name="l00706"></a>00706   Caller     : general
<a name="l00707"></a>00707   Status     : Stable
<a name="l00708"></a>00708 =cut
<a name="l00709"></a>00709 
<a name="l00710"></a>00710 sub fetch_all_by_dbname {
<a name="l00711"></a>00711   my ( $self, $name, $keen ) = @_;
<a name="l00712"></a>00712   <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;dbname&#39;</span>, $name }, $keen );
<a name="l00713"></a>00713 }
<a name="l00714"></a>00714 
<a name="l00715"></a>00715 =head2 fetch_all_by_name_pattern
<a name="l00716"></a>00716   Arg        : Regular expression matching of genome
<a name="l00717"></a>00717   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00718"></a>00718   Description: Fetch genome info <span class="keywordflow">for</span> specified species
<a name="l00719"></a>00719   Returntype : Arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00720"></a>00720   Exceptions : none
<a name="l00721"></a>00721   Caller     : general
<a name="l00722"></a>00722   Status     : Stable
<a name="l00723"></a>00723 =cut
<a name="l00724"></a>00724 
<a name="l00725"></a>00725 sub fetch_all_by_name_pattern {
<a name="l00726"></a>00726   my ( $self, $name, $keen ) = @_;
<a name="l00727"></a>00727   my $orgs = $self-&gt;_organism_adaptor()-&gt;fetch_all_by_name_pattern($name);
<a name="l00728"></a>00728   <span class="keywordflow">return</span> $self-&gt;fetch_all_by_organisms( $orgs, $keen );
<a name="l00729"></a>00729 }
<a name="l00730"></a>00730 
<a name="l00731"></a>00731 =head2 fetch_by_alias
<a name="l00732"></a>00732   Arg        : Alias of genome
<a name="l00733"></a>00733   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00734"></a>00734   Description: Fetch genome info <span class="keywordflow">for</span> specified species
<a name="l00735"></a>00735   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00736"></a>00736   Exceptions : none
<a name="l00737"></a>00737   Caller     : general
<a name="l00738"></a>00738   Status     : Stable
<a name="l00739"></a>00739 =cut
<a name="l00740"></a>00740 
<a name="l00741"></a>00741 sub fetch_by_alias {
<a name="l00742"></a>00742   my ( $self, $name, $keen ) = @_;
<a name="l00743"></a>00743   my $org = $self-&gt;_organism_adaptor()-&gt;fetch_all_by_name_pattern($name);
<a name="l00744"></a>00744   <span class="keywordflow">return</span> $self-&gt;fetch_by_organism( $org, $keen );
<a name="l00745"></a>00745 }
<a name="l00746"></a>00746 
<a name="l00747"></a>00747 =head2 fetch_all_with_variation
<a name="l00748"></a>00748   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00749"></a>00749   Description: Fetch all genome info that have variation data
<a name="l00750"></a>00750   Returntype : arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00751"></a>00751   Exceptions : none
<a name="l00752"></a>00752   Caller     : general
<a name="l00753"></a>00753   Status     : Stable
<a name="l00754"></a>00754 =cut
<a name="l00755"></a>00755 
<a name="l00756"></a>00756 sub fetch_all_with_variation {
<a name="l00757"></a>00757   my ( $self, $keen ) = @_;
<a name="l00758"></a>00758   <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_variations&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> }, $keen );
<a name="l00759"></a>00759 }
<a name="l00760"></a>00760 
<a name="l00761"></a>00761 =head2 fetch_all_with_peptide_compara
<a name="l00762"></a>00762   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00763"></a>00763   Description: Fetch all genome info that have peptide compara data
<a name="l00764"></a>00764   Returntype : arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00765"></a>00765   Exceptions : none
<a name="l00766"></a>00766   Caller     : general
<a name="l00767"></a>00767   Status     : Stable
<a name="l00768"></a>00768 =cut
<a name="l00769"></a>00769 
<a name="l00770"></a>00770 sub fetch_all_with_peptide_compara {
<a name="l00771"></a>00771   my ( $self, $keen ) = @_;
<a name="l00772"></a>00772   <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_peptide_compara&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> },
<a name="l00773"></a>00773                                           $keen );
<a name="l00774"></a>00774 }
<a name="l00775"></a>00775 
<a name="l00776"></a>00776 =head2 fetch_all_with_pan_compara
<a name="l00777"></a>00777   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00778"></a>00778   Description: Fetch all genome info that have pan comapra data
<a name="l00779"></a>00779   Returntype : arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00780"></a>00780   Exceptions : none
<a name="l00781"></a>00781   Caller     : general
<a name="l00782"></a>00782   Status     : Stable
<a name="l00783"></a>00783 =cut
<a name="l00784"></a>00784 
<a name="l00785"></a>00785 sub fetch_all_with_pan_compara {
<a name="l00786"></a>00786   my ( $self, $keen ) = @_;
<a name="l00787"></a>00787   <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_pan_compara&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> }, $keen );
<a name="l00788"></a>00788 }
<a name="l00789"></a>00789 
<a name="l00790"></a>00790 =head2 fetch_all_with_genome_alignments
<a name="l00791"></a>00791   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00792"></a>00792   Description: Fetch all genome info that have whole genome alignment data
<a name="l00793"></a>00793   Returntype : arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00794"></a>00794   Exceptions : none
<a name="l00795"></a>00795   Caller     : general
<a name="l00796"></a>00796   Status     : Stable
<a name="l00797"></a>00797 =cut
<a name="l00798"></a>00798 
<a name="l00799"></a>00799 sub fetch_all_with_genome_alignments {
<a name="l00800"></a>00800   my ( $self, $keen ) = @_;
<a name="l00801"></a>00801   <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_genome_alignments&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> },
<a name="l00802"></a>00802                                           $keen );
<a name="l00803"></a>00803 }
<a name="l00804"></a>00804 
<a name="l00805"></a>00805 =head2 fetch_all_with_compara
<a name="l00806"></a>00806   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00807"></a>00807   Description: Fetch all genome info that have any compara or whole genome alignment data
<a name="l00808"></a>00808   Returntype : arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00809"></a>00809   Exceptions : none
<a name="l00810"></a>00810   Caller     : general
<a name="l00811"></a>00811   Status     : Stable
<a name="l00812"></a>00812 =cut
<a name="l00813"></a>00813 
<a name="l00814"></a>00814 sub fetch_all_with_compara {
<a name="l00815"></a>00815   my ( $self, $keen ) = @_;
<a name="l00816"></a>00816   <span class="keywordflow">return</span> $self-&gt;_fetch_generic( $self-&gt;_get_base_sql() .
<a name="l00817"></a>00817 q/ where has_genome_alignments=1 or has_pan_compara=1 or has_peptide_compara=1/
<a name="l00818"></a>00818   );
<a name="l00819"></a>00819 }
<a name="l00820"></a>00820 
<a name="l00821"></a>00821 =head2 fetch_with_other_alignments
<a name="l00822"></a>00822   Arg        : (optional) <span class="keywordflow">if</span> 1, expand children of genome info
<a name="l00823"></a>00823   Description: Fetch all genome info that have other alignment data
<a name="l00824"></a>00824   Returntype : arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00825"></a>00825   Exceptions : none
<a name="l00826"></a>00826   Caller     : general
<a name="l00827"></a>00827   Status     : Stable
<a name="l00828"></a>00828 =cut
<a name="l00829"></a>00829 
<a name="l00830"></a>00830 sub fetch_all_with_other_alignments {
<a name="l00831"></a>00831   my ( $self, $keen ) = @_;
<a name="l00832"></a>00832   <span class="keywordflow">return</span> $self-&gt;_fetch_generic_with_args( { <span class="stringliteral">&#39;has_other_alignments&#39;</span> =&gt; <span class="charliteral">&#39;1&#39;</span> },
<a name="l00833"></a>00833                                           $keen );
<a name="l00834"></a>00834 }
<a name="l00835"></a>00835 
<a name="l00836"></a>00836 =head1 INTERNAL METHODS
<a name="l00837"></a>00837 =head2 _fetch_assembly
<a name="l00838"></a>00838   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a> 
<a name="l00839"></a>00839   Description: Add assembly to supplied <span class="keywordtype">object</span>
<a name="l00840"></a>00840   Returntype : none
<a name="l00841"></a>00841   Exceptions : none
<a name="l00842"></a>00842   Caller     : <span class="keyword">internal</span>
<a name="l00843"></a>00843   Status     : Stable
<a name="l00844"></a>00844 =cut
<a name="l00845"></a>00845 
<a name="l00846"></a>00846 sub _fetch_assembly {
<a name="l00847"></a>00847   my ( $self, $genome ) = @_;
<a name="l00848"></a>00848   <span class="keywordflow">if</span> ( defined $genome-&gt;{assembly_id} ) {
<a name="l00849"></a>00849     $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#a7b0ee55d50bbe9a09eee62b9d7dbadc0">assembly</a>( $self-&gt;db()-&gt;get_GenomeAssemblyInfoAdaptor()
<a name="l00850"></a>00850                        -&gt;fetch_by_dbID( $genome-&gt;{assembly_id} ) );
<a name="l00851"></a>00851   }
<a name="l00852"></a>00852   <span class="keywordflow">return</span>;
<a name="l00853"></a>00853 }
<a name="l00854"></a>00854 
<a name="l00855"></a>00855 =head2 _fetch_data_release
<a name="l00856"></a>00856   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a> 
<a name="l00857"></a>00857   Description: Add data release to supplied <span class="keywordtype">object</span>
<a name="l00858"></a>00858   Returntype : none
<a name="l00859"></a>00859   Exceptions : none
<a name="l00860"></a>00860   Caller     : <span class="keyword">internal</span>
<a name="l00861"></a>00861   Status     : Stable
<a name="l00862"></a>00862 =cut
<a name="l00863"></a>00863 
<a name="l00864"></a>00864 sub _fetch_data_release {
<a name="l00865"></a>00865   my ( $self, $genome ) = @_;
<a name="l00866"></a>00866   <span class="keywordflow">if</span> ( defined $genome-&gt;{data_release_id} ) {
<a name="l00867"></a>00867     $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#a090b7b65151c6ce78babe02445961be7">data_release</a>( $self-&gt;db()-&gt;get_DataReleaseInfoAdaptor()
<a name="l00868"></a>00868                            -&gt;fetch_by_dbID( $genome-&gt;{data_release_id} ) );
<a name="l00869"></a>00869   }
<a name="l00870"></a>00870   <span class="keywordflow">return</span>;
<a name="l00871"></a>00871 }
<a name="l00872"></a>00872 
<a name="l00873"></a>00873 =head2 _fetch_variations
<a name="l00874"></a>00874   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a> 
<a name="l00875"></a>00875   Description: Add variations to supplied <span class="keywordtype">object</span>
<a name="l00876"></a>00876   Returntype : none
<a name="l00877"></a>00877   Exceptions : none
<a name="l00878"></a>00878   Caller     : <span class="keyword">internal</span>
<a name="l00879"></a>00879   Status     : Stable
<a name="l00880"></a>00880 =cut
<a name="l00881"></a>00881 
<a name="l00882"></a>00882 sub _fetch_variations {
<a name="l00883"></a>00883   my ( $self, $genome ) = @_;
<a name="l00884"></a>00884   croak
<a name="l00885"></a>00885     <span class="stringliteral">&quot;Cannot fetch variations for a GenomeInfo object that has not been stored&quot;</span>
<a name="l00886"></a>00886     <span class="keywordflow">if</span> !defined $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#ae2ae7289199643ff4352d80941bdb6cd">dbID</a>();
<a name="l00887"></a>00887   my $variations = {};
<a name="l00888"></a>00888   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_no_return(
<a name="l00889"></a>00889     -SQL =&gt; <span class="stringliteral">&#39;select type,name,count from genome_variation where genome_id=?&#39;</span>,
<a name="l00890"></a>00890     -CALLBACK =&gt; sub {
<a name="l00891"></a>00891       my @row = @{ shift @_ };
<a name="l00892"></a>00892       $variations-&gt;{ $row[0] }-&gt;{ $row[1] } = $row[2];
<a name="l00893"></a>00893       <span class="keywordflow">return</span>;
<a name="l00894"></a>00894     },
<a name="l00895"></a>00895     -PARAMS =&gt; [ $genome-&gt;dbID() ] );
<a name="l00896"></a>00896   $genome-&gt;variations($variations);
<a name="l00897"></a>00897   $genome-&gt;has_variations();
<a name="l00898"></a>00898   <span class="keywordflow">return</span>;
<a name="l00899"></a>00899 }
<a name="l00900"></a>00900 
<a name="l00901"></a>00901 =head2 _fetch_databases
<a name="l00902"></a>00902   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a> 
<a name="l00903"></a>00903   Description: Add databases to supplied <span class="keywordtype">object</span>
<a name="l00904"></a>00904   Returntype : none
<a name="l00905"></a>00905   Exceptions : none
<a name="l00906"></a>00906   Caller     : <span class="keyword">internal</span>
<a name="l00907"></a>00907   Status     : Stable
<a name="l00908"></a>00908 =cut
<a name="l00909"></a>00909 
<a name="l00910"></a>00910 sub _fetch_databases {
<a name="l00911"></a>00911   my ( $self, $genome ) = @_;
<a name="l00912"></a>00912   croak
<a name="l00913"></a>00913     <span class="stringliteral">&quot;Cannot fetch databases for a GenomeInfo object that has not been stored&quot;</span>
<a name="l00914"></a>00914     <span class="keywordflow">if</span> !defined $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#ae2ae7289199643ff4352d80941bdb6cd">dbID</a>();
<a name="l00915"></a>00915   $genome-&gt;{databases} = $self-&gt;db()-&gt;get_DatabaseInfoAdaptor()-&gt;fetch_databases($genome);
<a name="l00916"></a>00916   <span class="keywordflow">return</span>;
<a name="l00917"></a>00917 }
<a name="l00918"></a>00918 
<a name="l00919"></a>00919 =head2 _fetch_other_alignments
<a name="l00920"></a>00920   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a> 
<a name="l00921"></a>00921   Description: Add other_alignments to supplied <span class="keywordtype">object</span>
<a name="l00922"></a>00922   Returntype : none
<a name="l00923"></a>00923   Exceptions : none
<a name="l00924"></a>00924   Caller     : <span class="keyword">internal</span>
<a name="l00925"></a>00925   Status     : Stable
<a name="l00926"></a>00926 =cut
<a name="l00927"></a>00927 
<a name="l00928"></a>00928 sub _fetch_other_alignments {
<a name="l00929"></a>00929   my ( $self, $genome ) = @_;
<a name="l00930"></a>00930   croak
<a name="l00931"></a>00931     <span class="stringliteral">&quot;Cannot fetch alignments for a GenomeInfo object that has not been stored&quot;</span>
<a name="l00932"></a>00932     <span class="keywordflow">if</span> !defined $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#ae2ae7289199643ff4352d80941bdb6cd">dbID</a>();
<a name="l00933"></a>00933   my $alignments = {};
<a name="l00934"></a>00934   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_no_return(
<a name="l00935"></a>00935     -SQL =&gt; <span class="stringliteral">&#39;select type,name,count from genome_alignment where genome_id=?&#39;</span>,
<a name="l00936"></a>00936     -CALLBACK =&gt; sub {
<a name="l00937"></a>00937       my @row = @{ shift @_ };
<a name="l00938"></a>00938       $alignments-&gt;{ $row[0] }-&gt;{ $row[1] } = $row[2];
<a name="l00939"></a>00939       <span class="keywordflow">return</span>;
<a name="l00940"></a>00940     },
<a name="l00941"></a>00941     -PARAMS =&gt; [ $genome-&gt;dbID() ] );
<a name="l00942"></a>00942   $genome-&gt;other_alignments($alignments);
<a name="l00943"></a>00943   $genome-&gt;has_other_alignments();
<a name="l00944"></a>00944   <span class="keywordflow">return</span>;
<a name="l00945"></a>00945 }
<a name="l00946"></a>00946 
<a name="l00947"></a>00947 =head2 _fetch_annotations
<a name="l00948"></a>00948   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a> 
<a name="l00949"></a>00949   Description: Add annotations to supplied <span class="keywordtype">object</span>
<a name="l00950"></a>00950   Returntype : none
<a name="l00951"></a>00951   Exceptions : none
<a name="l00952"></a>00952   Caller     : <span class="keyword">internal</span>
<a name="l00953"></a>00953   Status     : Stable
<a name="l00954"></a>00954 =cut
<a name="l00955"></a>00955 
<a name="l00956"></a>00956 sub _fetch_annotations {
<a name="l00957"></a>00957   my ( $self, $genome ) = @_;
<a name="l00958"></a>00958   croak
<a name="l00959"></a>00959     <span class="stringliteral">&quot;Cannot fetch annotations for a GenomeInfo object that has not been stored&quot;</span>
<a name="l00960"></a>00960     <span class="keywordflow">if</span> !defined $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#ae2ae7289199643ff4352d80941bdb6cd">dbID</a>();
<a name="l00961"></a>00961   my $annotations = {};
<a name="l00962"></a>00962   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_no_return(
<a name="l00963"></a>00963     -SQL      =&gt; <span class="stringliteral">&#39;select type,count from genome_annotation where genome_id=?&#39;</span>,
<a name="l00964"></a>00964     -CALLBACK =&gt; sub {
<a name="l00965"></a>00965       my @row = @{ shift @_ };
<a name="l00966"></a>00966       $annotations-&gt;{ $row[0] } = $row[1];
<a name="l00967"></a>00967       <span class="keywordflow">return</span>;
<a name="l00968"></a>00968     },
<a name="l00969"></a>00969     -PARAMS =&gt; [ $genome-&gt;dbID() ] );
<a name="l00970"></a>00970   $genome-&gt;annotations($annotations);
<a name="l00971"></a>00971   <span class="keywordflow">return</span>;
<a name="l00972"></a>00972 }
<a name="l00973"></a>00973 
<a name="l00974"></a>00974 =head2 _fetch_features
<a name="l00975"></a>00975   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a> 
<a name="l00976"></a>00976   Description: Add features to supplied <span class="keywordtype">object</span>
<a name="l00977"></a>00977   Returntype : none
<a name="l00978"></a>00978   Exceptions : none
<a name="l00979"></a>00979   Caller     : <span class="keyword">internal</span>
<a name="l00980"></a>00980   Status     : Stable
<a name="l00981"></a>00981 =cut
<a name="l00982"></a>00982 
<a name="l00983"></a>00983 sub _fetch_features {
<a name="l00984"></a>00984   my ( $self, $genome ) = @_;
<a name="l00985"></a>00985   croak
<a name="l00986"></a>00986     <span class="stringliteral">&quot;Cannot fetch features  for a GenomeInfo object that has not been stored&quot;</span>
<a name="l00987"></a>00987     <span class="keywordflow">if</span> !defined $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#ae2ae7289199643ff4352d80941bdb6cd">dbID</a>();
<a name="l00988"></a>00988   my $features = {};
<a name="l00989"></a>00989   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_no_return(
<a name="l00990"></a>00990     -SQL =&gt; <span class="stringliteral">&#39;select type,analysis,count from genome_feature where genome_id=?&#39;</span>,
<a name="l00991"></a>00991     -CALLBACK =&gt; sub {
<a name="l00992"></a>00992       my @row = @{ shift @_ };
<a name="l00993"></a>00993       $features-&gt;{ $row[0] }-&gt;{ $row[1] } = $row[2];
<a name="l00994"></a>00994       <span class="keywordflow">return</span>;
<a name="l00995"></a>00995     },
<a name="l00996"></a>00996     -PARAMS =&gt; [ $genome-&gt;dbID() ] );
<a name="l00997"></a>00997   $genome-&gt;features($features);
<a name="l00998"></a>00998   <span class="keywordflow">return</span>;
<a name="l00999"></a>00999 }
<a name="l01000"></a>01000 
<a name="l01001"></a>01001 =head2 _fetch_comparas
<a name="l01002"></a>01002   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a> 
<a name="l01003"></a>01003   Description: Add compara info to supplied <span class="keywordtype">object</span>
<a name="l01004"></a>01004   Returntype : none
<a name="l01005"></a>01005   Exceptions : none
<a name="l01006"></a>01006   Caller     : <span class="keyword">internal</span>
<a name="l01007"></a>01007   Status     : Stable
<a name="l01008"></a>01008 =cut
<a name="l01009"></a>01009 
<a name="l01010"></a>01010 sub _fetch_comparas {
<a name="l01011"></a>01011   my ( $self, $genome ) = @_;
<a name="l01012"></a>01012   my $comparas = [];
<a name="l01013"></a>01013   <span class="keywordflow">for</span> my $id (
<a name="l01014"></a>01014     @{$self-&gt;dbc()-&gt;sql_helper()-&gt;execute_simple(
<a name="l01015"></a>01015         -SQL =&gt; q/select distinct compara_analysis_id from compara_analysis 
<a name="l01016"></a>01016   join genome_compara_analysis <span class="keyword">using</span> (compara_analysis_id)
<a name="l01017"></a>01017   where genome_id=? and method in (<span class="stringliteral">&#39;BLASTZ_NET&#39;</span>,<span class="stringliteral">&#39;LASTZ_NET&#39;</span>,<span class="stringliteral">&#39;TRANSLATED_BLAT_NET&#39;</span>, <span class="stringliteral">&#39;PROTEIN_TREES&#39;</span>, <span class="stringliteral">&#39;ATAC&#39;</span>)/,
<a name="l01018"></a>01018         -PARAMS =&gt; [ $genome-&gt;dbID() ] ) } )
<a name="l01019"></a>01019   {
<a name="l01020"></a>01020     push @$comparas,
<a name="l01021"></a>01021       $self-&gt;db()-&gt;get_GenomeComparaInfoAdaptor()-&gt;fetch_by_dbID($id);
<a name="l01022"></a>01022   }
<a name="l01023"></a>01023   $genome-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#a25c4bbab1a034a6663e73326a0a7fb89">compara</a>($comparas);
<a name="l01024"></a>01024   $genome-&gt;has_pan_compara();
<a name="l01025"></a>01025   $genome-&gt;has_genome_alignments();
<a name="l01026"></a>01026   $genome-&gt;has_peptide_compara();
<a name="l01027"></a>01027   <span class="keywordflow">return</span>;
<a name="l01028"></a>01028 }
<a name="l01029"></a>01029 
<a name="l01030"></a>01030 =head2 _fetch_children
<a name="l01031"></a>01031   Arg        : Arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l01032"></a>01032   Description: Fetch all children of specified genome info <span class="keywordtype">object</span>
<a name="l01033"></a>01033   Returntype : none
<a name="l01034"></a>01034   Exceptions : none
<a name="l01035"></a>01035   Caller     : <span class="keyword">internal</span>
<a name="l01036"></a>01036   Status     : Stable
<a name="l01037"></a>01037 =cut
<a name="l01038"></a>01038 
<a name="l01039"></a>01039 sub _fetch_children {
<a name="l01040"></a>01040   my ( $self, $genome ) = @_;
<a name="l01041"></a>01041   $self-&gt;_fetch_databases($genome);
<a name="l01042"></a>01042   $self-&gt;_fetch_assembly($genome);
<a name="l01043"></a>01043   $self-&gt;_fetch_data_release($genome);
<a name="l01044"></a>01044   $self-&gt;_fetch_variations($genome);
<a name="l01045"></a>01045   $self-&gt;_fetch_annotations($genome);
<a name="l01046"></a>01046   $self-&gt;_fetch_other_alignments($genome);
<a name="l01047"></a>01047   $self-&gt;_fetch_comparas($genome);
<a name="l01048"></a>01048   <span class="keywordflow">return</span>;
<a name="l01049"></a>01049 }
<a name="l01050"></a>01050 
<a name="l01051"></a>01051 
<a name="l01052"></a>01052 =head2 _store_features
<a name="l01053"></a>01053   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l01054"></a>01054   Description: Stores the features <span class="keywordflow">for</span> the supplied <span class="keywordtype">object</span>
<a name="l01055"></a>01055   Returntype : None
<a name="l01056"></a>01056   Exceptions : none
<a name="l01057"></a>01057   Caller     : <span class="keyword">internal</span>
<a name="l01058"></a>01058   Status     : Stable
<a name="l01059"></a>01059 =cut
<a name="l01060"></a>01060 
<a name="l01061"></a>01061 sub _store_features {
<a name="l01062"></a>01062   my ( $self, $genome ) = @_;
<a name="l01063"></a>01063 
<a name="l01064"></a>01064   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01065"></a>01065                         -SQL =&gt; q/<span class="keyword">delete</span> from genome_feature where genome_id=?/,
<a name="l01066"></a>01066                         -PARAMS =&gt; [ $genome-&gt;dbID() ] );
<a name="l01067"></a>01067 
<a name="l01068"></a>01068   <span class="keywordflow">while</span> ( my ( $type, $f ) = each %{ $genome-&gt;features() } ) {
<a name="l01069"></a>01069     <span class="keywordflow">while</span> ( my ( $analysis, $count ) = each %$f ) {
<a name="l01070"></a>01070       $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01071"></a>01071         -SQL =&gt; q/insert into genome_feature(genome_id,type,analysis,count)
<a name="l01072"></a>01072         values(?,?,?,?)/,
<a name="l01073"></a>01073         -PARAMS =&gt; [ $genome-&gt;dbID(), $type, $analysis, $count ] );
<a name="l01074"></a>01074     }
<a name="l01075"></a>01075   }
<a name="l01076"></a>01076   <span class="keywordflow">return</span>;
<a name="l01077"></a>01077 }
<a name="l01078"></a>01078 
<a name="l01079"></a>01079 =head2 _store_databases
<a name="l01080"></a>01080   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l01081"></a>01081   Description: Stores the databases <span class="keywordflow">for</span> the supplied <span class="keywordtype">object</span>
<a name="l01082"></a>01082   Returntype : None
<a name="l01083"></a>01083   Exceptions : none
<a name="l01084"></a>01084   Caller     : <span class="keyword">internal</span>
<a name="l01085"></a>01085   Status     : Stable
<a name="l01086"></a>01086 =cut
<a name="l01087"></a>01087 
<a name="l01088"></a>01088 sub _store_databases {
<a name="l01089"></a>01089   my ( $self, $genome ) = @_;
<a name="l01090"></a>01090   <span class="keywordflow">for</span> my $database ( @{ $genome-&gt;databases() } ) {
<a name="l01091"></a>01091     $self-&gt;db()-&gt;get_DatabaseInfoAdaptor()-&gt;store($database);
<a name="l01092"></a>01092   }
<a name="l01093"></a>01093   <span class="keywordflow">return</span>;
<a name="l01094"></a>01094 }
<a name="l01095"></a>01095 
<a name="l01096"></a>01096 =head2 _store_annotations
<a name="l01097"></a>01097   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l01098"></a>01098   Description: Stores the annotations <span class="keywordflow">for</span> the supplied <span class="keywordtype">object</span>
<a name="l01099"></a>01099   Returntype : None
<a name="l01100"></a>01100   Exceptions : none
<a name="l01101"></a>01101   Caller     : <span class="keyword">internal</span>
<a name="l01102"></a>01102   Status     : Stable
<a name="l01103"></a>01103 =cut
<a name="l01104"></a>01104 
<a name="l01105"></a>01105 sub _store_annotations {
<a name="l01106"></a>01106   my ( $self, $genome ) = @_;
<a name="l01107"></a>01107 
<a name="l01108"></a>01108   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01109"></a>01109                      -SQL =&gt; q/<span class="keyword">delete</span> from genome_annotation where genome_id=?/,
<a name="l01110"></a>01110                      -PARAMS =&gt; [ $genome-&gt;dbID() ] );
<a name="l01111"></a>01111 
<a name="l01112"></a>01112   <span class="keywordflow">while</span> ( my ( $type, $count ) = each %{ $genome-&gt;annotations() } ) {
<a name="l01113"></a>01113     $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01114"></a>01114       -SQL =&gt; q/insert into genome_annotation(genome_id,type,count)
<a name="l01115"></a>01115         values(?,?,?)/,
<a name="l01116"></a>01116       -PARAMS =&gt; [ $genome-&gt;dbID(), $type, $count ] );
<a name="l01117"></a>01117   }
<a name="l01118"></a>01118   <span class="keywordflow">return</span>;
<a name="l01119"></a>01119 }
<a name="l01120"></a>01120 
<a name="l01121"></a>01121 =head2 _store_variations
<a name="l01122"></a>01122   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l01123"></a>01123   Description: Stores the variations <span class="keywordflow">for</span> the supplied <span class="keywordtype">object</span>
<a name="l01124"></a>01124   Returntype : None
<a name="l01125"></a>01125   Exceptions : none
<a name="l01126"></a>01126   Caller     : <span class="keyword">internal</span>
<a name="l01127"></a>01127   Status     : Stable
<a name="l01128"></a>01128 =cut
<a name="l01129"></a>01129 
<a name="l01130"></a>01130 sub _store_variations {
<a name="l01131"></a>01131   my ( $self, $genome ) = @_;
<a name="l01132"></a>01132 
<a name="l01133"></a>01133   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01134"></a>01134                       -SQL =&gt; q/<span class="keyword">delete</span> from genome_variation where genome_id=?/,
<a name="l01135"></a>01135                       -PARAMS =&gt; [ $genome-&gt;dbID() ] );
<a name="l01136"></a>01136 
<a name="l01137"></a>01137   <span class="keywordflow">while</span> ( my ( $type, $f ) = each %{ $genome-&gt;variations() } ) {
<a name="l01138"></a>01138     <span class="keywordflow">while</span> ( my ( $key, $count ) = each %$f ) {
<a name="l01139"></a>01139       $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01140"></a>01140         -SQL =&gt; q/insert into genome_variation(genome_id,type,name,count)
<a name="l01141"></a>01141         values(?,?,?,?)/,
<a name="l01142"></a>01142         -PARAMS =&gt; [ $genome-&gt;dbID(), $type, $key, $count ] );
<a name="l01143"></a>01143     }
<a name="l01144"></a>01144   }
<a name="l01145"></a>01145   <span class="keywordflow">return</span>;
<a name="l01146"></a>01146 }
<a name="l01147"></a>01147 
<a name="l01148"></a>01148 =head2 _store_alignments
<a name="l01149"></a>01149   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l01150"></a>01150   Description: Stores the alignments <span class="keywordflow">for</span> the supplied <span class="keywordtype">object</span>
<a name="l01151"></a>01151   Returntype : None
<a name="l01152"></a>01152   Exceptions : none
<a name="l01153"></a>01153   Caller     : <span class="keyword">internal</span>
<a name="l01154"></a>01154   Status     : Stable
<a name="l01155"></a>01155 =cut
<a name="l01156"></a>01156 
<a name="l01157"></a>01157 sub _store_alignments {
<a name="l01158"></a>01158   my ( $self, $genome ) = @_;
<a name="l01159"></a>01159 
<a name="l01160"></a>01160   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01161"></a>01161                       -SQL =&gt; q/<span class="keyword">delete</span> from genome_alignment where genome_id=?/,
<a name="l01162"></a>01162                       -PARAMS =&gt; [ $genome-&gt;dbID() ] );
<a name="l01163"></a>01163 
<a name="l01164"></a>01164   <span class="keywordflow">while</span> ( my ( $type, $f ) = each %{ $genome-&gt;other_alignments() } ) {
<a name="l01165"></a>01165     <span class="keywordflow">while</span> ( my ( $key, $count ) = each %$f ) {
<a name="l01166"></a>01166       $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01167"></a>01167         -SQL =&gt; q/insert into genome_alignment(genome_id,type,name,count)
<a name="l01168"></a>01168         values(?,?,?,?)/,
<a name="l01169"></a>01169         -PARAMS =&gt; [ $genome-&gt;dbID(), $type, $key, $count ] );
<a name="l01170"></a>01170     }
<a name="l01171"></a>01171   }
<a name="l01172"></a>01172   <span class="keywordflow">return</span>;
<a name="l01173"></a>01173 }
<a name="l01174"></a>01174 =head2 _store_compara
<a name="l01175"></a>01175   Arg        : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l01176"></a>01176   Description: Stores the compara analyses <span class="keywordflow">for</span> the supplied <span class="keywordtype">object</span>
<a name="l01177"></a>01177   Returntype : None
<a name="l01178"></a>01178   Exceptions : none
<a name="l01179"></a>01179   Caller     : <span class="keyword">internal</span>
<a name="l01180"></a>01180   Status     : Stable
<a name="l01181"></a>01181 =cut
<a name="l01182"></a>01182 sub _store_compara {
<a name="l01183"></a>01183   my ( $self, $genome ) = @_;
<a name="l01184"></a>01184 
<a name="l01185"></a>01185   $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01186"></a>01186                -SQL =&gt; q/<span class="keyword">delete</span> from genome_compara_analysis where genome_id=?/,
<a name="l01187"></a>01187                -PARAMS =&gt; [ $genome-&gt;dbID() ] );
<a name="l01188"></a>01188   <span class="keywordflow">if</span> ( defined $genome-&gt;compara() ) {
<a name="l01189"></a>01189     <span class="keywordflow">for</span> my $compara ( @{ $genome-&gt;compara() } ) {
<a name="l01190"></a>01190       <span class="keywordflow">if</span> ( !defined $compara-&gt;dbID() ) {
<a name="l01191"></a>01191         $self-&gt;db()-&gt;get_GenomeComparaInfoAdaptor()-&gt;store($compara);
<a name="l01192"></a>01192       }
<a name="l01193"></a>01193       $self-&gt;dbc()-&gt;sql_helper()-&gt;execute_update(
<a name="l01194"></a>01194         -SQL =&gt;
<a name="l01195"></a>01195 q/insert into genome_compara_analysis(genome_id,compara_analysis_id) values(?,?)/,
<a name="l01196"></a>01196         -PARAMS =&gt; [ $genome-&gt;dbID(), $compara-&gt;dbID() ] );
<a name="l01197"></a>01197     }
<a name="l01198"></a>01198   }
<a name="l01199"></a>01199   <span class="keywordflow">return</span>;
<a name="l01200"></a>01200 
<a name="l01201"></a>01201 }
<a name="l01202"></a>01202 
<a name="l01203"></a>01203 
<a name="l01204"></a>01204 my $base_genome_fetch_sql =
<a name="l01205"></a>01205   q/select genome_id as dbID, division.<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#a099657bcf23824f8c7342ea680b04a1e">name</a> as division, genebuild, 
<a name="l01206"></a>01206 has_pan_compara, has_variations, has_peptide_compara, 
<a name="l01207"></a>01207 has_genome_alignments, has_synteny, has_other_alignments, 
<a name="l01208"></a>01208 assembly_id, data_release_id
<a name="l01209"></a>01209 from genome join division <span class="keyword">using</span> (division_id)/;
<a name="l01210"></a>01210 
<a name="l01211"></a>01211 sub _get_base_sql {
<a name="l01212"></a>01212   <span class="keywordflow">return</span> $base_genome_fetch_sql;
<a name="l01213"></a>01213 }
<a name="l01214"></a>01214 
<a name="l01215"></a>01215 sub _get_id_field {
<a name="l01216"></a>01216   <span class="keywordflow">return</span> <span class="stringliteral">&#39;genome_id&#39;</span>;
<a name="l01217"></a>01217 }
<a name="l01218"></a>01218 
<a name="l01219"></a>01219 sub _get_obj_class {
<a name="l01220"></a>01220   <span class="keywordflow">return</span> <span class="stringliteral">&#39;Bio::EnsEMBL::MetaData::GenomeInfo&#39;</span>;
<a name="l01221"></a>01221 }
<a name="l01222"></a>01222 
<a name="l01223"></a>01223 <span class="preprocessor"># override to add release clause</span>
<a name="l01224"></a>01224 <span class="preprocessor"></span>sub _args_to_sql {
<a name="l01225"></a>01225   my ( $self, $sql_in, $args ) = @_;
<a name="l01226"></a>01226   <span class="keywordflow">if</span> ( !defined $args-&gt;{ _get_id_field() } ) {
<a name="l01227"></a>01227 <span class="preprocessor">    # if we&#39;re not searching by dbID, add release as a clause</span>
<a name="l01228"></a>01228 <span class="preprocessor"></span>    <span class="keywordflow">if</span> ( defined $self-&gt;data_release()-&gt;dbID() ) {
<a name="l01229"></a>01229       $args-&gt;{data_release_id} = $self-&gt;data_release()-&gt;dbID();
<a name="l01230"></a>01230     }
<a name="l01231"></a>01231   }
<a name="l01232"></a>01232   <span class="keywordflow">return</span> $self-&gt;SUPER::_args_to_sql( $sql_in, $args );
<a name="l01233"></a>01233 }
<a name="l01234"></a>01234 
<a name="l01235"></a>01235 sub _organism_adaptor {
<a name="l01236"></a>01236   my ($self) = @_;
<a name="l01237"></a>01237   <span class="keywordflow">if</span> ( !defined $self-&gt;{organism_adaptor} ) {
<a name="l01238"></a>01238     $self-&gt;{organism_adaptor} = $self-&gt;db()-&gt;get_GenomeOrganismInfoAdaptor();
<a name="l01239"></a>01239   }
<a name="l01240"></a>01240   <span class="keywordflow">return</span> $self-&gt;{organism_adaptor};
<a name="l01241"></a>01241 }
<a name="l01242"></a>01242 
<a name="l01243"></a>01243 sub _assembly_adaptor {
<a name="l01244"></a>01244   my ($self) = @_;
<a name="l01245"></a>01245   <span class="keywordflow">if</span> ( !defined $self-&gt;{assembly_adaptor} ) {
<a name="l01246"></a>01246     $self-&gt;{assembly_adaptor} = $self-&gt;db()-&gt;get_GenomeAssemblyInfoAdaptor();
<a name="l01247"></a>01247   }
<a name="l01248"></a>01248   <span class="keywordflow">return</span> $self-&gt;{assembly_adaptor};
<a name="l01249"></a>01249 }
<a name="l01250"></a>01250 
<a name="l01251"></a>01251 1;
</pre></div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="GenomeInfoAdaptor_8pm.html">GenomeInfoAdaptor.pm</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
