<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Ensembl Metadata: Bio/EnsEMBL/MetaData/GenomeInfo.pm Source File</title>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('GenomeInfo_8pm.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">GenomeInfo.pm</div>  </div>
</div>
<div class="contents">
<a href="GenomeInfo_8pm.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 
<a name="l00002"></a>00002 =head1 LICENSE
<a name="l00003"></a>00003 
<a name="l00004"></a>00004 Copyright [1999-2014] EMBL-European Bioinformatics Institute
<a name="l00005"></a>00005 
<a name="l00006"></a>00006 Licensed under the Apache License, Version 2.0 (the <span class="stringliteral">&quot;License&quot;</span>);
<a name="l00007"></a>00007 you may not use <span class="keyword">this</span> file except in compliance with the License.
<a name="l00008"></a>00008 You may obtain a copy of the License at
<a name="l00009"></a>00009 
<a name="l00010"></a>00010      http:<span class="comment">//www.apache.org/licenses/LICENSE-2.0</span>
<a name="l00011"></a>00011 
<a name="l00012"></a>00012 Unless required by applicable law or agreed to in writing, software
<a name="l00013"></a>00013 distributed under the License is distributed on an <span class="stringliteral">&quot;AS IS&quot;</span> BASIS,
<a name="l00014"></a>00014 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
<a name="l00015"></a>00015 See the License <span class="keywordflow">for</span> the specific language governing permissions and
<a name="l00016"></a>00016 limitations under the License.
<a name="l00017"></a>00017 
<a name="l00018"></a>00018 =cut
<a name="l00019"></a>00019 
<a name="l00020"></a>00020 =pod
<a name="l00021"></a>00021 
<a name="l00022"></a>00022 =head1 NAME
<a name="l00023"></a>00023 
<a name="l00024"></a>00024 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00025"></a>00025 
<a name="l00026"></a>00026 =head1 SYNOPSIS
<a name="l00027"></a>00027 
<a name="l00028"></a>00028   my $genome = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html#a8a7161c7738910c10ba3e8cd9d33b7bf">new</a>(
<a name="l00029"></a>00029       -species    =&gt; $dba-&gt;species(),
<a name="l00030"></a>00030       -species_id =&gt; $dba-&gt;species_id(),
<a name="l00031"></a>00031       -division   =&gt; $meta-&gt;get_division() || <span class="stringliteral">&#39;Ensembl&#39;</span>,
<a name="l00032"></a>00032       -dbname     =&gt; $dbname);
<a name="l00033"></a>00033       
<a name="l00034"></a>00034   print Dumper($genome-&gt;to_hash());
<a name="l00035"></a>00035 
<a name="l00036"></a>00036 =head1 DESCRIPTION
<a name="l00037"></a>00037 
<a name="l00038"></a>00038 Object encapsulating meta information about a genome in Ensembl Genomes. 
<a name="l00039"></a><a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">00039</a> 
<a name="l00040"></a>00040 Can be used to render information about a genome e.g.
<a name="l00041"></a>00041 
<a name="l00042"></a>00042 print $genome-&gt;name().<span class="stringliteral">&quot; (&quot;</span>.$genome-&gt;species.<span class="stringliteral">&quot;)\n&quot;</span>;
<a name="l00043"></a>00043 print <span class="stringliteral">&quot;Sequences: &quot;</span>.scalar(@{$genome-&gt;sequences()}).<span class="stringliteral">&quot;\n&quot;</span>;
<a name="l00044"></a>00044 <span class="keywordflow">if</span>($genome-&gt;has_variations()) {
<a name="l00045"></a>00045     print <span class="stringliteral">&quot;Variations: \n&quot;</span>;
<a name="l00046"></a>00046 <span class="preprocessor">    # variations is a hash with type as the key</span>
<a name="l00047"></a>00047 <span class="preprocessor"></span>    <span class="keywordflow">while</span>(my ($type,$value) = each %{$genome-&gt;variations()}) {
<a name="l00048"></a>00048         print <span class="stringliteral">&quot;- $type\n&quot;</span>;
<a name="l00049"></a>00049     }
<a name="l00050"></a>00050 }
<a name="l00051"></a>00051 print <span class="stringliteral">&quot;Compara analyses: &quot;</span>.scalar(@{$genome-&gt;compara()}).<span class="stringliteral">&quot;\n&quot;</span>;
<a name="l00052"></a>00052 
<a name="l00053"></a>00053 =head1 AUTHOR
<a name="l00054"></a>00054 
<a name="l00055"></a>00055 Dan Staines
<a name="l00056"></a>00056 
<a name="l00057"></a>00057 =cut
<a name="l00058"></a>00058 
<a name="l00059"></a>00059 use strict;
<a name="l00060"></a>00060 use warnings;
<a name="l00061"></a>00061 
<a name="l00062"></a>00062 <span class="keyword">package </span>Bio::EnsEMBL::MetaData::GenomeInfo;
<a name="l00063"></a>00063 use base qw/<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1BaseInfo.html">Bio::EnsEMBL::MetaData::BaseInfo</a>/;
<a name="l00064"></a>00064 use Bio::EnsEMBL::Utils::Argument qw(rearrange);
<a name="l00065"></a>00065 use Bio::EnsEMBL::Utils::Exception qw/<span class="keywordflow">throw</span>/;
<a name="l00066"></a>00066 use <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DataReleaseInfo.html">Bio::EnsEMBL::MetaData::DataReleaseInfo</a>;
<a name="l00067"></a>00067 use <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeAssemblyInfo.html">Bio::EnsEMBL::MetaData::GenomeAssemblyInfo</a>;
<a name="l00068"></a>00068 use <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeOrganismInfo.html">Bio::EnsEMBL::MetaData::GenomeOrganismInfo</a>;
<a name="l00069"></a>00069 
<a name="l00070"></a>00070 =head1 CONSTRUCTOR
<a name="l00071"></a>00071 =head2 <span class="keyword">new</span>
<a name="l00072"></a>00072   Arg [-DISPLAY_NAME]  : 
<a name="l00073"></a>00073        <span class="keywordtype">string</span> - human readable version of the name of the genome
<a name="l00074"></a>00074   Arg [-NAME]    : 
<a name="l00075"></a>00075        <span class="keywordtype">string</span> - computable version of the name of the genome (lower <span class="keywordflow">case</span>, no spaces)
<a name="l00076"></a>00076   Arg [-DBNAME] : 
<a name="l00077"></a>00077        <span class="keywordtype">string</span> - name of the core database in which the genome can be found
<a name="l00078"></a>00078   Arg [-SPECIES_ID]  : 
<a name="l00079"></a>00079        <span class="keywordtype">int</span> - identifier of the species within the core database <span class="keywordflow">for</span> <span class="keyword">this</span> genome
<a name="l00080"></a>00080   Arg [-TAXONOMY_ID] :
<a name="l00081"></a>00081         <span class="keywordtype">string</span> - NCBI taxonomy identifier
<a name="l00082"></a>00082   Arg [-SPECIES_TAXONOMY_ID] :
<a name="l00083"></a>00083         <span class="keywordtype">string</span> - NCBI taxonomy identifier of species to which <span class="keyword">this</span> genome belongs
<a name="l00084"></a>00084   Arg [-ASSEMBLY_NAME] :
<a name="l00085"></a>00085         <span class="keywordtype">string</span> - name of the assembly
<a name="l00086"></a>00086   Arg [-ASSEMBLY_ID] :
<a name="l00087"></a>00087         <span class="keywordtype">string</span> - INSDC assembly accession
<a name="l00088"></a>00088   Arg [-ASSEMBLY_LEVEL] :
<a name="l00089"></a>00089         <span class="keywordtype">string</span> - highest assembly level (chromosome, supercontig etc.)
<a name="l00090"></a>00090   Arg [-GENEBUILD]:
<a name="l00091"></a>00091         <span class="keywordtype">string</span> - identifier for genebuild
<a name="l00092"></a>00092   Arg [-DIVISION]:
<a name="l00093"></a>00093         <span class="keywordtype">string</span> - name of Ensembl Genomes division (e.g. EnsemblBacteria, EnsemblPlants)
<a name="l00094"></a>00094   Arg [-STRAIN]:
<a name="l00095"></a>00095         <span class="keywordtype">string</span> - name of strain to which genome belongs
<a name="l00096"></a>00096   Arg [-SEROTYPE]:
<a name="l00097"></a>00097         <span class="keywordtype">string</span> - name of serotype to which genome belongs
<a name="l00098"></a>00098   Arg [-IS_REFERENCE]:
<a name="l00099"></a>00099         <span class="keywordtype">bool</span> - 1 if this genome is the reference for its species
<a name="l00100"></a>00100 
<a name="l00101"></a>00101   Example    : $info = Bio::EnsEMBL::MetaData::GenomeInfo-&gt;new(...);
<a name="l00102"></a>00102   Description: Creates a new info <span class="keywordtype">object</span>
<a name="l00103"></a>00103   Returntype : Bio::EnsEMBL::MetaData::GenomeInfo
<a name="l00104"></a>00104   Exceptions : none
<a name="l00105"></a>00105   Caller     : general
<a name="l00106"></a>00106   Status     : Stable
<a name="l00107"></a>00107 
<a name="l00108"></a>00108 =cut
<a name="l00109"></a>00109 
<a name="l00110"></a>00110 sub new {
<a name="l00111"></a>00111   my ( $class, @args ) = @_;
<a name="l00112"></a>00112   my $self = $class-&gt;SUPER::new(@args);
<a name="l00113"></a>00113   my ( $name,          $display_name, $dbname,
<a name="l00114"></a>00114        $species_id,    $taxonomy_id,  $species_taxonomy_id,
<a name="l00115"></a>00115        $assembly_name, $assembly_id,  $assembly_level,
<a name="l00116"></a>00116        $strain,        $serotype,     $is_reference,
<a name="l00117"></a>00117        $organism );
<a name="l00118"></a>00118   ( $name,              $display_name,     $dbname,
<a name="l00119"></a>00119     $species_id,        $taxonomy_id,      $species_taxonomy_id,
<a name="l00120"></a>00120     $assembly_name,     $assembly_id,      $assembly_level,
<a name="l00121"></a>00121     $self-&gt;{genebuild}, $self-&gt;{division}, $strain,
<a name="l00122"></a>00122     $serotype,          $is_reference,     $self-&gt;{assembly},
<a name="l00123"></a>00123     $organism,          $self-&gt;{data_release} )
<a name="l00124"></a>00124     = rearrange( [ <span class="stringliteral">&#39;NAME&#39;</span>,           <span class="stringliteral">&#39;DISPLAY_NAME&#39;</span>,
<a name="l00125"></a>00125                    <span class="stringliteral">&#39;DBNAME&#39;</span>,         <span class="stringliteral">&#39;SPECIES_ID&#39;</span>,
<a name="l00126"></a>00126                    <span class="stringliteral">&#39;TAXONOMY_ID&#39;</span>,    <span class="stringliteral">&#39;SPECIES_TAXONOMY_ID&#39;</span>,
<a name="l00127"></a>00127                    <span class="stringliteral">&#39;ASSEMBLY_NAME&#39;</span>,  <span class="stringliteral">&#39;ASSEMBLY_ID&#39;</span>,
<a name="l00128"></a>00128                    <span class="stringliteral">&#39;ASSEMBLY_LEVEL&#39;</span>, <span class="stringliteral">&#39;GENEBUILD&#39;</span>,
<a name="l00129"></a>00129                    <span class="stringliteral">&#39;DIVISION&#39;</span>,       <span class="stringliteral">&#39;STRAIN&#39;</span>,
<a name="l00130"></a>00130                    <span class="stringliteral">&#39;SEROTYPE&#39;</span>,       <span class="stringliteral">&#39;IS_REFERENCE&#39;</span>,
<a name="l00131"></a>00131                    <span class="stringliteral">&#39;ASSEMBLY&#39;</span>,       <span class="stringliteral">&#39;ORGANISM&#39;</span>,
<a name="l00132"></a>00132                    <span class="stringliteral">&#39;DATA_RELEASE&#39;</span> ],
<a name="l00133"></a>00133                  @args );
<a name="l00134"></a>00134 
<a name="l00135"></a>00135   <span class="keywordflow">if</span> ( defined $dbname ) {
<a name="l00136"></a>00136     $self-&gt;add_database( $dbname, $species_id );
<a name="l00137"></a>00137   }
<a name="l00138"></a>00138 
<a name="l00139"></a>00139   <span class="keywordflow">if</span> ( !defined $self-&gt;assembly() ) {
<a name="l00140"></a>00140     my $ass =
<a name="l00141"></a>00141       <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeAssemblyInfo.html">Bio::EnsEMBL::MetaData::GenomeAssemblyInfo</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeAssemblyInfo.html#a17fe417a1b13a698d46932828ab4aa15">new</a>(
<a name="l00142"></a>00142                                    -ASSEMBLY_NAME       =&gt; $assembly_name,
<a name="l00143"></a>00143                                    -ASSEMBLY_ID         =&gt; $assembly_id,
<a name="l00144"></a>00144                                    -ASSEMBLY_LEVEL      =&gt; $assembly_level,
<a name="l00145"></a>00145                                    -DISPLAY_NAME        =&gt; $display_name,
<a name="l00146"></a>00146                                    -NAME                =&gt; $name,
<a name="l00147"></a>00147                                    -STRAIN              =&gt; $strain,
<a name="l00148"></a>00148                                    -SEROTYPE            =&gt; $serotype,
<a name="l00149"></a>00149                                    -TAXONOMY_ID         =&gt; $taxonomy_id,
<a name="l00150"></a>00150                                    -SPECIES_TAXONOMY_ID =&gt; $species_taxonomy_id,
<a name="l00151"></a>00151                                    -IS_REFERENCE        =&gt; $is_reference,
<a name="l00152"></a>00152                                    -ORGANISM            =&gt; $organism );
<a name="l00153"></a>00153     $ass-&gt;adaptor( $self-&gt;adaptor() ) <span class="keywordflow">if</span> defined $self-&gt;adaptor();
<a name="l00154"></a>00154     $self-&gt;assembly($ass);
<a name="l00155"></a>00155   }
<a name="l00156"></a>00156   <span class="keywordflow">return</span> $self;
<a name="l00157"></a>00157 } ## end sub <span class="keyword">new</span>
<a name="l00158"></a>00158 
<a name="l00159"></a>00159 =head1 ATTRIBUTE METHODS
<a name="l00160"></a>00160 
<a name="l00161"></a>00161 =head2 dbname
<a name="l00162"></a>00162   Description: Gets name of core database from which genome comes
<a name="l00163"></a>00163   Returntype : <span class="keywordtype">string</span>
<a name="l00164"></a>00164   Exceptions : none
<a name="l00165"></a>00165   Caller     : general
<a name="l00166"></a>00166   Status     : Stable
<a name="l00167"></a>00167 =cut
<a name="l00168"></a>00168 
<a name="l00169"></a>00169 sub dbname {
<a name="l00170"></a>00170   my ( $self ) = @_;
<a name="l00171"></a>00171   <span class="keywordflow">return</span> $self-&gt;_get_core()-&gt;dbname();
<a name="l00172"></a>00172 }
<a name="l00173"></a>00173 
<a name="l00174"></a>00174 =head2 species_id
<a name="l00175"></a>00175   Description: Gets species_id of genome within core database
<a name="l00176"></a>00176   Returntype : <span class="keywordtype">string</span>
<a name="l00177"></a>00177   Exceptions : none
<a name="l00178"></a>00178   Caller     : general
<a name="l00179"></a>00179   Status     : Stable
<a name="l00180"></a>00180 =cut
<a name="l00181"></a>00181 
<a name="l00182"></a>00182 sub species_id {
<a name="l00183"></a>00183   my ( $self ) = @_;
<a name="l00184"></a>00184   <span class="keywordflow">return</span> $self-&gt;_get_core()-&gt;species_id();
<a name="l00185"></a>00185 }
<a name="l00186"></a>00186 
<a name="l00187"></a>00187 sub add_database {
<a name="l00188"></a>00188   my ( $self, $dbname, $species_id ) = @_;
<a name="l00189"></a>00189   push @{$self-&gt;{databases}}, <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DatabaseInfo.html">Bio::EnsEMBL::MetaData::DatabaseInfo</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DatabaseInfo.html#acfbda123d7a8008ed849f44ad444e9b4">new</a>(-SUBJECT=&gt;$self, -DBNAME=&gt;$dbname, -SPECIES_ID=&gt;$species_id);
<a name="l00190"></a>00190   <span class="keywordflow">return</span>;
<a name="l00191"></a>00191 }
<a name="l00192"></a>00192 
<a name="l00193"></a>00193 sub databases {
<a name="l00194"></a>00194   my ($self, $databases) = @_;
<a name="l00195"></a>00195   <span class="keywordflow">if</span>(defined $databases) {
<a name="l00196"></a>00196     $self-&gt;{databases} = $databases;
<a name="l00197"></a>00197   }
<a name="l00198"></a>00198   $self-&gt;_load_child( <span class="stringliteral">&#39;databases&#39;</span>, <span class="stringliteral">&#39;_fetch_databases&#39;</span> );
<a name="l00199"></a>00199   <span class="keywordflow">return</span> $self-&gt;{databases};
<a name="l00200"></a>00200 }
<a name="l00201"></a>00201 
<a name="l00202"></a>00202 =head2 data_release
<a name="l00203"></a>00203   Arg        : (optional) data_release <span class="keywordtype">object</span> to <span class="keyword">set</span>
<a name="l00204"></a>00204   Description: Gets/sets data_release to which genome belongs
<a name="l00205"></a>00205   Returntype : Bio::EnsEMBL::MetaData::Data_ReleaseInfo
<a name="l00206"></a>00206   Exceptions : none
<a name="l00207"></a>00207   Caller     : general
<a name="l00208"></a>00208   Status     : Stable
<a name="l00209"></a>00209 =cut
<a name="l00210"></a>00210 
<a name="l00211"></a>00211 sub data_release {
<a name="l00212"></a>00212   my ( $self, $data_release ) = @_;
<a name="l00213"></a>00213   $self-&gt;{data_release} = $data_release <span class="keywordflow">if</span> ( defined $data_release );
<a name="l00214"></a>00214   $self-&gt;_load_child( <span class="stringliteral">&#39;data_release&#39;</span>, <span class="stringliteral">&#39;_fetch_data_release&#39;</span> );
<a name="l00215"></a>00215   <span class="keywordflow">return</span> $self-&gt;{data_release};
<a name="l00216"></a>00216 }
<a name="l00217"></a>00217 
<a name="l00218"></a>00218 =head2 organism
<a name="l00219"></a>00219   Arg        : (optional) organism <span class="keywordtype">object</span> to <span class="keyword">set</span>
<a name="l00220"></a>00220   Description: Gets/sets organism to which genome belongs
<a name="l00221"></a>00221   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeOrganismInfo.html">Bio::EnsEMBL::MetaData::GenomeOrganismInfo</a>
<a name="l00222"></a>00222   Exceptions : none
<a name="l00223"></a>00223   Caller     : general
<a name="l00224"></a>00224   Status     : Stable
<a name="l00225"></a>00225 =cut
<a name="l00226"></a>00226 
<a name="l00227"></a>00227 sub organism {
<a name="l00228"></a>00228   my ( $self, $organism ) = @_;
<a name="l00229"></a>00229   <span class="keywordflow">return</span> $self-&gt;assembly()-&gt;organism();
<a name="l00230"></a>00230 }
<a name="l00231"></a>00231 
<a name="l00232"></a>00232 =head2 species
<a name="l00233"></a>00233   Arg        : (optional) species to <span class="keyword">set</span>
<a name="l00234"></a>00234   Description: Gets/sets readable name
<a name="l00235"></a>00235   Returntype : <span class="keywordtype">string</span>
<a name="l00236"></a>00236   Exceptions : none
<a name="l00237"></a>00237   Caller     : general
<a name="l00238"></a>00238   Status     : Stable
<a name="l00239"></a>00239 =cut
<a name="l00240"></a>00240 
<a name="l00241"></a>00241 sub display_name {
<a name="l00242"></a>00242   my ( $self, $name ) = @_;
<a name="l00243"></a>00243   <span class="keywordflow">return</span> $self-&gt;organism()-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeOrganismInfo.html#aeb74f73b231cd8aa2ad0224a679d8c1f">display_name</a>($name);
<a name="l00244"></a>00244 }
<a name="l00245"></a>00245 
<a name="l00246"></a>00246 =head2 strain
<a name="l00247"></a>00247   Arg        : (optional) strain to <span class="keyword">set</span>
<a name="l00248"></a>00248   Description: Gets/sets strain of genome
<a name="l00249"></a>00249   Returntype : <span class="keywordtype">string</span>
<a name="l00250"></a>00250   Exceptions : none
<a name="l00251"></a>00251   Caller     : general
<a name="l00252"></a>00252   Status     : Stable
<a name="l00253"></a>00253 =cut
<a name="l00254"></a>00254 
<a name="l00255"></a>00255 sub strain {
<a name="l00256"></a>00256   my ( $self, $arg ) = @_;
<a name="l00257"></a>00257   <span class="keywordflow">return</span> $self-&gt;organism()-&gt;strain($arg);
<a name="l00258"></a>00258 }
<a name="l00259"></a>00259 
<a name="l00260"></a>00260 =head2 serotype
<a name="l00261"></a>00261   Arg        : (optional) serotype to <span class="keyword">set</span>
<a name="l00262"></a>00262   Description: Gets/sets serotype
<a name="l00263"></a>00263   Returntype : <span class="keywordtype">string</span>
<a name="l00264"></a>00264   Exceptions : none
<a name="l00265"></a>00265   Caller     : general
<a name="l00266"></a>00266   Status     : Stable
<a name="l00267"></a>00267 =cut
<a name="l00268"></a>00268 
<a name="l00269"></a>00269 sub serotype {
<a name="l00270"></a>00270   my ( $self, $arg ) = @_;
<a name="l00271"></a>00271   <span class="keywordflow">return</span> $self-&gt;organism()-&gt;serotype($arg);
<a name="l00272"></a>00272 }
<a name="l00273"></a>00273 
<a name="l00274"></a>00274 =head2 name
<a name="l00275"></a>00275   Arg        : (optional) name to <span class="keyword">set</span>
<a name="l00276"></a>00276   Description: Gets/sets name <span class="keywordflow">for</span> genome
<a name="l00277"></a>00277   Returntype : <span class="keywordtype">string</span>
<a name="l00278"></a>00278   Exceptions : none
<a name="l00279"></a>00279   Caller     : general
<a name="l00280"></a>00280   Status     : Stable
<a name="l00281"></a>00281 =cut
<a name="l00282"></a>00282 
<a name="l00283"></a>00283 sub name {
<a name="l00284"></a>00284   my ( $self, $arg ) = @_;
<a name="l00285"></a>00285   <span class="keywordflow">return</span> $self-&gt;organism()-&gt;name($arg);
<a name="l00286"></a>00286 }
<a name="l00287"></a>00287 
<a name="l00288"></a>00288 =head2 taxonomy_id
<a name="l00289"></a>00289   Arg        : (optional) taxonomy_id to <span class="keyword">set</span>
<a name="l00290"></a>00290   Description: Gets/sets NCBI taxonomy ID
<a name="l00291"></a>00291   Returntype : <span class="keywordtype">string</span>
<a name="l00292"></a>00292   Exceptions : none
<a name="l00293"></a>00293   Caller     : general
<a name="l00294"></a>00294   Status     : Stable
<a name="l00295"></a>00295 =cut
<a name="l00296"></a>00296 
<a name="l00297"></a>00297 sub taxonomy_id {
<a name="l00298"></a>00298   my ( $self, $taxonomy_id ) = @_;
<a name="l00299"></a>00299   <span class="keywordflow">return</span> $self-&gt;organism()-&gt;taxonomy_id($taxonomy_id);
<a name="l00300"></a>00300 }
<a name="l00301"></a>00301 
<a name="l00302"></a>00302 =head2 species_taxonomy_id
<a name="l00303"></a>00303   Arg        : (optional) taxonomy_id to <span class="keyword">set</span>
<a name="l00304"></a>00304   Description: Gets/sets NCBI taxonomy ID of species to which <span class="keyword">this</span> belongs
<a name="l00305"></a>00305   Returntype : <span class="keywordtype">string</span>
<a name="l00306"></a>00306   Exceptions : none
<a name="l00307"></a>00307   Caller     : general
<a name="l00308"></a>00308   Status     : Stable
<a name="l00309"></a>00309 =cut
<a name="l00310"></a>00310 
<a name="l00311"></a>00311 sub species_taxonomy_id {
<a name="l00312"></a>00312   my ( $self, $taxonomy_id ) = @_;
<a name="l00313"></a>00313   <span class="keywordflow">return</span> $self-&gt;organism()-&gt;species_taxonomy_id($taxonomy_id);
<a name="l00314"></a>00314 }
<a name="l00315"></a>00315 
<a name="l00316"></a>00316 =head2 assembly
<a name="l00317"></a>00317   Arg        : (optional) assembly to <span class="keyword">set</span>
<a name="l00318"></a>00318   Description: Gets/sets assembly <span class="keywordtype">object</span>
<a name="l00319"></a>00319   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeAssemblyInfo.html">Bio::EnsEMBL::MetaData::GenomeAssemblyInfo</a>
<a name="l00320"></a>00320   Exceptions : none
<a name="l00321"></a>00321   Caller     : general
<a name="l00322"></a>00322   Status     : Stable
<a name="l00323"></a>00323 =cut
<a name="l00324"></a>00324 
<a name="l00325"></a>00325 sub assembly {
<a name="l00326"></a>00326   my ( $self, $assembly ) = @_;
<a name="l00327"></a>00327   $self-&gt;{assembly} = $assembly <span class="keywordflow">if</span> ( defined $assembly );
<a name="l00328"></a>00328   $self-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1BaseInfo.html#a1a34c3fa71d4be6076403fee9cbda557">_load_child</a>( <span class="stringliteral">&#39;assembly&#39;</span>, <span class="stringliteral">&#39;_fetch_assembly&#39;</span> );
<a name="l00329"></a>00329   <span class="keywordflow">return</span> $self-&gt;{assembly};
<a name="l00330"></a>00330 }
<a name="l00331"></a>00331 
<a name="l00332"></a>00332 =head2 assembly_name
<a name="l00333"></a>00333   Arg        : (optional) assembly_name to <span class="keyword">set</span>
<a name="l00334"></a>00334   Description: Gets/sets name of assembly
<a name="l00335"></a>00335   Returntype : <span class="keywordtype">string</span>
<a name="l00336"></a>00336   Exceptions : none
<a name="l00337"></a>00337   Caller     : general
<a name="l00338"></a>00338   Status     : Stable
<a name="l00339"></a>00339 =cut
<a name="l00340"></a>00340 
<a name="l00341"></a>00341 sub assembly_name {
<a name="l00342"></a>00342   my ( $self, $assembly_name ) = @_;
<a name="l00343"></a>00343   <span class="keywordflow">return</span> $self-&gt;assembly()-&gt;assembly_name($assembly_name);
<a name="l00344"></a>00344 }
<a name="l00345"></a>00345 
<a name="l00346"></a>00346 =head2 assembly_accession
<a name="l00347"></a>00347   Arg        : (optional) assembly_id to <span class="keyword">set</span>
<a name="l00348"></a>00348   Description: Gets/sets INSDC accession <span class="keywordflow">for</span> assembly
<a name="l00349"></a>00349   Returntype : <span class="keywordtype">string</span>
<a name="l00350"></a>00350   Exceptions : none
<a name="l00351"></a>00351   Caller     : general
<a name="l00352"></a>00352   Status     : Stable
<a name="l00353"></a>00353 =cut
<a name="l00354"></a>00354 
<a name="l00355"></a>00355 sub assembly_accession {
<a name="l00356"></a>00356   my ( $self, $assembly_id ) = @_;
<a name="l00357"></a>00357   <span class="keywordflow">return</span> $self-&gt;assembly()-&gt;assembly_accession($assembly_id);
<a name="l00358"></a>00358 }
<a name="l00359"></a>00359 
<a name="l00360"></a>00360 =head2 assembly_level
<a name="l00361"></a>00361   Arg        : (optional) assembly_level to <span class="keyword">set</span>
<a name="l00362"></a>00362   Description: Gets/sets highest level of assembly (chromosome, supercontig etc.)
<a name="l00363"></a>00363   Returntype : <span class="keywordtype">string</span>
<a name="l00364"></a>00364   Exceptions : none
<a name="l00365"></a>00365   Caller     : general
<a name="l00366"></a>00366   Status     : Stable
<a name="l00367"></a>00367 =cut
<a name="l00368"></a>00368 
<a name="l00369"></a>00369 sub assembly_level {
<a name="l00370"></a>00370   my ( $self, $assembly_level ) = @_;
<a name="l00371"></a>00371   <span class="keywordflow">return</span> $self-&gt;assembly()-&gt;assembly_level($assembly_level);
<a name="l00372"></a>00372 }
<a name="l00373"></a>00373 
<a name="l00374"></a>00374 =head2 genebuild
<a name="l00375"></a>00375   Arg        : (optional) genebuild to <span class="keyword">set</span>
<a name="l00376"></a>00376   Description: Gets/sets identifier <span class="keywordflow">for</span> genebuild
<a name="l00377"></a>00377   Returntype : <span class="keywordtype">string</span>
<a name="l00378"></a>00378   Exceptions : none
<a name="l00379"></a>00379   Caller     : general
<a name="l00380"></a>00380   Status     : Stable
<a name="l00381"></a>00381 =cut
<a name="l00382"></a>00382 
<a name="l00383"></a>00383 sub genebuild {
<a name="l00384"></a>00384   my ( $self, $genebuild ) = @_;
<a name="l00385"></a>00385   $self-&gt;{genebuild} = $genebuild <span class="keywordflow">if</span> ( defined $genebuild );
<a name="l00386"></a>00386   <span class="keywordflow">return</span> $self-&gt;{genebuild};
<a name="l00387"></a>00387 }
<a name="l00388"></a>00388 
<a name="l00389"></a>00389 =head2 division
<a name="l00390"></a>00390   Arg        : (optional) division to <span class="keyword">set</span>
<a name="l00391"></a>00391   Description: Gets/sets Ensembl Genomes division
<a name="l00392"></a>00392   Returntype : <span class="keywordtype">string</span>
<a name="l00393"></a>00393   Exceptions : none
<a name="l00394"></a>00394   Caller     : general
<a name="l00395"></a>00395   Status     : Stable
<a name="l00396"></a>00396 =cut
<a name="l00397"></a>00397 
<a name="l00398"></a>00398 sub division {
<a name="l00399"></a>00399   my ( $self, $division ) = @_;
<a name="l00400"></a>00400   $self-&gt;{division} = $division <span class="keywordflow">if</span> ( defined $division );
<a name="l00401"></a>00401   <span class="keywordflow">return</span> $self-&gt;{division};
<a name="l00402"></a>00402 }
<a name="l00403"></a>00403 
<a name="l00404"></a>00404 =head2 is_reference
<a name="l00405"></a>00405   Arg        : (optional) value of is_reference
<a name="l00406"></a>00406   Description: Gets/sets whether <span class="keyword">this</span> is a reference <span class="keywordflow">for</span> the species
<a name="l00407"></a>00407   Returntype : <span class="keywordtype">bool</span>
<a name="l00408"></a>00408   Exceptions : none
<a name="l00409"></a>00409   Caller     : general
<a name="l00410"></a>00410   Status     : Stable
<a name="l00411"></a>00411 =cut
<a name="l00412"></a>00412 
<a name="l00413"></a>00413 sub is_reference {
<a name="l00414"></a>00414   my ( $self, $is_ref ) = @_;
<a name="l00415"></a>00415   <span class="keywordflow">return</span> $self-&gt;organism()-&gt;is_reference($is_ref);
<a name="l00416"></a>00416 }
<a name="l00417"></a>00417 
<a name="l00418"></a>00418 =head2 db_size
<a name="l00419"></a>00419   Arg        : (optional) db_size to <span class="keyword">set</span>
<a name="l00420"></a>00420   Description: Gets/sets size of database containing core
<a name="l00421"></a>00421   Returntype : <span class="keywordtype">string</span>
<a name="l00422"></a>00422   Exceptions : none
<a name="l00423"></a>00423   Caller     : general
<a name="l00424"></a>00424   Status     : Stable
<a name="l00425"></a>00425 =cut
<a name="l00426"></a>00426 
<a name="l00427"></a>00427 sub db_size {
<a name="l00428"></a>00428   my ( $self, $arg ) = @_;
<a name="l00429"></a>00429   $self-&gt;{db_size} = $arg <span class="keywordflow">if</span> ( defined $arg );
<a name="l00430"></a>00430   <span class="keywordflow">return</span> $self-&gt;{db_size};
<a name="l00431"></a>00431 }
<a name="l00432"></a>00432 
<a name="l00433"></a>00433 =head2 base_count
<a name="l00434"></a>00434   Arg        : (optional) base_count to <span class="keyword">set</span>
<a name="l00435"></a>00435   Description: Gets/sets total number of bases in assembled genome
<a name="l00436"></a>00436   Returntype : integer
<a name="l00437"></a>00437   Exceptions : none
<a name="l00438"></a>00438   Caller     : general
<a name="l00439"></a>00439   Status     : Stable
<a name="l00440"></a>00440 =cut
<a name="l00441"></a>00441 
<a name="l00442"></a>00442 sub base_count {
<a name="l00443"></a>00443   my ( $self, $base_count ) = @_;
<a name="l00444"></a>00444   <span class="keywordflow">return</span> $self-&gt;assembly()-&gt;base_count($base_count);
<a name="l00445"></a>00445 }
<a name="l00446"></a>00446 
<a name="l00447"></a>00447 =head2 aliases
<a name="l00448"></a>00448   Arg        : (optional) arrayref of aliases to <span class="keyword">set</span>
<a name="l00449"></a>00449   Description: Gets/sets aliases by which the genome is also known 
<a name="l00450"></a>00450   Returntype : Arrayref of aliases
<a name="l00451"></a>00451   Exceptions : none
<a name="l00452"></a>00452   Caller     : general
<a name="l00453"></a>00453   Status     : Stable
<a name="l00454"></a>00454 =cut
<a name="l00455"></a>00455 
<a name="l00456"></a>00456 sub aliases {
<a name="l00457"></a>00457   my ( $self, $aliases ) = @_;
<a name="l00458"></a>00458   <span class="keywordflow">return</span> $self-&gt;organism()-&gt;aliases($aliases);
<a name="l00459"></a>00459 }
<a name="l00460"></a>00460 
<a name="l00461"></a>00461 =head2 compara
<a name="l00462"></a>00462   Arg        : (optional) arrayref of GenomeComparaInfo objects to <span class="keyword">set</span>
<a name="l00463"></a>00463   Description: Gets/sets GenomeComparaInfo describing comparative analyses applied to the genome
<a name="l00464"></a>00464   Returntype : Arrayref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a>
<a name="l00465"></a>00465   Exceptions : none
<a name="l00466"></a>00466   Caller     : general
<a name="l00467"></a>00467   Status     : Stable
<a name="l00468"></a>00468 =cut
<a name="l00469"></a>00469 
<a name="l00470"></a>00470 sub compara {
<a name="l00471"></a>00471   my ( $self, $compara ) = @_;
<a name="l00472"></a>00472   <span class="keywordflow">if</span> ( defined $compara ) {
<a name="l00473"></a>00473     $self-&gt;{compara}               = $compara;
<a name="l00474"></a>00474     $self-&gt;{has_peptide_compara}   = undef;
<a name="l00475"></a>00475     $self-&gt;{has_synteny}           = undef;
<a name="l00476"></a>00476     $self-&gt;{has_genome_alignments} = undef;
<a name="l00477"></a>00477     $self-&gt;{has_pan_compara}       = undef;
<a name="l00478"></a>00478   }
<a name="l00479"></a>00479   $self-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1BaseInfo.html#a1a34c3fa71d4be6076403fee9cbda557">_load_child</a>( <span class="stringliteral">&#39;compara&#39;</span>, <span class="stringliteral">&#39;_fetch_comparas&#39;</span> );
<a name="l00480"></a>00480   <span class="keywordflow">return</span> $self-&gt;{compara};
<a name="l00481"></a>00481 }
<a name="l00482"></a>00482 
<a name="l00483"></a>00483 =head2 sequences
<a name="l00484"></a>00484   Arg        : (optional) arrayref of sequences to <span class="keyword">set</span>
<a name="l00485"></a>00485   Description: Gets/sets array of hashrefs describing sequences from the assembly. Elements are hashrefs with name and acc as keys
<a name="l00486"></a>00486   Returntype : Arrayref
<a name="l00487"></a>00487   Exceptions : none
<a name="l00488"></a>00488   Caller     : general
<a name="l00489"></a>00489   Status     : Stable
<a name="l00490"></a>00490 =cut
<a name="l00491"></a>00491 
<a name="l00492"></a>00492 sub sequences {
<a name="l00493"></a>00493   my ( $self, $sequences ) = @_;
<a name="l00494"></a>00494   <span class="keywordflow">return</span> $self-&gt;assembly()-&gt;sequences($sequences);
<a name="l00495"></a>00495 }
<a name="l00496"></a>00496 
<a name="l00497"></a>00497 =head2 publications
<a name="l00498"></a>00498   Arg        : (optional) arrayref of pubmed IDs to <span class="keyword">set</span>
<a name="l00499"></a>00499   Description: Gets/sets PubMed IDs <span class="keywordflow">for</span> publications associated with the genome
<a name="l00500"></a>00500   Returntype : Arrayref of PubMed IDs
<a name="l00501"></a>00501   Exceptions : none
<a name="l00502"></a>00502   Caller     : general
<a name="l00503"></a>00503   Status     : Stable
<a name="l00504"></a>00504 =cut
<a name="l00505"></a>00505 
<a name="l00506"></a>00506 sub publications {
<a name="l00507"></a>00507   my ( $self, $publications ) = @_;
<a name="l00508"></a>00508   <span class="keywordflow">return</span> $self-&gt;organism()-&gt;publications($publications);
<a name="l00509"></a>00509 }
<a name="l00510"></a>00510 
<a name="l00511"></a>00511 =head2 variations
<a name="l00512"></a>00512   Arg        : (optional) variations to <span class="keyword">set</span>
<a name="l00513"></a>00513   Description: Gets/sets variations associated with genomes as hashref 
<a name="l00514"></a>00514                (variations,structural variations,genotypes,phenotypes), 
<a name="l00515"></a>00515                further broken down into counts by type/source
<a name="l00516"></a>00516   Returntype : Arrayref
<a name="l00517"></a>00517   Exceptions : none
<a name="l00518"></a>00518   Caller     : general
<a name="l00519"></a>00519   Status     : Stable
<a name="l00520"></a>00520 =cut
<a name="l00521"></a>00521 
<a name="l00522"></a>00522 sub variations {
<a name="l00523"></a>00523   my ( $self, $variations ) = @_;
<a name="l00524"></a>00524   <span class="keywordflow">if</span> ( defined $variations ) {
<a name="l00525"></a>00525     $self-&gt;{variations}     = $variations;
<a name="l00526"></a>00526     $self-&gt;{has_variations} = undef;
<a name="l00527"></a>00527   }
<a name="l00528"></a>00528   $self-&gt;_load_child( <span class="stringliteral">&#39;variations&#39;</span>, <span class="stringliteral">&#39;_fetch_variations&#39;</span> );
<a name="l00529"></a>00529   <span class="keywordflow">return</span> $self-&gt;{variations};
<a name="l00530"></a>00530 }
<a name="l00531"></a>00531 
<a name="l00532"></a>00532 =head2 features
<a name="l00533"></a>00533   Arg        : (optional) features to <span class="keyword">set</span>
<a name="l00534"></a>00534   Description: Gets/sets general genomic features associated with the genome as hashref
<a name="l00535"></a>00535                keyed by type (e.g. repeatFeatures,simpleFeatures), further broken down into
<a name="l00536"></a>00536                counts by analysis
<a name="l00537"></a>00537   Returntype : Hashref
<a name="l00538"></a>00538   Exceptions : none
<a name="l00539"></a>00539   Caller     : general
<a name="l00540"></a>00540   Status     : Stable
<a name="l00541"></a>00541 =cut
<a name="l00542"></a>00542 
<a name="l00543"></a>00543 sub features {
<a name="l00544"></a>00544   my ( $self, $features ) = @_;
<a name="l00545"></a>00545   <span class="keywordflow">if</span> ( defined $features ) {
<a name="l00546"></a>00546     $self-&gt;{features} = $features;
<a name="l00547"></a>00547   }
<a name="l00548"></a>00548   $self-&gt;_load_child( <span class="stringliteral">&#39;features&#39;</span>, <span class="stringliteral">&#39;_fetch_features&#39;</span> );
<a name="l00549"></a>00549   <span class="keywordflow">return</span> $self-&gt;{features};
<a name="l00550"></a>00550 }
<a name="l00551"></a>00551 
<a name="l00552"></a>00552 =head2 annotations
<a name="l00553"></a>00553   Arg        : (optional) annotations to <span class="keyword">set</span>
<a name="l00554"></a>00554   Description: Gets/sets summary information about gene annotation as hashref, with
<a name="l00555"></a>00555                annotation type as key and count as value
<a name="l00556"></a>00556   Returntype : hashref
<a name="l00557"></a>00557   Exceptions : none
<a name="l00558"></a>00558   Caller     : general
<a name="l00559"></a>00559   Status     : Stable
<a name="l00560"></a>00560 =cut
<a name="l00561"></a>00561 
<a name="l00562"></a>00562 sub annotations {
<a name="l00563"></a>00563   my ( $self, $annotation ) = @_;
<a name="l00564"></a>00564   <span class="keywordflow">if</span> ( defined $annotation ) {
<a name="l00565"></a>00565     $self-&gt;{annotations} = $annotation;
<a name="l00566"></a>00566   }
<a name="l00567"></a>00567   $self-&gt;_load_child( <span class="stringliteral">&#39;annotations&#39;</span>, <span class="stringliteral">&#39;_fetch_annotations&#39;</span> );
<a name="l00568"></a>00568   <span class="keywordflow">return</span> $self-&gt;{annotations};
<a name="l00569"></a>00569 }
<a name="l00570"></a>00570 
<a name="l00571"></a>00571 =head2 other_alignments
<a name="l00572"></a>00572   Arg        : (optional) other alignments to <span class="keyword">set</span>
<a name="l00573"></a>00573   Description: Gets/sets other alignments as hashref, keyed by type (dnaAlignFeatures,proteinAlignFeatures)
<a name="l00574"></a>00574                with values as logic_name-count pairs 
<a name="l00575"></a>00575   Returntype : Hashref
<a name="l00576"></a>00576   Exceptions : none
<a name="l00577"></a>00577   Caller     : general
<a name="l00578"></a>00578   Status     : Stable
<a name="l00579"></a>00579 =cut
<a name="l00580"></a>00580 
<a name="l00581"></a>00581 sub other_alignments {
<a name="l00582"></a>00582   my ( $self, $other_alignments ) = @_;
<a name="l00583"></a>00583   <span class="keywordflow">if</span> ( defined $other_alignments ) {
<a name="l00584"></a>00584     $self-&gt;{other_alignments}     = $other_alignments;
<a name="l00585"></a>00585     $self-&gt;{has_other_alignments} = undef;
<a name="l00586"></a>00586   }
<a name="l00587"></a>00587   $self-&gt;_load_child( <span class="stringliteral">&#39;other_alignments&#39;</span>, <span class="stringliteral">&#39;_fetch_other_alignments&#39;</span> );
<a name="l00588"></a>00588   <span class="keywordflow">return</span> $self-&gt;{other_alignments} || 0;
<a name="l00589"></a>00589 }
<a name="l00590"></a>00590 
<a name="l00591"></a>00591 =head1 utility methods
<a name="l00592"></a>00592 =head2 has_variations
<a name="l00593"></a>00593   Arg        : (optional) 1/0 to <span class="keyword">set</span> <span class="keywordflow">if</span> genome has variation
<a name="l00594"></a>00594   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> genome has variation, 0 <span class="keywordflow">if</span> not
<a name="l00595"></a>00595   Returntype : 1 or 0
<a name="l00596"></a>00596   Exceptions : none
<a name="l00597"></a>00597   Caller     : general
<a name="l00598"></a>00598   Status     : Stable
<a name="l00599"></a>00599 =cut
<a name="l00600"></a>00600 
<a name="l00601"></a>00601 sub has_variations {
<a name="l00602"></a>00602   my ( $self, $arg ) = @_;
<a name="l00603"></a>00603   <span class="keywordflow">if</span> ( defined $arg ) {
<a name="l00604"></a>00604     $self-&gt;{has_variations} = $arg;
<a name="l00605"></a>00605   }
<a name="l00606"></a>00606   elsif ( !defined( $self-&gt;{has_variations} ) ) {
<a name="l00607"></a>00607     $self-&gt;{has_variations} = $self-&gt;count_variation() &gt; 0 ? 1 : 0;
<a name="l00608"></a>00608   }
<a name="l00609"></a>00609   <span class="keywordflow">return</span> $self-&gt;{has_variations};
<a name="l00610"></a>00610 }
<a name="l00611"></a>00611 
<a name="l00612"></a>00612 =head2 has_genome_alignments
<a name="l00613"></a>00613   Arg        : (optional) 1/0 to <span class="keyword">set</span> <span class="keywordflow">if</span> genome has genome alignments
<a name="l00614"></a>00614   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> genome has genome alignments, 0 <span class="keywordflow">if</span> not
<a name="l00615"></a>00615   Returntype : 1 or 0
<a name="l00616"></a>00616   Exceptions : none
<a name="l00617"></a>00617   Caller     : general
<a name="l00618"></a>00618   Status     : Stable
<a name="l00619"></a>00619 =cut
<a name="l00620"></a>00620 
<a name="l00621"></a>00621 sub has_genome_alignments {
<a name="l00622"></a>00622   my ( $self, $arg ) = @_;
<a name="l00623"></a>00623   <span class="keywordflow">if</span> ( defined $arg ) {
<a name="l00624"></a>00624     $self-&gt;{has_genome_alignments} = $arg;
<a name="l00625"></a>00625   }
<a name="l00626"></a>00626   elsif ( !defined( $self-&gt;{has_genome_alignments} ) &amp;&amp;
<a name="l00627"></a>00627           defined $self-&gt;compara() )
<a name="l00628"></a>00628   {
<a name="l00629"></a>00629     $self-&gt;{has_genome_alignments} = 0;
<a name="l00630"></a>00630     <span class="keywordflow">for</span> my $compara ( @{ $self-&gt;compara() } ) {
<a name="l00631"></a>00631       <span class="keywordflow">if</span> ( $compara-&gt;is_dna_compara() ) {
<a name="l00632"></a>00632         $self-&gt;{has_genome_alignments} = 1;
<a name="l00633"></a>00633         last;
<a name="l00634"></a>00634       }
<a name="l00635"></a>00635     }
<a name="l00636"></a>00636   }
<a name="l00637"></a>00637   <span class="keywordflow">return</span> $self-&gt;{has_genome_alignments} || 0;
<a name="l00638"></a>00638 }
<a name="l00639"></a>00639 
<a name="l00640"></a>00640 =head2 has_synteny
<a name="l00641"></a>00641   Arg        : (optional) 1/0 to <span class="keyword">set</span> <span class="keywordflow">if</span> genome has synteny
<a name="l00642"></a>00642   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> genome has synteny, 0 <span class="keywordflow">if</span> not
<a name="l00643"></a>00643   Returntype : 1 or 0
<a name="l00644"></a>00644   Exceptions : none
<a name="l00645"></a>00645   Caller     : general
<a name="l00646"></a>00646   Status     : Stable
<a name="l00647"></a>00647 =cut
<a name="l00648"></a>00648 
<a name="l00649"></a>00649 sub has_synteny {
<a name="l00650"></a>00650   my ( $self, $arg ) = @_;
<a name="l00651"></a>00651   <span class="keywordflow">if</span> ( defined $arg ) {
<a name="l00652"></a>00652     $self-&gt;{has_synteny} = $arg;
<a name="l00653"></a>00653   }
<a name="l00654"></a>00654   elsif ( !defined( $self-&gt;{has_synteny} ) &amp;&amp; defined $self-&gt;compara() ) {
<a name="l00655"></a>00655     $self-&gt;{has_synteny} = 0;
<a name="l00656"></a>00656     <span class="keywordflow">for</span> my $compara ( @{ $self-&gt;compara() } ) {
<a name="l00657"></a>00657       <span class="keywordflow">if</span> ( $compara-&gt;is_synteny() ) {
<a name="l00658"></a>00658         $self-&gt;{has_synteny} = 1;
<a name="l00659"></a>00659         last;
<a name="l00660"></a>00660       }
<a name="l00661"></a>00661     }
<a name="l00662"></a>00662   }
<a name="l00663"></a>00663   <span class="keywordflow">return</span> $self-&gt;{has_synteny} || 0;
<a name="l00664"></a>00664 }
<a name="l00665"></a>00665 
<a name="l00666"></a>00666 =head2 has_peptide_compara
<a name="l00667"></a>00667   Arg        : (optional) 1/0 to <span class="keyword">set</span> <span class="keywordflow">if</span> genome has peptide compara
<a name="l00668"></a>00668   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> genome has peptide, 0 <span class="keywordflow">if</span> not
<a name="l00669"></a>00669   Returntype : 1 or 0
<a name="l00670"></a>00670   Exceptions : none
<a name="l00671"></a>00671   Caller     : general
<a name="l00672"></a>00672   Status     : Stable
<a name="l00673"></a>00673 =cut
<a name="l00674"></a>00674 
<a name="l00675"></a>00675 sub has_peptide_compara {
<a name="l00676"></a>00676   my ( $self, $arg ) = @_;
<a name="l00677"></a>00677   <span class="keywordflow">if</span> ( defined $arg ) {
<a name="l00678"></a>00678     $self-&gt;{has_peptide_compara} = $arg;
<a name="l00679"></a>00679   }
<a name="l00680"></a>00680   elsif ( !defined( $self-&gt;{has_peptide_compara} ) &amp;&amp; defined $self-&gt;compara() )
<a name="l00681"></a>00681   {
<a name="l00682"></a>00682     $self-&gt;{has_peptide_compara} = 0;
<a name="l00683"></a>00683     <span class="keywordflow">for</span> my $compara ( @{ $self-&gt;compara() } ) {
<a name="l00684"></a>00684       <span class="keywordflow">if</span> ( $compara-&gt;is_peptide_compara() &amp;&amp; !$compara-&gt;is_pan_compara() ) {
<a name="l00685"></a>00685         $self-&gt;{has_peptide_compara} = 1;
<a name="l00686"></a>00686         last;
<a name="l00687"></a>00687       }
<a name="l00688"></a>00688     }
<a name="l00689"></a>00689   }
<a name="l00690"></a>00690   <span class="keywordflow">return</span> $self-&gt;{has_peptide_compara} || 0;
<a name="l00691"></a>00691 }
<a name="l00692"></a>00692 
<a name="l00693"></a>00693 =head2 has_pan_compara
<a name="l00694"></a>00694   Arg        : (optional) 1/0 to <span class="keyword">set</span> <span class="keywordflow">if</span> genome is included in pan compara
<a name="l00695"></a>00695   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> genome is in pan compara, 0 <span class="keywordflow">if</span> not
<a name="l00696"></a>00696   Returntype : 1 or 0
<a name="l00697"></a>00697   Exceptions : none
<a name="l00698"></a>00698   Caller     : general
<a name="l00699"></a>00699   Status     : Stable
<a name="l00700"></a>00700 =cut
<a name="l00701"></a>00701 
<a name="l00702"></a>00702 sub has_pan_compara {
<a name="l00703"></a>00703   my ( $self, $arg ) = @_;
<a name="l00704"></a>00704   <span class="keywordflow">if</span> ( defined $arg ) {
<a name="l00705"></a>00705     $self-&gt;{has_pan_compara} = $arg;
<a name="l00706"></a>00706   }
<a name="l00707"></a>00707   elsif ( !defined( $self-&gt;{has_pan_compara} ) &amp;&amp; defined $self-&gt;compara() ) {
<a name="l00708"></a>00708     $self-&gt;{has_pan_compara} = 0;
<a name="l00709"></a>00709     <span class="keywordflow">for</span> my $compara ( @{ $self-&gt;compara() } ) {
<a name="l00710"></a>00710       <span class="keywordflow">if</span> ( $compara-&gt;is_pan_compara() ) {
<a name="l00711"></a>00711         $self-&gt;{has_pan_compara} = 1;
<a name="l00712"></a>00712         last;
<a name="l00713"></a>00713       }
<a name="l00714"></a>00714     }
<a name="l00715"></a>00715   }
<a name="l00716"></a>00716   <span class="keywordflow">return</span> $self-&gt;{has_pan_compara} || 0;
<a name="l00717"></a>00717 }
<a name="l00718"></a>00718 
<a name="l00719"></a>00719 =head2 has_other_alignments
<a name="l00720"></a>00720   Arg        : (optional) 1/0 to <span class="keyword">set</span> <span class="keywordflow">if</span> genome has other alignments
<a name="l00721"></a>00721   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> genome has other alignments, 0 <span class="keywordflow">if</span> not
<a name="l00722"></a>00722   Returntype : 1 or 0
<a name="l00723"></a>00723   Exceptions : none
<a name="l00724"></a>00724   Caller     : general
<a name="l00725"></a>00725   Status     : Stable
<a name="l00726"></a>00726 =cut
<a name="l00727"></a>00727 
<a name="l00728"></a>00728 sub has_other_alignments {
<a name="l00729"></a>00729   my ( $self, $arg ) = @_;
<a name="l00730"></a>00730   <span class="keywordflow">if</span> ( defined $arg ) {
<a name="l00731"></a>00731     $self-&gt;{has_other_alignments} = $arg;
<a name="l00732"></a>00732   }
<a name="l00733"></a>00733   elsif ( !defined( $self-&gt;{has_other_alignments} ) ) {
<a name="l00734"></a>00734     $self-&gt;{has_other_alignments} = $self-&gt;count_alignments() &gt; 0 ? 1 : 0;
<a name="l00735"></a>00735   }
<a name="l00736"></a>00736   <span class="keywordflow">return</span> $self-&gt;{has_other_alignments} || 0;
<a name="l00737"></a>00737 }
<a name="l00738"></a>00738 
<a name="l00739"></a>00739 =head2 count_variation
<a name="l00740"></a>00740   Description: Returns total number of variations and structural variations mapped to genome
<a name="l00741"></a>00741   Returntype : integer
<a name="l00742"></a>00742   Exceptions : none
<a name="l00743"></a>00743   Caller     : general
<a name="l00744"></a>00744   Status     : Stable
<a name="l00745"></a>00745 =cut
<a name="l00746"></a>00746 
<a name="l00747"></a>00747 sub count_variation {
<a name="l00748"></a>00748   my ($self) = @_;
<a name="l00749"></a>00749   <span class="keywordflow">return</span> $self-&gt;count_hash_values( $self-&gt;{variations}{variations} ) +
<a name="l00750"></a>00750     $self-&gt;count_hash_values( $self-&gt;{variations}{structural_variations} );
<a name="l00751"></a>00751 }
<a name="l00752"></a>00752 
<a name="l00753"></a>00753 =head2 count_alignments
<a name="l00754"></a>00754   Description: Returns total number of alignments to genome
<a name="l00755"></a>00755   Returntype : integer
<a name="l00756"></a>00756   Exceptions : none
<a name="l00757"></a>00757   Caller     : general
<a name="l00758"></a>00758   Status     : Stable
<a name="l00759"></a>00759 =cut
<a name="l00760"></a>00760 
<a name="l00761"></a>00761 sub count_alignments {
<a name="l00762"></a>00762   my ($self) = @_;
<a name="l00763"></a>00763   <span class="keywordflow">return</span> $self-&gt;count_hash_values( $self-&gt;{other_alignments}{bam} ) +
<a name="l00764"></a>00764     $self-&gt;count_hash_values( $self-&gt;{other_alignments}{proteinAlignFeatures} )
<a name="l00765"></a>00765     + $self-&gt;count_hash_values( $self-&gt;{other_alignments}{dnaAlignFeatures} );
<a name="l00766"></a>00766 }
<a name="l00767"></a>00767 
<a name="l00768"></a>00768 =head2 get_uniprot_coverage
<a name="l00769"></a>00769   Description: Get % of protein coding genes with a UniProt cross-reference
<a name="l00770"></a>00770   Returntype : uniprot coverage as percentage
<a name="l00771"></a>00771   Exceptions : none
<a name="l00772"></a>00772   Caller     : general
<a name="l00773"></a>00773   Status     : Stable
<a name="l00774"></a>00774 =cut
<a name="l00775"></a>00775 
<a name="l00776"></a>00776 sub get_uniprot_coverage {
<a name="l00777"></a>00777   my ($self) = @_;
<a name="l00778"></a>00778   <span class="keywordflow">return</span> 100.0*
<a name="l00779"></a>00779     ( $self-&gt;annotations()-&gt;{nProteinCodingUniProtKB} )/
<a name="l00780"></a>00780     $self-&gt;annotations()-&gt;{nProteinCoding};
<a name="l00781"></a>00781 }
<a name="l00782"></a>00782 
<a name="l00783"></a>00783 =head2 to_string
<a name="l00784"></a>00784   Description: Render genome as <span class="keywordtype">string</span> <span class="keywordflow">for</span> display
<a name="l00785"></a>00785   Returntype : String
<a name="l00786"></a>00786   Exceptions : none
<a name="l00787"></a>00787   Caller     : general
<a name="l00788"></a>00788   Status     : Stable
<a name="l00789"></a>00789 =cut
<a name="l00790"></a>00790 
<a name="l00791"></a>00791 sub to_string {
<a name="l00792"></a>00792    my ($self) = @_;
<a name="l00793"></a>00793    <span class="keywordflow">return</span> join(<span class="charliteral">&#39;:&#39;</span>,($self-&gt;dbID()||<span class="charliteral">&#39;-&#39;</span>),$self-&gt;name(),$self-&gt;dbname(),$self-&gt;species_id());
<a name="l00794"></a>00794 }
<a name="l00795"></a>00795 
<a name="l00796"></a>00796 =head2 to_hash
<a name="l00797"></a>00797   Description: Render genome as plain hash suitable <span class="keywordflow">for</span> export as JSON/XML
<a name="l00798"></a>00798   Argument   : (optional) <span class="keywordflow">if</span> <span class="keyword">set</span> to 1, force expansion of children
<a name="l00799"></a>00799   Returntype : Hashref
<a name="l00800"></a>00800   Exceptions : none
<a name="l00801"></a>00801   Caller     : general
<a name="l00802"></a>00802   Status     : Stable
<a name="l00803"></a>00803 =cut
<a name="l00804"></a>00804 
<a name="l00805"></a>00805 sub to_hash {
<a name="l00806"></a>00806   my ( $in, $keen ) = @_;
<a name="l00807"></a>00807   my $out;
<a name="l00808"></a>00808   my $type = ref $in;
<a name="l00809"></a>00809   <span class="keywordflow">if</span> ( defined $keen &amp;&amp;
<a name="l00810"></a>00810        $keen == 1 &amp;&amp;
<a name="l00811"></a>00811        $type =~ m/Bio::EnsEMBL::MetaData::[A-z]+Info/ )
<a name="l00812"></a>00812   {
<a name="l00813"></a>00813     $in-&gt;_preload();
<a name="l00814"></a>00814   }
<a name="l00815"></a>00815   <span class="keywordflow">if</span> ( $type eq <span class="stringliteral">&#39;ARRAY&#39;</span> ) {
<a name="l00816"></a>00816     $out = [];
<a name="l00817"></a>00817     <span class="keywordflow">for</span> my $item ( @{$in} ) {
<a name="l00818"></a>00818       push @{$out}, to_hash( $item, $keen );
<a name="l00819"></a>00819     }
<a name="l00820"></a>00820   }
<a name="l00821"></a>00821   elsif ( $type eq <span class="stringliteral">&#39;HASH&#39;</span> || $type eq <span class="stringliteral">&#39;Bio::EnsEMBL::MetaData::GenomeInfo&#39;</span> ) {
<a name="l00822"></a>00822     $out = {};
<a name="l00823"></a>00823     <span class="keywordflow">while</span> ( my ( $key, $val ) = each %$in ) {
<a name="l00824"></a>00824       <span class="keywordflow">if</span> ( $key ne <span class="stringliteral">&#39;dbID&#39;</span> &amp;&amp; $key ne <span class="stringliteral">&#39;adaptor&#39;</span> &amp;&amp; $key ne <span class="stringliteral">&#39;logger&#39;</span> ) {
<a name="l00825"></a>00825 
<a name="l00826"></a>00826 <span class="preprocessor"># deal with keys starting with numbers, which are not valid element names in XML</span>
<a name="l00827"></a>00827 <span class="preprocessor"></span>        <span class="keywordflow">if</span> ( $key =~ m/^[0-9].*/ ) {
<a name="l00828"></a>00828           $key = <span class="charliteral">&#39;_&#39;</span> . $key;
<a name="l00829"></a>00829         }
<a name="l00830"></a>00830         $out-&gt;{$key} = to_hash( $val, $keen );
<a name="l00831"></a>00831       }
<a name="l00832"></a>00832 
<a name="l00833"></a>00833     }
<a name="l00834"></a>00834   }
<a name="l00835"></a>00835   elsif ( $type =~ m/Bio::EnsEMBL::MetaData::[A-z]+Info/ ) {
<a name="l00836"></a>00836     $out = $in-&gt;to_hash($keen);
<a name="l00837"></a>00837   }
<a name="l00838"></a>00838   <span class="keywordflow">else</span> {
<a name="l00839"></a>00839     $out = $in;
<a name="l00840"></a>00840   }
<a name="l00841"></a>00841   <span class="keywordflow">if</span> ( defined $keen &amp;&amp;
<a name="l00842"></a>00842        $keen == 1 &amp;&amp;
<a name="l00843"></a>00843        $type =~ m/Bio::EnsEMBL::MetaData::[A-z]+Info/ )
<a name="l00844"></a>00844   {
<a name="l00845"></a>00845     $in-&gt;_preload();
<a name="l00846"></a>00846   }
<a name="l00847"></a>00847   <span class="keywordflow">return</span> $out;
<a name="l00848"></a>00848 } ## end sub to_hash
<a name="l00849"></a>00849 
<a name="l00850"></a>00850 =head1 INTERNAL METHODS
<a name="l00851"></a>00851 =head2 _get_core
<a name="l00852"></a>00852   Description: Convenience method to find core database
<a name="l00853"></a>00853   Returntype : DatabaseInfo
<a name="l00854"></a>00854   Exceptions : none
<a name="l00855"></a>00855   Caller     : <span class="keyword">internal</span>
<a name="l00856"></a>00856   Status     : Stable
<a name="l00857"></a>00857 =cut
<a name="l00858"></a>00858 sub _get_core {
<a name="l00859"></a>00859   my ($self) = @_;
<a name="l00860"></a>00860   <span class="keywordflow">if</span>(!defined $self-&gt;{core}) {
<a name="l00861"></a>00861     ($self-&gt;{core}) = grep {$_-&gt;type() eq <span class="stringliteral">&#39;core&#39;</span>}  @{$self-&gt;databases()}; 
<a name="l00862"></a>00862   }
<a name="l00863"></a>00863   <span class="keywordflow">return</span> $self-&gt;{core};
<a name="l00864"></a>00864 }
<a name="l00865"></a>00865 
<a name="l00866"></a>00866 
<a name="l00867"></a>00867 =head2 count_hash_values
<a name="l00868"></a>00868   Description: Sums values found in hash
<a name="l00869"></a>00869   Arg        : hashref
<a name="l00870"></a>00870   Returntype : integer
<a name="l00871"></a>00871   Exceptions : none
<a name="l00872"></a>00872   Caller     : <span class="keyword">internal</span>
<a name="l00873"></a>00873   Status     : Stable
<a name="l00874"></a>00874 =cut
<a name="l00875"></a>00875 sub count_hash_values {
<a name="l00876"></a>00876   my ( $self, $hash ) = @_;
<a name="l00877"></a>00877   my $tot = 0;
<a name="l00878"></a>00878   <span class="keywordflow">if</span> ( defined $hash ) {
<a name="l00879"></a>00879     <span class="keywordflow">for</span> my $v ( values %{$hash} ) {
<a name="l00880"></a>00880       $tot += $v;
<a name="l00881"></a>00881     }
<a name="l00882"></a>00882   }
<a name="l00883"></a>00883   <span class="keywordflow">return</span> $tot;
<a name="l00884"></a>00884 }
<a name="l00885"></a>00885 
<a name="l00886"></a>00886 =head2 count_hash_lengths
<a name="l00887"></a>00887   Description: Sums sizes of arrays found in hash as values
<a name="l00888"></a>00888   Arg        : hashref
<a name="l00889"></a>00889   Returntype : integer
<a name="l00890"></a>00890   Exceptions : none
<a name="l00891"></a>00891   Caller     : <span class="keyword">internal</span>
<a name="l00892"></a>00892   Status     : Stable
<a name="l00893"></a>00893 =cut
<a name="l00894"></a>00894 
<a name="l00895"></a>00895 sub count_hash_lengths {
<a name="l00896"></a>00896   my ( $self, $hash ) = @_;
<a name="l00897"></a>00897   my $tot = 0;
<a name="l00898"></a>00898   <span class="keywordflow">if</span> ( defined $hash ) {
<a name="l00899"></a>00899     <span class="keywordflow">for</span> my $v ( values %{$hash} ) {
<a name="l00900"></a>00900       $tot += scalar(@$v);
<a name="l00901"></a>00901     }
<a name="l00902"></a>00902   }
<a name="l00903"></a>00903   <span class="keywordflow">return</span> $tot;
<a name="l00904"></a>00904 }
<a name="l00905"></a>00905 
<a name="l00906"></a>00906 =head2 dbID
<a name="l00907"></a>00907   Arg        : (optional) dbID to <span class="keyword">set</span> <span class="keyword">set</span>
<a name="l00908"></a>00908   Description: Gets/sets <span class="keyword">internal</span> genome_id used as database primary key
<a name="l00909"></a>00909   Returntype : dbID <span class="keywordtype">string</span>
<a name="l00910"></a>00910   Exceptions : none
<a name="l00911"></a>00911   Caller     : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>
<a name="l00912"></a>00912   Status     : Stable
<a name="l00913"></a>00913 =cut
<a name="l00914"></a>00914 
<a name="l00915"></a>00915 sub dbID {
<a name="l00916"></a>00916   my ( $self, $id ) = @_;
<a name="l00917"></a>00917   $self-&gt;{dbID} = $id <span class="keywordflow">if</span> ( defined $id );
<a name="l00918"></a>00918   <span class="keywordflow">return</span> $self-&gt;{dbID};
<a name="l00919"></a>00919 }
<a name="l00920"></a>00920 
<a name="l00921"></a>00921 =head2 adaptor
<a name="l00922"></a>00922   Arg        : (optional) adaptor to <span class="keyword">set</span> <span class="keyword">set</span>
<a name="l00923"></a>00923   Description: Gets/sets GenomeInfoAdaptor
<a name="l00924"></a>00924   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>
<a name="l00925"></a>00925   Exceptions : none
<a name="l00926"></a>00926   Caller     : Internal
<a name="l00927"></a>00927   Status     : Stable
<a name="l00928"></a>00928 =cut
<a name="l00929"></a>00929 
<a name="l00930"></a>00930 sub adaptor {
<a name="l00931"></a>00931   my ( $self, $adaptor ) = @_;
<a name="l00932"></a>00932   $self-&gt;{adaptor} = $adaptor <span class="keywordflow">if</span> ( defined $adaptor );
<a name="l00933"></a>00933   <span class="keywordflow">return</span> $self-&gt;{adaptor};
<a name="l00934"></a>00934 }
<a name="l00935"></a>00935 
<a name="l00936"></a>00936 =head2 _preload
<a name="l00937"></a>00937   Description: Ensure all children are loaded (used <span class="keywordflow">for</span> hash transformation)
<a name="l00938"></a>00938   Returntype : none
<a name="l00939"></a>00939   Exceptions : none
<a name="l00940"></a>00940   Caller     : Internal
<a name="l00941"></a>00941   Status     : Stable
<a name="l00942"></a>00942 =cut
<a name="l00943"></a>00943 
<a name="l00944"></a>00944 sub _preload {
<a name="l00945"></a>00945   my ($self) = @_;
<a name="l00946"></a>00946   $self-&gt;annotations();
<a name="l00947"></a>00947   $self-&gt;compara();
<a name="l00948"></a>00948   $self-&gt;features();
<a name="l00949"></a>00949   $self-&gt;other_alignments();
<a name="l00950"></a>00950   $self-&gt;variations();
<a name="l00951"></a>00951   <span class="keywordflow">return</span>;
<a name="l00952"></a>00952 }
<a name="l00953"></a>00953 
<a name="l00954"></a>00954 =head2 _preload
<a name="l00955"></a>00955   Description: Remove all children (used after hash transformation to ensure <span class="keywordtype">object</span> is minimised)
<a name="l00956"></a>00956   Returntype : none
<a name="l00957"></a>00957   Exceptions : none
<a name="l00958"></a>00958   Caller     : dump_metadata.pl
<a name="l00959"></a>00959   Status     : Stable
<a name="l00960"></a>00960 =cut
<a name="l00961"></a>00961 
<a name="l00962"></a>00962 sub _unload {
<a name="l00963"></a>00963   my ($self) = @_;
<a name="l00964"></a>00964   $self-&gt;{annotations}      = undef;
<a name="l00965"></a>00965   $self-&gt;{compara}          = undef;
<a name="l00966"></a>00966   $self-&gt;{features}         = undef;
<a name="l00967"></a>00967   $self-&gt;{other_alignments} = undef;
<a name="l00968"></a>00968   $self-&gt;{variations}       = undef;
<a name="l00969"></a>00969   <span class="keywordflow">return</span>;
<a name="l00970"></a>00970 }
<a name="l00971"></a>00971 
<a name="l00972"></a>00972 1;
</pre></div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="GenomeInfo_8pm.html">GenomeInfo.pm</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
