<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Ensembl Metadata: Bio/EnsEMBL/MetaData/GenomeComparaInfo.pm Source File</title>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('GenomeComparaInfo_8pm.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">GenomeComparaInfo.pm</div>  </div>
</div>
<div class="contents">
<a href="GenomeComparaInfo_8pm.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 
<a name="l00002"></a>00002 =head1 LICENSE
<a name="l00003"></a>00003 
<a name="l00004"></a>00004 Copyright [1999-2014] EMBL-European Bioinformatics Institute
<a name="l00005"></a>00005 
<a name="l00006"></a>00006 Licensed under the Apache License, Version 2.0 (the <span class="stringliteral">&quot;License&quot;</span>);
<a name="l00007"></a>00007 you may not use <span class="keyword">this</span> file except in compliance with the License.
<a name="l00008"></a>00008 You may obtain a copy of the License at
<a name="l00009"></a>00009 
<a name="l00010"></a>00010      http:<span class="comment">//www.apache.org/licenses/LICENSE-2.0</span>
<a name="l00011"></a>00011 
<a name="l00012"></a>00012 Unless required by applicable law or agreed to in writing, software
<a name="l00013"></a>00013 distributed under the License is distributed on an <span class="stringliteral">&quot;AS IS&quot;</span> BASIS,
<a name="l00014"></a>00014 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
<a name="l00015"></a>00015 See the License <span class="keywordflow">for</span> the specific language governing permissions and
<a name="l00016"></a>00016 limitations under the License.
<a name="l00017"></a>00017 
<a name="l00018"></a>00018 =head1 CONTACT
<a name="l00019"></a>00019 
<a name="l00020"></a>00020   Please email comments or questions to the <span class="keyword">public</span> Ensembl
<a name="l00021"></a>00021   developers list at &lt;dev@ensembl.org&gt;.
<a name="l00022"></a>00022 
<a name="l00023"></a>00023   Questions may also be sent to the Ensembl help desk at
<a name="l00024"></a>00024   &lt;helpdesk@ensembl.org&gt;.
<a name="l00025"></a>00025 
<a name="l00026"></a>00026 =head1 NAME
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a>
<a name="l00029"></a><a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">00029</a> 
<a name="l00030"></a>00030 =head1 SYNOPSIS
<a name="l00031"></a>00031 
<a name="l00032"></a>00032       my $compara_info =
<a name="l00033"></a>00033         <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html#af724cfb183b4763d5f823acf84ae68a0">new</a>(
<a name="l00034"></a>00034                                    -DBNAME   =&gt; $compara-&gt;dbc()-&gt;dbname(),
<a name="l00035"></a>00035                                    -DIVISION =&gt; $division,
<a name="l00036"></a>00036                                    -METHOD   =&gt; $method,
<a name="l00037"></a>00037                                    -SET_NAME =&gt; $set_name,
<a name="l00038"></a>00038                                    -GENOMES  =&gt; [$genome1, $genome2],
<a name="l00039"></a>00039                                    -RELEASE  =&gt; $release);
<a name="l00040"></a>00040 
<a name="l00041"></a>00041 =head1 DESCRIPTION
<a name="l00042"></a>00042 
<a name="l00043"></a>00043 Object encapsulating information about a particular compara analysis and the genomes it involves
<a name="l00044"></a>00044 
<a name="l00045"></a>00045 =head1 SEE ALSO
<a name="l00046"></a>00046 
<a name="l00047"></a>00047 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1BaseInfo.html">Bio::EnsEMBL::MetaData::BaseInfo</a>
<a name="l00048"></a>00048 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00049"></a>00049 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeComparaInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeComparaInfoAdaptor</a>
<a name="l00050"></a>00050 
<a name="l00051"></a>00051 =head1 Author
<a name="l00052"></a>00052 
<a name="l00053"></a>00053 Dan Staines
<a name="l00054"></a>00054 
<a name="l00055"></a>00055 =cut
<a name="l00056"></a>00056 
<a name="l00057"></a>00057 <span class="keyword">package </span>Bio::EnsEMBL::MetaData::GenomeComparaInfo;
<a name="l00058"></a>00058 use base qw/<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1BaseInfo.html">Bio::EnsEMBL::MetaData::BaseInfo</a>/;
<a name="l00059"></a>00059 use Bio::EnsEMBL::Utils::Argument qw(rearrange);
<a name="l00060"></a>00060 use strict;
<a name="l00061"></a>00061 use warnings;
<a name="l00062"></a>00062 
<a name="l00063"></a>00063 =head1 CONSTRUCTOR
<a name="l00064"></a>00064 =head2 <span class="keyword">new</span>
<a name="l00065"></a>00065   Arg [-DIVISION]  : 
<a name="l00066"></a>00066        <span class="keywordtype">string</span> - Compara division e.g. plants, pan_homology
<a name="l00067"></a>00067   Arg [-METHOD]    : 
<a name="l00068"></a>00068        <span class="keywordtype">string</span> - compara method e.g. PROTEIN_TREES, LASTZ_NET
<a name="l00069"></a>00069   Arg [-DBNAME] : 
<a name="l00070"></a>00070        <span class="keywordtype">string</span> - name of the compara database in which the analysis can be found
<a name="l00071"></a>00071   Arg [-SET_NAME] : 
<a name="l00072"></a>00072        <span class="keywordtype">string</span> - optional name <span class="keywordflow">for</span> the analysis
<a name="l00073"></a>00073   Arg [-GENOMES]  : 
<a name="l00074"></a>00074        arrayref - list of genomes involved in analysis
<a name="l00075"></a>00075 
<a name="l00076"></a>00076   Example    : $info = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html#af724cfb183b4763d5f823acf84ae68a0">new</a>(...);
<a name="l00077"></a>00077   Description: Creates a <span class="keyword">new</span> info <span class="keywordtype">object</span>
<a name="l00078"></a>00078   Returntype : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a>
<a name="l00079"></a>00079   Exceptions : none
<a name="l00080"></a>00080   Caller     : general
<a name="l00081"></a>00081   Status     : Stable
<a name="l00082"></a>00082 
<a name="l00083"></a>00083 =cut
<a name="l00084"></a>00084 
<a name="l00085"></a>00085 sub <span class="keyword">new</span> {
<a name="l00086"></a>00086     my ( $class, @args ) = @_;
<a name="l00087"></a>00087     my $self = $class-&gt;SUPER::new(@args);
<a name="l00088"></a>00088   ( $self-&gt;{division}, $self-&gt;{method}, $self-&gt;{dbname},
<a name="l00089"></a>00089     $self-&gt;{set_name}, $self-&gt;{genomes}, $self-&gt;{release} )
<a name="l00090"></a>00090     = rearrange(
<a name="l00091"></a>00091                  [ <span class="stringliteral">&#39;DIVISION&#39;</span>, <span class="stringliteral">&#39;METHOD&#39;</span>, <span class="stringliteral">&#39;DBNAME&#39;</span>, <span class="stringliteral">&#39;SET_NAME&#39;</span>, <span class="stringliteral">&#39;GENOMES&#39;</span>, <span class="stringliteral">&#39;DATA_RELEASE&#39;</span>
<a name="l00092"></a>00092                  ],
<a name="l00093"></a>00093                  @args );
<a name="l00094"></a>00094   <span class="keywordflow">return</span> $self;
<a name="l00095"></a>00095 }
<a name="l00096"></a>00096 
<a name="l00097"></a>00097 =head1 ATTRIBUTE METHODS
<a name="l00098"></a>00098 =head2 dbname
<a name="l00099"></a>00099   Arg        : (optional) dbname to <span class="keyword">set</span>
<a name="l00100"></a>00100   Description: Gets/sets name of compara database
<a name="l00101"></a>00101   Returntype : <span class="keywordtype">string</span>
<a name="l00102"></a>00102   Exceptions : none
<a name="l00103"></a>00103   Caller     : general
<a name="l00104"></a>00104   Status     : Stable
<a name="l00105"></a>00105 =cut
<a name="l00106"></a>00106 
<a name="l00107"></a>00107 sub dbname {
<a name="l00108"></a>00108   my ( $self, $arg ) = @_;
<a name="l00109"></a>00109   $self-&gt;{dbname} = $arg <span class="keywordflow">if</span> ( defined $arg );
<a name="l00110"></a>00110   <span class="keywordflow">return</span> $self-&gt;{dbname};
<a name="l00111"></a>00111 }
<a name="l00112"></a>00112 
<a name="l00113"></a>00113 =head2 method
<a name="l00114"></a>00114   Arg        : (optional) method to <span class="keyword">set</span>
<a name="l00115"></a>00115   Description: Gets/sets name of compara method
<a name="l00116"></a>00116   Returntype : <span class="keywordtype">string</span>
<a name="l00117"></a>00117   Exceptions : none
<a name="l00118"></a>00118   Caller     : general
<a name="l00119"></a>00119   Status     : Stable
<a name="l00120"></a>00120 =cut
<a name="l00121"></a>00121 
<a name="l00122"></a>00122 sub method {
<a name="l00123"></a>00123   my ( $self, $arg ) = @_;
<a name="l00124"></a>00124   $self-&gt;{method} = $arg <span class="keywordflow">if</span> ( defined $arg );
<a name="l00125"></a>00125   <span class="keywordflow">return</span> $self-&gt;{method};
<a name="l00126"></a>00126 }
<a name="l00127"></a>00127 
<a name="l00128"></a>00128 =head2 set_name
<a name="l00129"></a>00129   Arg        : (optional) species <span class="keyword">set</span> name to <span class="keyword">set</span>
<a name="l00130"></a>00130   Description: Gets/sets name of species <span class="keyword">set</span> used in compara analysis
<a name="l00131"></a>00131   Returntype : <span class="keywordtype">string</span>
<a name="l00132"></a>00132   Exceptions : none
<a name="l00133"></a>00133   Caller     : general
<a name="l00134"></a>00134   Status     : Stable
<a name="l00135"></a>00135 =cut
<a name="l00136"></a>00136 
<a name="l00137"></a>00137 sub set_name {
<a name="l00138"></a>00138   my ( $self, $arg ) = @_;
<a name="l00139"></a>00139   $self-&gt;{set_name} = $arg <span class="keywordflow">if</span> ( defined $arg );
<a name="l00140"></a>00140   <span class="keywordflow">return</span> $self-&gt;{set_name};
<a name="l00141"></a>00141 }
<a name="l00142"></a>00142 
<a name="l00143"></a>00143 =head2 division
<a name="l00144"></a>00144   Arg        : (optional) division to <span class="keyword">set</span>
<a name="l00145"></a>00145   Description: Gets/sets Ensembl compara division
<a name="l00146"></a>00146   Returntype : <span class="keywordtype">string</span>
<a name="l00147"></a>00147   Exceptions : none
<a name="l00148"></a>00148   Caller     : general
<a name="l00149"></a>00149   Status     : Stable
<a name="l00150"></a>00150 =cut
<a name="l00151"></a>00151 
<a name="l00152"></a>00152 sub division {
<a name="l00153"></a>00153   my ( $self, $arg ) = @_;
<a name="l00154"></a>00154   $self-&gt;{division} = $arg <span class="keywordflow">if</span> ( defined $arg );
<a name="l00155"></a>00155   <span class="keywordflow">return</span> $self-&gt;{division};
<a name="l00156"></a>00156 }
<a name="l00157"></a>00157 
<a name="l00158"></a>00158 =head2 genomes
<a name="l00159"></a>00159   Arg        : (optional) genomes to <span class="keyword">set</span>
<a name="l00160"></a>00160   Description: Gets/sets arrayref of genomes 
<a name="l00161"></a>00161   Returntype : arrayref
<a name="l00162"></a>00162   Exceptions : none
<a name="l00163"></a>00163   Caller     : general
<a name="l00164"></a>00164   Status     : Stable
<a name="l00165"></a>00165 =cut
<a name="l00166"></a>00166 
<a name="l00167"></a>00167 sub genomes {
<a name="l00168"></a>00168   my ( $self, $arg ) = @_;
<a name="l00169"></a>00169   $self-&gt;{genomes} = $arg <span class="keywordflow">if</span> ( defined $arg );
<a name="l00170"></a>00170   $self-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1BaseInfo.html#a1b236763fc118984bee34ca2618e687f">_load_child</a>(<span class="stringliteral">&quot;genomes&quot;</span>,<span class="stringliteral">&quot;_fetch_compara_genomes&quot;</span>);
<a name="l00171"></a>00171   <span class="keywordflow">return</span> $self-&gt;{genomes};
<a name="l00172"></a>00172 }
<a name="l00173"></a>00173 
<a name="l00174"></a>00174 =head1 UTILITY METHODS
<a name="l00175"></a>00175 =head2 is_pan_compara
<a name="l00176"></a>00176   Arg        : (optional) 1/0 to <span class="keyword">set</span> 
<a name="l00177"></a>00177   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> analysis is part of the pan compara, 0 <span class="keywordflow">if</span> not
<a name="l00178"></a>00178   Returntype : 1 or 0
<a name="l00179"></a>00179   Exceptions : none
<a name="l00180"></a>00180   Caller     : general
<a name="l00181"></a>00181   Status     : Stable
<a name="l00182"></a>00182 =cut
<a name="l00183"></a>00183 
<a name="l00184"></a>00184 sub is_pan_compara {
<a name="l00185"></a>00185   my ($self) = @_;
<a name="l00186"></a>00186   <span class="keywordflow">return</span> $self-&gt;{division} eq <span class="stringliteral">&#39;EnsemblPan&#39;</span> ? 1 : 0;
<a name="l00187"></a>00187 }
<a name="l00188"></a>00188 
<a name="l00189"></a>00189 =head2 is_peptide_compara
<a name="l00190"></a>00190   Arg        : (optional) 1/0 to <span class="keyword">set</span> 
<a name="l00191"></a>00191   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> analysis is part of a peptide compara, 0 <span class="keywordflow">if</span> not
<a name="l00192"></a>00192   Returntype : 1 or 0
<a name="l00193"></a>00193   Exceptions : none
<a name="l00194"></a>00194   Caller     : general
<a name="l00195"></a>00195   Status     : Stable
<a name="l00196"></a>00196 =cut
<a name="l00197"></a>00197 
<a name="l00198"></a>00198 sub is_peptide_compara {
<a name="l00199"></a>00199   my ($self) = @_;
<a name="l00200"></a>00200   <span class="keywordflow">return</span> ( $self-&gt;{division} ne <span class="stringliteral">&#39;EnsemblPan&#39;</span> &amp;&amp;
<a name="l00201"></a>00201            $self-&gt;{method} eq <span class="stringliteral">&#39;PROTEIN_TREES&#39;</span> ) ? 1 : 0;
<a name="l00202"></a>00202 }
<a name="l00203"></a>00203 
<a name="l00204"></a>00204 =head2 is_dna_compara
<a name="l00205"></a>00205   Arg        : (optional) 1/0 to <span class="keyword">set</span> 
<a name="l00206"></a>00206   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> analysis is part of a DNA compara, 0 <span class="keywordflow">if</span> not
<a name="l00207"></a>00207   Returntype : 1 or 0
<a name="l00208"></a>00208   Exceptions : none
<a name="l00209"></a>00209   Caller     : general
<a name="l00210"></a>00210   Status     : Stable
<a name="l00211"></a>00211 =cut
<a name="l00212"></a>00212 
<a name="l00213"></a>00213 sub is_dna_compara {
<a name="l00214"></a>00214   my ($self) = @_;
<a name="l00215"></a>00215   <span class="keywordflow">return</span> ( $self-&gt;{method} eq <span class="stringliteral">&#39;TRANSLATED_BLAT_NET&#39;</span> ||
<a name="l00216"></a>00216            $self-&gt;{method} eq <span class="stringliteral">&#39;LASTZ_NET&#39;</span>   ||
<a name="l00217"></a>00217            $self-&gt;{method} eq <span class="stringliteral">&#39;TBLAT&#39;</span>       ||
<a name="l00218"></a>00218            $self-&gt;{method} eq <span class="stringliteral">&#39;ATAC&#39;</span>        ||
<a name="l00219"></a>00219            $self-&gt;{method} eq <span class="stringliteral">&#39;BLASTZ_NET&#39;</span> ) ? 1 : 0;
<a name="l00220"></a>00220 }
<a name="l00221"></a>00221 
<a name="l00222"></a>00222 =head2 is_synteny
<a name="l00223"></a>00223   Arg        : (optional) 1/0 to <span class="keyword">set</span> 
<a name="l00224"></a>00224   Description: Boolean-style method, returns 1 <span class="keywordflow">if</span> analysis is part of a synteny compara, 0 <span class="keywordflow">if</span> not
<a name="l00225"></a>00225   Returntype : 1 or 0
<a name="l00226"></a>00226   Exceptions : none
<a name="l00227"></a>00227   Caller     : general
<a name="l00228"></a>00228   Status     : Stable
<a name="l00229"></a>00229 =cut
<a name="l00230"></a>00230 
<a name="l00231"></a>00231 sub is_synteny {
<a name="l00232"></a>00232   my ($self) = @_;
<a name="l00233"></a>00233   <span class="keywordflow">return</span> ( $self-&gt;{method} eq <span class="stringliteral">&#39;SYNTENY&#39;</span> ) ? 1 : 0;
<a name="l00234"></a>00234 }
<a name="l00235"></a>00235 
<a name="l00236"></a>00236 =head2 to_hash
<a name="l00237"></a>00237   Description: Render compara as plain hash suitable <span class="keywordflow">for</span> export as JSON/XML
<a name="l00238"></a>00238   Returntype : Hashref
<a name="l00239"></a>00239   Exceptions : none
<a name="l00240"></a>00240   Caller     : general
<a name="l00241"></a>00241   Status     : Stable
<a name="l00242"></a>00242 =cut
<a name="l00243"></a>00243 
<a name="l00244"></a>00244 sub to_hash {
<a name="l00245"></a>00245   my ($in) = @_;
<a name="l00246"></a>00246   my $out = { method             =&gt; $in-&gt;{method},
<a name="l00247"></a>00247               division           =&gt; $in-&gt;{division},
<a name="l00248"></a>00248               dbname             =&gt; $in-&gt;{dbname},
<a name="l00249"></a>00249               set_name           =&gt; $in-&gt;{set_name},
<a name="l00250"></a>00250               is_pan_compara     =&gt; $in-&gt;is_pan_compara(),
<a name="l00251"></a>00251               is_peptide_compara =&gt; $in-&gt;is_peptide_compara(),
<a name="l00252"></a>00252               is_dna_compara     =&gt; $in-&gt;is_dna_compara(), };
<a name="l00253"></a>00253   $out-&gt;{genomes} = [];
<a name="l00254"></a>00254   <span class="keywordflow">for</span> my $genome ( @{ $in-&gt;genomes() } ) {
<a name="l00255"></a>00255     push @{ $out-&gt;{genomes} }, $genome-&gt;species();
<a name="l00256"></a>00256   }
<a name="l00257"></a>00257   <span class="keywordflow">return</span> $out;
<a name="l00258"></a>00258 }
<a name="l00259"></a>00259 
<a name="l00260"></a>00260 =head2 to_string
<a name="l00261"></a>00261   Description: Render as plain <span class="keywordtype">string</span>
<a name="l00262"></a>00262   Returntype : <span class="keywordtype">string</span>
<a name="l00263"></a>00263   Exceptions : none
<a name="l00264"></a>00264   Caller     : general
<a name="l00265"></a>00265   Status     : Stable
<a name="l00266"></a>00266 =cut
<a name="l00267"></a>00267 sub to_string {
<a name="l00268"></a>00268   my ($self) = @_;
<a name="l00269"></a>00269   <span class="keywordflow">return</span>
<a name="l00270"></a>00270     join( <span class="charliteral">&#39;/&#39;</span>, $self-&gt;division(), $self-&gt;method(), ($self-&gt;set_name()||<span class="charliteral">&#39;-&#39;</span>) );
<a name="l00271"></a>00271 }
<a name="l00272"></a>00272 
<a name="l00273"></a>00273 1;
</pre></div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="GenomeComparaInfo_8pm.html">GenomeComparaInfo.pm</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
