<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Ensembl Metadata: Bio/EnsEMBL/LookUp/LocalLookUp.pm Source File</title>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('LocalLookUp_8pm.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">LocalLookUp.pm</div>  </div>
</div>
<div class="contents">
<a href="LocalLookUp_8pm.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 
<a name="l00002"></a>00002 =head1 LICENSE
<a name="l00003"></a>00003 
<a name="l00004"></a>00004 Copyright [2009-2014] EMBL-European Bioinformatics Institute
<a name="l00005"></a>00005 
<a name="l00006"></a>00006 Licensed under the Apache License, Version 2.0 (the <span class="stringliteral">&quot;License&quot;</span>);
<a name="l00007"></a>00007 you may not use <span class="keyword">this</span> file except in compliance with the License.
<a name="l00008"></a>00008 You may obtain a copy of the License at
<a name="l00009"></a>00009 
<a name="l00010"></a>00010      http:<span class="comment">//www.apache.org/licenses/LICENSE-2.0</span>
<a name="l00011"></a>00011 
<a name="l00012"></a>00012 Unless required by applicable law or agreed to in writing, software
<a name="l00013"></a>00013 distributed under the License is distributed on an <span class="stringliteral">&quot;AS IS&quot;</span> BASIS,
<a name="l00014"></a><a class="code" href="namespaceBio.html">00014</a> WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
<a name="l00015"></a><a class="code" href="namespaceBio_1_1EnsEMBL.html">00015</a> See the License <span class="keywordflow">for</span> the specific language governing permissions and
<a name="l00016"></a><a class="code" href="namespaceBio_1_1EnsEMBL_1_1LookUp.html">00016</a> limitations under the License.
<a name="l00017"></a>00017 
<a name="l00018"></a>00018 =cut
<a name="l00019"></a>00019 
<a name="l00020"></a>00020 =pod
<a name="l00021"></a>00021 
<a name="l00022"></a>00022 =head1 CONTACT
<a name="l00023"></a>00023 
<a name="l00024"></a>00024   Please email comments or questions to the <span class="keyword">public</span> Ensembl
<a name="l00025"></a>00025   developers list at &lt;dev@ensembl.org&gt;.
<a name="l00026"></a>00026 
<a name="l00027"></a>00027   Questions may also be sent to the Ensembl help desk at
<a name="l00028"></a>00028   &lt;helpdesk@ensembl.org&gt;.
<a name="l00029"></a>00029 
<a name="l00030"></a>00030 =head1 NAME
<a name="l00031"></a>00031 
<a name="l00032"></a>00032 Bio::EnsEMBL::LookUp::LocalLookUp
<a name="l00033"></a>00033 
<a name="l00034"></a>00034 =head1 SYNOPSIS
<a name="l00035"></a>00035 
<a name="l00036"></a>00036 <span class="preprocessor"># creation from a database server</span>
<a name="l00037"></a>00037 <span class="preprocessor"></span>Bio::EnsEMBL::LookUp::LocalLookUp-&gt;register_all_dbs( $conf-&gt;{host},
<a name="l00038"></a>00038        $conf-&gt;{port}, $conf-&gt;{user}, $conf-&gt;{pass}, $conf-&gt;{db});
<a name="l00039"></a>00039 my $lookup = Bio::EnsEMBL::LookUp::LocalLookUp-&gt;new();
<a name="l00040"></a>00040 my $dbas = $lookup-&gt;registry()-&gt;get_all();
<a name="l00041"></a>00041 $dbas = $lookup-&gt;get_all_by_taxon_id(388919);
<a name="l00042"></a>00042 $dbas = $lookup-&gt;get_by_name_pattern(<span class="stringliteral">&quot;Escherichia.*&quot;</span>);
<a name="l00043"></a>00043 
<a name="l00044"></a>00044 <span class="preprocessor"># creation from a URL</span>
<a name="l00045"></a>00045 <span class="preprocessor"></span>my $lookup = Bio::EnsEMBL::LookUp::::LocalLookUp-&gt;new(-URL=&gt;<span class="stringliteral">&#39;http://bacteria.ensembl.org/registry.json&#39;</span>);
<a name="l00046"></a>00046 
<a name="l00047"></a>00047 =head1 DESCRIPTION
<a name="l00048"></a>00048 
<a name="l00049"></a>00049 This module is an implementation of <a class="code" href="classBio_1_1EnsEMBL_1_1LookUp.html">Bio::EnsEMBL::LookUp</a> that uses a local hash to store information about available Ensembl Genomes data. 
<a name="l00050"></a>00050 It can be constructed in a number of different ways:
<a name="l00051"></a>00051 
<a name="l00052"></a>00052 A remote cache file containing a JSON representation of the hash can be supplied:
<a name="l00053"></a>00053 
<a name="l00054"></a>00054     my $lookup = Bio::EnsEMBL::LookUp::LocalLookUp-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1LookUp.html#a004a9eccc10f89787bab3b32bb4a1366">new</a>(-URL=&gt;<span class="stringliteral">&#39;http://bacteria.ensembl.org/registry.json&#39;</span>);
<a name="l00055"></a>00055 
<a name="l00056"></a>00056 Alternatively, a local file containing the required JSON can be specified instead:
<a name="l00057"></a>00057 
<a name="l00058"></a>00058     my $lookup = Bio::EnsEMBL::LookUp::LocalLookUp-&gt;new(-FILE=&gt;<span class="stringliteral">&quot;/path/to/reg.json&quot;</span>);
<a name="l00059"></a>00059 
<a name="l00060"></a>00060 Finally, a Registry already loaded with core databases can be supplied:
<a name="l00061"></a>00061 
<a name="l00062"></a><a class="code" href="classBio_1_1EnsEMBL_1_1LookUp_1_1LocalLookUp.html">00062</a>     my $lookup = Bio::EnsEMBL::LookUp::LocalLookUp-&gt;new(-REGISTRY=&gt;<span class="stringliteral">&#39;Bio::EnsEMBL::Registry&#39;</span>);
<a name="l00063"></a>00063 
<a name="l00064"></a>00064 If the standard Registry is used, the argument can be omitted completely:
<a name="l00065"></a>00065 
<a name="l00066"></a>00066     my $lookup = Bio::EnsEMBL::LookUp::LocalLookUp-&gt;new();
<a name="l00067"></a>00067 
<a name="l00068"></a>00068 To populate the registry with just the Ensembl Genomes databases <span class="keywordflow">for</span> the current software 
<a name="l00069"></a>00069 release on a specified server, the following method can be used:
<a name="l00070"></a>00070 
<a name="l00071"></a>00071     Bio::EnsEMBL::LookUp::LocalLookUp-&gt;register_all_dbs( $host,
<a name="l00072"></a>00072        $port, $user, $pass);
<a name="l00073"></a>00073 
<a name="l00074"></a>00074 Once a lookup has been created, it can be used as specified in <a class="code" href="classBio_1_1EnsEMBL_1_1LookUp.html">Bio::EnsEMBL::LookUp</a>
<a name="l00075"></a>00075 
<a name="l00076"></a>00076 =head1 AUTHOR
<a name="l00077"></a>00077 
<a name="l00078"></a>00078 dstaines
<a name="l00079"></a>00079 
<a name="l00080"></a>00080 =head1 MAINTANER
<a name="l00081"></a>00081 
<a name="l00082"></a>00082 $Author$
<a name="l00083"></a>00083 
<a name="l00084"></a>00084 =head1 VERSION
<a name="l00085"></a>00085 
<a name="l00086"></a>00086 $Revision$
<a name="l00087"></a>00087 
<a name="l00088"></a>00088 =cut
<a name="l00089"></a>00089 
<a name="l00090"></a>00090 <span class="keyword">package </span>Bio::EnsEMBL::LookUp::LocalLookUp;
<a name="l00091"></a>00091 
<a name="l00092"></a>00092 use warnings;
<a name="l00093"></a>00093 use strict;
<a name="l00094"></a>00094 use Bio::EnsEMBL::ApiVersion;
<a name="l00095"></a>00095 use Bio::EnsEMBL::Utils::ConfigRegistry;
<a name="l00096"></a>00096 use Bio::EnsEMBL::DBSQL::DBAdaptor;
<a name="l00097"></a>00097 use Bio::EnsEMBL::Utils::Argument qw(rearrange);
<a name="l00098"></a>00098 use Bio::EnsEMBL::Utils::Scalar qw(assert_ref check_ref);
<a name="l00099"></a>00099 use Bio::EnsEMBL::Utils::Exception qw(<span class="keywordflow">throw</span> warning);
<a name="l00100"></a>00100 use Bio::EnsEMBL::Taxonomy::DBSQL::TaxonomyNodeAdaptor;
<a name="l00101"></a>00101 use List::MoreUtils qw(uniq);
<a name="l00102"></a>00102 use Scalar::Util qw(looks_like_number);
<a name="l00103"></a>00103 use DBI;
<a name="l00104"></a>00104 use JSON;
<a name="l00105"></a>00105 use LWP::Simple;
<a name="l00106"></a>00106 use Carp;
<a name="l00107"></a>00107 
<a name="l00108"></a>00108 my $default_cache_file = qw/lookup_cache.json/;
<a name="l00109"></a>00109 
<a name="l00110"></a>00110 =head1 SUBROUTINES/METHODS
<a name="l00111"></a>00111 
<a name="l00112"></a>00112 =head2 <span class="keyword">new</span>
<a name="l00113"></a>00113 
<a name="l00114"></a>00114   Description       : Creates a <span class="keyword">new</span> instance of <span class="keyword">this</span> <span class="keywordtype">object</span>. 
<a name="l00115"></a>00115   Arg [-REGISTRY]   : (optional) Registry module to use (<span class="keywordflow">default</span> is Bio::EnsEMBL::Registry)
<a name="l00116"></a>00116   Arg [-NO_CACHE]   : (optional) <span class="keywordtype">int</span> 1
<a name="l00117"></a>00117                This option will turn off the use of a local cache file <span class="keywordflow">for</span> storing species details
<a name="l00118"></a>00118   Arg [-CACHE_FILE] : (optional) String (<span class="keywordflow">default</span> lookup_cache.json)
<a name="l00119"></a>00119                 This option allows use of a user specified local cache file
<a name="l00120"></a>00120   Arg [-URL]        : (optional) 
<a name="l00121"></a>00121               This option allows the use of a remote resource supplying species details in JSON format. This is not used <span class="keywordflow">if</span> a local cache exists and -NO_CACHE is not <span class="keyword">set</span>
<a name="l00122"></a>00122   Arg [-FILE]       : (optional) 
<a name="l00123"></a>00123               This option allows the use of a file supplying species details in JSON format. This is not used <span class="keywordflow">if</span> a local cache exists and -NO_CACHE is not <span class="keyword">set</span>
<a name="l00124"></a>00124   Returntype        : Instance of lookup
<a name="l00125"></a>00125   Status            : Stable
<a name="l00126"></a>00126 
<a name="l00127"></a>00127   Example           : 
<a name="l00128"></a>00128   my $lookup = Bio::EnsEMBL::LookUp::LocalLookUp-&gt;new(
<a name="l00129"></a>00129                                         -REGISTRY =&gt; $reg);
<a name="l00130"></a>00130 =cut
<a name="l00131"></a>00131 
<a name="l00132"></a>00132 sub <span class="keyword">new</span> {
<a name="l00133"></a>00133   my ( $class, @args ) = @_;
<a name="l00134"></a>00134   my $self = bless( {}, ref($class) || $class );
<a name="l00135"></a>00135   my ( $reg, $url, $file, $nocache, $cache_file, $clear_cache, $skip_contigs ) =
<a name="l00136"></a>00136     rearrange( [qw(registry url file no_cache cache_file clear_cache skip_contigs)],
<a name="l00137"></a>00137                @args );
<a name="l00138"></a>00138   <span class="keywordflow">if</span> ( !defined $reg ) { $reg = <span class="stringliteral">&#39;Bio::EnsEMBL::Registry&#39;</span> }
<a name="l00139"></a>00139   $self-&gt;{registry}      = $reg;
<a name="l00140"></a>00140   $self-&gt;{dbas_by_taxid} = {};
<a name="l00141"></a>00141   $self-&gt;{dbas_by_name}  = {};
<a name="l00142"></a>00142   $self-&gt;{dbas_by_acc}   = {};
<a name="l00143"></a>00143   $self-&gt;{dbas_by_vacc}  = {};
<a name="l00144"></a>00144   $self-&gt;{dbas_by_gc}    = {};
<a name="l00145"></a>00145   $self-&gt;{dbas_by_vgc}   = {};
<a name="l00146"></a>00146   $self-&gt;{dbas}          = {};
<a name="l00147"></a>00147   $self-&gt;{dbc_meta}      = {};
<a name="l00148"></a>00148   $self-&gt;{skip_contigs}  = $skip_contigs;
<a name="l00149"></a>00149   $cache_file ||= $default_cache_file;
<a name="l00150"></a>00150   $self-&gt;cache_file($cache_file);
<a name="l00151"></a>00151 
<a name="l00152"></a>00152   <span class="keywordflow">if</span> ( defined $clear_cache &amp;&amp; $clear_cache == 1 ) {
<a name="l00153"></a>00153     $self-&gt;clear_cache();
<a name="l00154"></a>00154   }
<a name="l00155"></a>00155   <span class="keywordflow">if</span> ( !$nocache &amp;&amp; !defined $file &amp;&amp; -e $cache_file ) {
<a name="l00156"></a>00156     $file = $cache_file;
<a name="l00157"></a>00157   }
<a name="l00158"></a>00158   <span class="keywordflow">if</span> ( defined $file ) {
<a name="l00159"></a>00159     $self-&gt;load_registry_from_file($file);
<a name="l00160"></a>00160   }
<a name="l00161"></a>00161   elsif ( defined $url ) {
<a name="l00162"></a>00162     $self-&gt;load_registry_from_url($url);
<a name="l00163"></a>00163   }
<a name="l00164"></a>00164   <span class="keywordflow">else</span> {
<a name="l00165"></a>00165     $self-&gt;update_from_registry();
<a name="l00166"></a>00166   }
<a name="l00167"></a>00167   <span class="keywordflow">if</span> ( !$nocache &amp;&amp; !-e $cache_file ) {
<a name="l00168"></a>00168     $self-&gt;write_registry_to_file($cache_file);
<a name="l00169"></a>00169   }
<a name="l00170"></a>00170   <span class="keywordflow">return</span> $self;
<a name="l00171"></a>00171 } ## end sub <span class="keyword">new</span>
<a name="l00172"></a>00172 
<a name="l00173"></a>00173 =head2 registry
<a name="l00174"></a>00174 
<a name="l00175"></a>00175   Arg [1]     : Registry module to use (<span class="keywordflow">default</span> is Bio::EnsEMBL::Registry)
<a name="l00176"></a>00176   Description : Sets and retrieves the Registry 
<a name="l00177"></a>00177   Returntype  : Registry <span class="keywordflow">if</span> <span class="keyword">set</span>; otherwise undef
<a name="l00178"></a>00178   Exceptions  : <span class="keywordflow">if</span> an attempt is made to <span class="keyword">set</span> the value more than once
<a name="l00179"></a>00179   Status      : Stable
<a name="l00180"></a>00180 =cut
<a name="l00181"></a>00181 
<a name="l00182"></a>00182 sub registry {
<a name="l00183"></a>00183   my ( $self, $registry ) = @_;
<a name="l00184"></a>00184   <span class="keywordflow">if</span> ( defined $registry ) {
<a name="l00185"></a>00185     <span class="keywordflow">if</span> ( exists $self-&gt;{registry} ) {
<a name="l00186"></a>00186       <span class="keywordflow">throw</span>(<span class="stringliteral">&#39;Cannot reset the Registry object; already defined &#39;</span>);
<a name="l00187"></a>00187     }
<a name="l00188"></a>00188     $self-&gt;{registry} = $registry;
<a name="l00189"></a>00189     $self-&gt;update_from_registry();
<a name="l00190"></a>00190   }
<a name="l00191"></a>00191   <span class="keywordflow">return</span> $self-&gt;{registry};
<a name="l00192"></a>00192 }
<a name="l00193"></a>00193 
<a name="l00194"></a>00194 =head2 cache_file
<a name="l00195"></a>00195   Arg [1]     : File to use <span class="keywordflow">for</span> local caching
<a name="l00196"></a>00196   Description : Sets and retrieves the local cache file 
<a name="l00197"></a>00197   Returntype  : File name <span class="keywordflow">if</span> <span class="keyword">set</span>; otherwise undef
<a name="l00198"></a>00198   Status      : Stable
<a name="l00199"></a>00199 =cut
<a name="l00200"></a>00200 
<a name="l00201"></a>00201 sub cache_file {
<a name="l00202"></a>00202   my ( $self, $cache_file ) = @_;
<a name="l00203"></a>00203   <span class="keywordflow">if</span> ( defined $cache_file ) {
<a name="l00204"></a>00204     $self-&gt;{cache_file} = $cache_file;
<a name="l00205"></a>00205   }
<a name="l00206"></a>00206   <span class="keywordflow">return</span> $self-&gt;{cache_file};
<a name="l00207"></a>00207 }
<a name="l00208"></a>00208 
<a name="l00209"></a>00209 =head2  update_from_registry
<a name="l00210"></a>00210     Description : Update <span class="keyword">internal</span> hashes of database adaptors by name/taxids from the registry. Invoke when registry has been updated independently.
<a name="l00211"></a>00211     Argument    : (optional) If <span class="keyword">set</span> to 1, <span class="keywordflow">do</span> not update from a DBAdaptor <span class="keywordflow">if</span> already processed (allows updates of an existing LookUp instance)
<a name="l00212"></a>00212     Exceptions  : None
<a name="l00213"></a>00213 =cut
<a name="l00214"></a>00214 
<a name="l00215"></a>00215 sub update_from_registry {
<a name="l00216"></a>00216   my ( $self, $update_only ) = @_;
<a name="l00217"></a>00217   $self-&gt;_intern_db_connections();
<a name="l00218"></a>00218   <span class="keywordflow">for</span> my $dba (
<a name="l00219"></a>00219         @{ $self-&gt;registry()-&gt;get_all_DBAdaptors( -group =&gt; <span class="stringliteral">&#39;core&#39;</span> ) } )
<a name="l00220"></a>00220   {
<a name="l00221"></a>00221     $self-&gt;_hash_dba( $dba, $update_only );
<a name="l00222"></a>00222         $dba-&gt;dbc()-&gt;disconnect_if_idle();
<a name="l00223"></a>00223   }
<a name="l00224"></a>00224   <span class="keywordflow">return</span>;
<a name="l00225"></a>00225 }
<a name="l00226"></a>00226 
<a name="l00227"></a>00227 sub taxonomy_adaptor {
<a name="l00228"></a>00228   my ( $self, $adaptor ) = @_;
<a name="l00229"></a>00229   <span class="keywordflow">if</span> ( defined $adaptor ) {
<a name="l00230"></a>00230     $self-&gt;{taxonomy_adaptor} = $adaptor;
<a name="l00231"></a>00231   }
<a name="l00232"></a>00232   elsif ( !defined $self-&gt;{taxonomy_adaptor} ) {
<a name="l00233"></a>00233     $self-&gt;{taxonomy_adaptor} =
<a name="l00234"></a>00234       Bio::EnsEMBL::Taxonomy::DBSQL::TaxonomyNodeAdaptor-&gt;new_public();
<a name="l00235"></a>00235   }
<a name="l00236"></a>00236   <span class="keywordflow">return</span> $self-&gt;{taxonomy_adaptor};
<a name="l00237"></a>00237 }
<a name="l00238"></a>00238 
<a name="l00239"></a>00239 =head2 _register_dba
<a name="l00240"></a>00240     Description : Add a single DBAdaptor to the registry and the <span class="keyword">internal</span> hashes of details
<a name="l00241"></a>00241     Argument    : Bio::EnsEMBL::DBAdaptor
<a name="l00242"></a>00242     Return      : None
<a name="l00243"></a>00243 =cut
<a name="l00244"></a>00244 
<a name="l00245"></a>00245 sub _register_dba {
<a name="l00246"></a>00246   my ( $self, $species, $dba ) = @_;
<a name="l00247"></a>00247   $self-&gt;registry()-&gt;add_DBAdaptor( $species, <span class="stringliteral">&quot;core&quot;</span>, $dba );
<a name="l00248"></a>00248   $self-&gt;_hash_dba($dba);
<a name="l00249"></a>00249   <span class="keywordflow">return</span>;
<a name="l00250"></a>00250 }
<a name="l00251"></a>00251 
<a name="l00252"></a>00252 sub _dba_id {
<a name="l00253"></a>00253   my ( $self, $dba ) = @_;
<a name="l00254"></a>00254   <span class="keywordflow">return</span> $dba-&gt;dbc()-&gt;dbname() . <span class="charliteral">&#39;/&#39;</span> . $dba-&gt;species_id();
<a name="l00255"></a>00255 }
<a name="l00256"></a>00256 
<a name="l00257"></a>00257 sub _get_dbc_meta {
<a name="l00258"></a>00258   my ( $self, $dbc ) = @_;
<a name="l00259"></a>00259   my $name = $dbc-&gt;dbname();
<a name="l00260"></a>00260   <span class="keywordflow">if</span> ( !defined $self-&gt;{dbc_meta}{$name} ) {
<a name="l00261"></a>00261 <span class="preprocessor">    # load metadata in batch for the whole database</span>
<a name="l00262"></a>00262 <span class="preprocessor"></span>    $self-&gt;{dbc_meta}{$name} = {};
<a name="l00263"></a>00263 <span class="preprocessor">    # taxid</span>
<a name="l00264"></a>00264 <span class="preprocessor"></span>
<a name="l00265"></a>00265     $dbc-&gt;sql_helper()-&gt;execute(
<a name="l00266"></a>00266       -SQL =&gt;
<a name="l00267"></a>00267 q/select species_id,meta_value from meta where meta_key=<span class="stringliteral">&#39;species.taxonomy_id&#39;</span>/,
<a name="l00268"></a>00268       -PARAMS   =&gt; [],
<a name="l00269"></a>00269       -CALLBACK =&gt; sub {
<a name="l00270"></a>00270         my $row = shift @_;
<a name="l00271"></a>00271         $self-&gt;{dbc_meta}{$name}{ $row-&gt;[0] }{taxid} = $row-&gt;[1];
<a name="l00272"></a>00272       } );
<a name="l00273"></a>00273 <span class="preprocessor">    # aliases</span>
<a name="l00274"></a>00274 <span class="preprocessor"></span>    $dbc-&gt;sql_helper()-&gt;execute(
<a name="l00275"></a>00275       -SQL =&gt;
<a name="l00276"></a>00276 <span class="stringliteral">&#39;select species_id,meta_value from meta where meta_key=\&#39;species.alias\&#39;&#39;</span>,
<a name="l00277"></a>00277       -PARAMS   =&gt; [],
<a name="l00278"></a>00278       -CALLBACK =&gt; sub {
<a name="l00279"></a>00279         my @row = @{ shift @_ };
<a name="l00280"></a>00280         push @{ $self-&gt;{dbc_meta}{$name}{ $row[0] }{aliases} }, $row[1];
<a name="l00281"></a>00281       } );
<a name="l00282"></a>00282 <span class="preprocessor">    # assembly.accession</span>
<a name="l00283"></a>00283 <span class="preprocessor"></span>    $dbc-&gt;sql_helper()-&gt;execute(
<a name="l00284"></a>00284       -SQL =&gt;
<a name="l00285"></a>00285 <span class="stringliteral">&#39;select species_id,meta_value from meta where meta_key=\&#39;assembly.accession\&#39;&#39;</span>,
<a name="l00286"></a>00286       -PARAMS   =&gt; [],
<a name="l00287"></a>00287       -CALLBACK =&gt; sub {
<a name="l00288"></a>00288         my @row = @{ shift @_ };
<a name="l00289"></a>00289         $self-&gt;{dbc_meta}{$name}{ $row[0] }{assembly_accession} =
<a name="l00290"></a>00290           $row[1];
<a name="l00291"></a>00291       } );
<a name="l00292"></a>00292 
<a name="l00293"></a>00293         <span class="keywordflow">if</span>(!defined $self-&gt;{skip_contigs}) {
<a name="l00294"></a>00294 <span class="preprocessor">    # seq_region_names and synonyms</span>
<a name="l00295"></a>00295 <span class="preprocessor"></span>    $dbc-&gt;sql_helper()-&gt;execute(
<a name="l00296"></a>00296       -SQL =&gt;
<a name="l00297"></a>00297         q/select cs.species_id,sr.name,ss.synonym from coord_system cs 
<a name="l00298"></a>00298       join seq_region sr using (coord_system_id) 
<a name="l00299"></a>00299       left join seq_region_synonym ss <span class="keyword">using</span> (seq_region_id)/,
<a name="l00300"></a>00300       -PARAMS   =&gt; [],
<a name="l00301"></a>00301       -CALLBACK =&gt; sub {
<a name="l00302"></a>00302         my @row = @{ shift @_ };
<a name="l00303"></a>00303         push @{ $self-&gt;{dbc_meta}{$name}{ $row[0] }{accessions} },
<a name="l00304"></a>00304           $row[1];
<a name="l00305"></a>00305         <span class="keywordflow">if</span> ( defined $row[2] ) {
<a name="l00306"></a>00306           push @{ $self-&gt;{dbc_meta}{$name}{ $row[0] }{accessions} },
<a name="l00307"></a>00307             $row[2];
<a name="l00308"></a>00308         }
<a name="l00309"></a>00309       } );
<a name="l00310"></a>00310         }
<a name="l00311"></a>00311   } ## end <span class="keywordflow">if</span> ( !defined $self-&gt;{...})
<a name="l00312"></a>00312   <span class="keywordflow">return</span> $self-&gt;{dbc_meta}{$name};
<a name="l00313"></a>00313 } ## end sub _get_dbc_meta
<a name="l00314"></a>00314 
<a name="l00315"></a>00315 =head2 _hash_dba
<a name="l00316"></a>00316     Description : Add details from a DBAdaptor to the <span class="keyword">internal</span> hashes of details
<a name="l00317"></a>00317     Argument    : Bio::EnsEMBL::DBAdaptor
<a name="l00318"></a>00318     Argument    : (optional) If <span class="keyword">set</span> to 1, <span class="keywordflow">do</span> not update from a DBAdaptor <span class="keywordflow">if</span> already processed (allows updates of an existing LookUp instance)
<a name="l00319"></a>00319 =cut
<a name="l00320"></a>00320 
<a name="l00321"></a>00321 sub _hash_dba {
<a name="l00322"></a>00322   my ( $self, $dba, $update_only ) = @_;
<a name="l00323"></a>00323   my $nom = $self-&gt;_dba_id($dba);
<a name="l00324"></a>00324   <span class="keywordflow">if</span> ( !defined $update_only || !defined $self-&gt;{dbas}{$nom} ) {
<a name="l00325"></a>00325     my $dbc_meta = $self-&gt;_get_dbc_meta( $dba-&gt;dbc() );
<a name="l00326"></a>00326     my $dba_meta = $dbc_meta-&gt;{ $dba-&gt;species_id() };
<a name="l00327"></a>00327     $self-&gt;_hash_dba_from_values( $dba,
<a name="l00328"></a>00328                                   [ $dba_meta-&gt;{taxid} ],
<a name="l00329"></a>00329                                   $dba_meta-&gt;{aliases},
<a name="l00330"></a>00330                                   $dba_meta-&gt;{accessions},
<a name="l00331"></a>00331                                   $dba_meta-&gt;{assembly_accession} );
<a name="l00332"></a>00332   }
<a name="l00333"></a>00333   <span class="keywordflow">return</span>;
<a name="l00334"></a>00334 }
<a name="l00335"></a>00335 
<a name="l00336"></a>00336 =head2 _hash_dba_from_values
<a name="l00337"></a>00337     Description : Add supplied details to the <span class="keyword">internal</span> hashes of details
<a name="l00338"></a>00338     Argument    : Bio::EnsEMBL::DBAdaptor to hash
<a name="l00339"></a>00339     Argument    : Arrayref of taxonomy IDs to use as keys
<a name="l00340"></a>00340     Argument    : Arrayref of aliases to use as keys
<a name="l00341"></a>00341     Argument    : Arrayref of ENA accessions to use as keys
<a name="l00342"></a>00342 =cut
<a name="l00343"></a>00343 
<a name="l00344"></a>00344 sub _hash_dba_from_values {
<a name="l00345"></a>00345   my ( $self, $dba, $taxids, $aliases, $accessions, $vgc ) = @_;
<a name="l00346"></a>00346 
<a name="l00347"></a>00347   <span class="keywordflow">for</span> my $taxid ( @{$taxids} ) {
<a name="l00348"></a>00348     <span class="keywordflow">if</span> ( defined $taxid ) {
<a name="l00349"></a>00349       push @{ $self-&gt;{dbas_by_taxid}{$taxid} }, $dba;
<a name="l00350"></a>00350     }
<a name="l00351"></a>00351   }
<a name="l00352"></a>00352   <span class="keywordflow">for</span> my $name ( uniq( @{$aliases} ) ) {
<a name="l00353"></a>00353     push @{ $self-&gt;{dbas_by_name}{$name} }, $dba;
<a name="l00354"></a>00354   }
<a name="l00355"></a>00355   <span class="keywordflow">for</span> my $acc ( uniq( @{$accessions} ) ) {
<a name="l00356"></a>00356     push @{ $self-&gt;{dbas_by_vacc}{$acc} }, $dba;
<a name="l00357"></a>00357     $acc =~ s/\.[0-9]+$<span class="comment">//x;</span>
<a name="l00358"></a>00358     push @{ $self-&gt;{dbas_by_acc}{$acc} }, $dba;
<a name="l00359"></a>00359   }
<a name="l00360"></a>00360   <span class="keywordflow">if</span> ( defined $vgc ) {
<a name="l00361"></a>00361     <span class="keywordflow">if</span> ( defined $self-&gt;{dbas_by_vgc}{$vgc} ) {
<a name="l00362"></a>00362       croak <span class="stringliteral">&quot;DBA with GC accession $vgc already registered&quot;</span>;
<a name="l00363"></a>00363     }
<a name="l00364"></a>00364     $self-&gt;{dbas_by_vgc}{$vgc} = $dba;
<a name="l00365"></a>00365     $vgc =~ s/\.[0-9]+$<span class="comment">//x;</span>
<a name="l00366"></a>00366     <span class="keywordflow">if</span> ( defined $self-&gt;{dbas_by_gc}{$vgc} ) {
<a name="l00367"></a>00367       croak <span class="stringliteral">&quot;DBA with GC set chain $vgc already registered&quot;</span>;
<a name="l00368"></a>00368     }
<a name="l00369"></a>00369     $self-&gt;{dbas_by_gc}{$vgc} = $dba;
<a name="l00370"></a>00370   }
<a name="l00371"></a>00371   $self-&gt;{dbas}{ $self-&gt;_dba_id($dba) } = $dba;
<a name="l00372"></a>00372   <span class="keywordflow">return</span>;
<a name="l00373"></a>00373 } ## end sub _hash_dba_from_values
<a name="l00374"></a>00374 
<a name="l00375"></a>00375 sub _invert_dba_hash {
<a name="l00376"></a>00376   my ($hash) = @_;
<a name="l00377"></a>00377   my $inv_hash;
<a name="l00378"></a>00378   <span class="keywordflow">while</span> ( my ( $item, $dbas ) = each( %{$hash} ) ) {
<a name="l00379"></a>00379     <span class="keywordflow">if</span> ( ref $dbas ne <span class="stringliteral">&#39;ARRAY&#39;</span> ) {
<a name="l00380"></a>00380       $dbas = [$dbas];
<a name="l00381"></a>00381     }
<a name="l00382"></a>00382     <span class="keywordflow">for</span> my $dba ( @{$dbas} ) {
<a name="l00383"></a>00383       my $dba_loc = _dba_to_locator($dba);
<a name="l00384"></a>00384       push @{ $inv_hash-&gt;{$dba_loc} }, $item;
<a name="l00385"></a>00385     }
<a name="l00386"></a>00386   }
<a name="l00387"></a>00387   <span class="keywordflow">return</span> $inv_hash;
<a name="l00388"></a>00388 }
<a name="l00389"></a>00389 
<a name="l00390"></a>00390 =head2 _registry_to_hash
<a name="l00391"></a>00391     Description : Generate a hash array structure <span class="keywordflow">for</span> the current registry and species details <span class="keywordflow">for</span> turning to JSON
<a name="l00392"></a>00392     Returns     : Arrayref of hashes
<a name="l00393"></a>00393 =cut
<a name="l00394"></a>00394 
<a name="l00395"></a>00395 sub _registry_to_hash {
<a name="l00396"></a>00396   my ($self) = @_;
<a name="l00397"></a>00397 <span class="preprocessor">  # hash dbcs and dbas by locators</span>
<a name="l00398"></a>00398 <span class="preprocessor"></span>  my $dbc_hash;
<a name="l00399"></a>00399   my $dba_hash;
<a name="l00400"></a>00400   <span class="keywordflow">for</span> my $dba ( values %{ $self-&gt;{dbas} } ) {
<a name="l00401"></a>00401     my $dbc_loc = _dbc_to_locator( $dba-&gt;dbc() );
<a name="l00402"></a>00402     $dbc_hash-&gt;{$dbc_loc} = $dba-&gt;dbc();
<a name="l00403"></a>00403     push @{ $dba_hash-&gt;{$dbc_loc} }, $dba;
<a name="l00404"></a>00404   }
<a name="l00405"></a>00405 <span class="preprocessor">  # create auxillary hashes</span>
<a name="l00406"></a>00406 <span class="preprocessor"></span>  my $acc_hash   = _invert_dba_hash( $self-&gt;{dbas_by_vacc} );
<a name="l00407"></a>00407   my $name_hash  = _invert_dba_hash( $self-&gt;{dbas_by_name} );
<a name="l00408"></a>00408   my $taxid_hash = _invert_dba_hash( $self-&gt;{dbas_by_taxid} );
<a name="l00409"></a>00409   my $gc_hash    = _invert_dba_hash( $self-&gt;{dbas_by_vgc} );
<a name="l00410"></a>00410 <span class="preprocessor">  # create array of hashes</span>
<a name="l00411"></a>00411 <span class="preprocessor"></span>  my $out_arr;
<a name="l00412"></a>00412   <span class="keywordflow">while</span> ( my ( $dbc_loc, $dbconn ) = each( %{$dbc_hash} ) ) {
<a name="l00413"></a>00413     my $dbc_h = { driver   =&gt; $dbconn-&gt;driver(),
<a name="l00414"></a>00414                   host     =&gt; $dbconn-&gt;host(),
<a name="l00415"></a>00415                   port     =&gt; $dbconn-&gt;port(),
<a name="l00416"></a>00416                   username =&gt; $dbconn-&gt;username(),
<a name="l00417"></a>00417                   password =&gt; $dbconn-&gt;password(),
<a name="l00418"></a>00418                   dbname   =&gt; $dbconn-&gt;dbname(),
<a name="l00419"></a>00419                   species  =&gt; [] };
<a name="l00420"></a>00420     <span class="keywordflow">for</span> my $dba ( @{ $dba_hash-&gt;{$dbc_loc} } ) {
<a name="l00421"></a>00421       my $dba_loc = _dba_to_locator($dba);
<a name="l00422"></a>00422       my $gc      = $gc_hash-&gt;{$dba_loc};
<a name="l00423"></a>00423       <span class="keywordflow">if</span> ( ref $gc eq <span class="stringliteral">&#39;ARRAY&#39;</span> ) {
<a name="l00424"></a>00424         $gc = $gc-&gt;[0];
<a name="l00425"></a>00425       }
<a name="l00426"></a>00426       push @{ $dbc_h-&gt;{species} },
<a name="l00427"></a>00427         { species_id =&gt; $dba-&gt;species_id(),
<a name="l00428"></a>00428           species    =&gt; $dba-&gt;species(),
<a name="l00429"></a>00429           taxids     =&gt; $taxid_hash-&gt;{$dba_loc},
<a name="l00430"></a>00430           aliases    =&gt; $name_hash-&gt;{$dba_loc},
<a name="l00431"></a>00431           accessions =&gt; $acc_hash-&gt;{$dba_loc},
<a name="l00432"></a>00432           gc         =&gt; $gc };
<a name="l00433"></a>00433     }
<a name="l00434"></a>00434     push @{$out_arr}, $dbc_h;
<a name="l00435"></a>00435   }
<a name="l00436"></a>00436   <span class="keywordflow">return</span> $out_arr;
<a name="l00437"></a>00437 } ## end sub _registry_to_hash
<a name="l00438"></a>00438 
<a name="l00439"></a>00439 sub _registry_to_json {
<a name="l00440"></a>00440   my ($self) = @_;
<a name="l00441"></a>00441   my $out_arr = $self-&gt;_registry_to_hash();
<a name="l00442"></a>00442   my $json;
<a name="l00443"></a>00443   <span class="keywordflow">if</span> ($out_arr) {
<a name="l00444"></a>00444     $json = encode_json($out_arr);
<a name="l00445"></a>00445   }
<a name="l00446"></a>00446   <span class="keywordflow">else</span> {
<a name="l00447"></a>00447     carp(<span class="stringliteral">&#39;No DBAs found&#39;</span>);
<a name="l00448"></a>00448   }
<a name="l00449"></a>00449   <span class="keywordflow">return</span> $json;
<a name="l00450"></a>00450 }
<a name="l00451"></a>00451 
<a name="l00452"></a>00452 =head2 write_registry_to_file
<a name="l00453"></a>00453     Description : Write the contents of the registry and species lists to a JSON file
<a name="l00454"></a>00454     Argument    : File name
<a name="l00455"></a>00455     Return      : None
<a name="l00456"></a>00456 =cut
<a name="l00457"></a>00457 
<a name="l00458"></a>00458 sub write_registry_to_file {
<a name="l00459"></a>00459   my ( $self, $file ) = @_;
<a name="l00460"></a>00460   my $json = $self-&gt;_registry_to_json();
<a name="l00461"></a>00461   <span class="keywordflow">if</span> ($json) {
<a name="l00462"></a>00462     open( my $fh, <span class="charliteral">&#39;&gt;&#39;</span>, $file ) ||
<a name="l00463"></a>00463       croak <span class="stringliteral">&quot;Cannot open &#39;${file}&#39; for writing: $!&quot;</span>;
<a name="l00464"></a>00464     print $fh $json;
<a name="l00465"></a>00465     close($fh);
<a name="l00466"></a>00466   }
<a name="l00467"></a>00467   <span class="keywordflow">return</span>;
<a name="l00468"></a>00468 }
<a name="l00469"></a>00469 
<a name="l00470"></a>00470 sub clear_cache {
<a name="l00471"></a>00471   my ($self) = @_;
<a name="l00472"></a>00472   <span class="keywordflow">if</span> ( defined $self-&gt;cache_file() &amp;&amp; -e $self-&gt;cache_file() ) {
<a name="l00473"></a>00473     unlink $self-&gt;cache_file();
<a name="l00474"></a>00474   }
<a name="l00475"></a>00475   <span class="keywordflow">return</span>;
<a name="l00476"></a>00476 }
<a name="l00477"></a>00477 
<a name="l00478"></a>00478 =head2 _load_registry_from_json
<a name="l00479"></a>00479     Description : load the registry from the supplied JSON <span class="keywordtype">string</span>
<a name="l00480"></a>00480     Argument    : JSON <span class="keywordtype">string</span>
<a name="l00481"></a>00481 =cut
<a name="l00482"></a>00482 
<a name="l00483"></a>00483 sub _load_registry_from_json {
<a name="l00484"></a>00484   my ( $self, $json ) = @_;
<a name="l00485"></a>00485   my $reg_arr = decode_json($json);
<a name="l00486"></a>00486   <span class="keywordflow">for</span> my $dbc_h ( @{$reg_arr} ) {
<a name="l00487"></a>00487     my $dbc =
<a name="l00488"></a>00488       Bio::EnsEMBL::DBSQL::DBConnection-&gt;new(
<a name="l00489"></a>00489                                             -driver =&gt; $dbc_h-&gt;{driver},
<a name="l00490"></a>00490                                             -host   =&gt; $dbc_h-&gt;{host},
<a name="l00491"></a>00491                                             -port   =&gt; $dbc_h-&gt;{port},
<a name="l00492"></a>00492                                             -user =&gt; $dbc_h-&gt;{username},
<a name="l00493"></a>00493                                             -pass =&gt; $dbc_h-&gt;{password},
<a name="l00494"></a>00494                                             -dbname =&gt; $dbc_h-&gt;{dbname}
<a name="l00495"></a>00495       );
<a name="l00496"></a>00496     <span class="keywordflow">for</span> my $species ( @{ $dbc_h-&gt;{species} } ) {
<a name="l00497"></a>00497       my $dba =
<a name="l00498"></a>00498         Bio::EnsEMBL::DBSQL::DBAdaptor-&gt;new(
<a name="l00499"></a>00499                                   -DBCONN     =&gt; $dbc,
<a name="l00500"></a>00500                                   -species    =&gt; $species-&gt;{species},
<a name="l00501"></a>00501                                   -species_id =&gt; $species-&gt;{species_id},
<a name="l00502"></a>00502                                   -multispecies_db =&gt; 1,
<a name="l00503"></a>00503                                   -group           =&gt; <span class="stringliteral">&#39;core&#39;</span>, );
<a name="l00504"></a>00504 <span class="preprocessor">      # no need to register - found automagically</span>
<a name="l00505"></a>00505 <span class="preprocessor"></span>      my $gc = $species-&gt;{gc};
<a name="l00506"></a>00506       <span class="keywordflow">if</span> ( ref $gc eq <span class="stringliteral">&#39;ARRAY&#39;</span> ) { $gc = $gc-&gt;[0]; }
<a name="l00507"></a>00507       $self-&gt;_hash_dba_from_values( $dba,
<a name="l00508"></a>00508                                     $species-&gt;{taxids},
<a name="l00509"></a>00509                                     $species-&gt;{aliases},
<a name="l00510"></a>00510                                     $species-&gt;{accessions},
<a name="l00511"></a>00511                                     $species-&gt;{gc} );
<a name="l00512"></a>00512     }
<a name="l00513"></a>00513   } ## end <span class="keywordflow">for</span> my $dbc_h ( @{$reg_arr...})
<a name="l00514"></a>00514   $self-&gt;_intern_db_connections();
<a name="l00515"></a>00515   <span class="keywordflow">return</span>;
<a name="l00516"></a>00516 } ## end sub _load_registry_from_json
<a name="l00517"></a>00517 
<a name="l00518"></a>00518 sub load_registry_from_file {
<a name="l00519"></a>00519   my ( $self, $file ) = @_;
<a name="l00520"></a>00520   open( my $fh, <span class="charliteral">&#39;&lt;&#39;</span>, $file ) or croak &quot;Cannot open file $file: $!&quot;;
<a name="l00521"></a>00521   my $data = do { local $/ = undef; &lt;$fh&gt; };
<a name="l00522"></a>00522   close($fh);
<a name="l00523"></a>00523   <span class="keywordflow">if</span> ( defined $data &amp;&amp; $data ne <span class="stringliteral">&#39;&#39;</span> ) {
<a name="l00524"></a>00524     $self-&gt;_load_registry_from_json($data);
<a name="l00525"></a>00525   }
<a name="l00526"></a>00526   <span class="keywordflow">return</span>;
<a name="l00527"></a>00527 }
<a name="l00528"></a>00528 
<a name="l00529"></a>00529 sub load_registry_from_url {
<a name="l00530"></a>00530   my ( $self, $url ) = @_;
<a name="l00531"></a>00531   my $json = <span class="keyword">get</span>($url);
<a name="l00532"></a>00532   croak <span class="stringliteral">&quot;Could not retrieve JSON from $url: $@&quot;</span> unless defined $json;
<a name="l00533"></a>00533   $self-&gt;_load_registry_from_json($json);
<a name="l00534"></a>00534   <span class="keywordflow">return</span>;
<a name="l00535"></a>00535 }
<a name="l00536"></a>00536 
<a name="l00537"></a>00537 =head2 dba_to_args
<a name="l00538"></a>00538     Description : Dump the arguments needed <span class="keywordflow">for</span> contructing a DBA
<a name="l00539"></a>00539     Argument    : Bio::EnsEMBL::DBAdaptor
<a name="l00540"></a>00540     Return      : Arrayref of args
<a name="l00541"></a>00541 =cut
<a name="l00542"></a>00542 
<a name="l00543"></a>00543 sub dba_to_args {
<a name="l00544"></a>00544   my ( $self, $dba ) = @_;
<a name="l00545"></a>00545   my $dbc = $dba-&gt;dbc();
<a name="l00546"></a>00546   my $args = [ -species         =&gt; $dba-&gt;species(),
<a name="l00547"></a>00547                -multispecies_db =&gt; $dba-&gt;is_multispecies(),
<a name="l00548"></a>00548                -species_id      =&gt; $dba-&gt;species_id(),
<a name="l00549"></a>00549                -group           =&gt; $dba-&gt;group(),
<a name="l00550"></a>00550                -host            =&gt; $dbc-&gt;host(),
<a name="l00551"></a>00551                -port            =&gt; $dbc-&gt;port(),
<a name="l00552"></a>00552                -user            =&gt; $dbc-&gt;username(),
<a name="l00553"></a>00553                -pass            =&gt; $dbc-&gt;password(),
<a name="l00554"></a>00554                -driver          =&gt; $dbc-&gt;driver(),
<a name="l00555"></a>00555                -dbname          =&gt; $dbc-&gt;dbname() ];
<a name="l00556"></a>00556   <span class="keywordflow">return</span> $args;
<a name="l00557"></a>00557 }
<a name="l00558"></a>00558 
<a name="l00559"></a>00559 =head2 _dba_to_locator
<a name="l00560"></a>00560     Description : <span class="keywordflow">return</span> a hash key <span class="keywordflow">for</span> a DBAdaptor
<a name="l00561"></a>00561     Argument    : Bio::EnsEMBL::DBAdaptor
<a name="l00562"></a>00562 =cut
<a name="l00563"></a>00563 
<a name="l00564"></a>00564 sub _dba_to_locator {
<a name="l00565"></a>00565   my ($dba) = @_;
<a name="l00566"></a>00566   confess(
<a name="l00567"></a>00567     <span class="stringliteral">&quot;Argument must be Bio::EnsEMBL::DBSQL::DBAdaptor not &quot;</span> . ref($dba) )
<a name="l00568"></a>00568     unless ref($dba) eq &quot;Bio::EnsEMBL::DBSQL::DBAdaptor&quot;;
<a name="l00569"></a>00569   my $locator =
<a name="l00570"></a>00570     join( q{!-!}, $dba-&gt;species_id(), _dbc_to_locator( $dba-&gt;dbc() ) );
<a name="l00571"></a>00571   <span class="keywordflow">return</span> $locator;
<a name="l00572"></a>00572 }
<a name="l00573"></a>00573 
<a name="l00574"></a>00574 =head2 _dbc_to_locator
<a name="l00575"></a>00575     Description : <span class="keywordflow">return</span> a hash key <span class="keywordflow">for</span> a DBConnection
<a name="l00576"></a>00576     Argument    : Bio::EnsEMBL::DBConnection
<a name="l00577"></a>00577 =cut
<a name="l00578"></a>00578 
<a name="l00579"></a>00579 sub _dbc_to_locator {
<a name="l00580"></a>00580   my ($dbc) = @_;
<a name="l00581"></a>00581   my $locator = join( q{!-!},
<a name="l00582"></a>00582                       $dbc-&gt;host(),   $dbc-&gt;dbname(),
<a name="l00583"></a>00583                       $dbc-&gt;driver(), $dbc-&gt;port(),
<a name="l00584"></a>00584                       $dbc-&gt;username() );
<a name="l00585"></a>00585   <span class="keywordflow">return</span> $locator;
<a name="l00586"></a>00586 }
<a name="l00587"></a>00587 
<a name="l00588"></a>00588 =head2 _intern_db_connections()
<a name="l00589"></a>00589   Description : Go through all available DBAdaptors of registry and ensure they use the same
<a name="l00590"></a>00590                 DBConnection instance.
<a name="l00591"></a>00591   Exceptions  : None
<a name="l00592"></a>00592   Status      : At Risk
<a name="l00593"></a>00593 =cut
<a name="l00594"></a>00594 
<a name="l00595"></a>00595 sub _intern_db_connections {
<a name="l00596"></a>00596   my ($self) = @_;
<a name="l00597"></a>00597   my $adaptors = $self-&gt;registry()-&gt;get_all_DBAdaptors();
<a name="l00598"></a>00598   my %dbc_intern;
<a name="l00599"></a>00599   <span class="keywordflow">foreach</span> my $dba ( @{$adaptors} ) {
<a name="l00600"></a>00600     my $dbc     = $dba-&gt;dbc();
<a name="l00601"></a>00601     my $locator = _dbc_to_locator($dbc);
<a name="l00602"></a>00602     <span class="keywordflow">if</span> ( exists $dbc_intern{$locator} ) {
<a name="l00603"></a>00603       $dba-&gt;dbc( $dbc_intern{$locator} );
<a name="l00604"></a>00604     }
<a name="l00605"></a>00605     <span class="keywordflow">else</span> {
<a name="l00606"></a>00606       $dbc_intern{$locator} = $dba-&gt;dbc();
<a name="l00607"></a>00607     }
<a name="l00608"></a>00608   }
<a name="l00609"></a>00609   <span class="keywordflow">return</span>;
<a name="l00610"></a>00610 }
<a name="l00611"></a>00611 
<a name="l00612"></a>00612 =head2 get_all_DBConnections
<a name="l00613"></a>00613     Description : Return all database connections used by the DBAs retrieved from the registry
<a name="l00614"></a>00614     Argument    : None
<a name="l00615"></a>00615     Exceptions  : None
<a name="l00616"></a>00616     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DBConnection
<a name="l00617"></a>00617 =cut
<a name="l00618"></a>00618 
<a name="l00619"></a>00619 sub get_all_DBConnections {
<a name="l00620"></a>00620   my ($self) = @_;
<a name="l00621"></a>00621   my $adaptors = $self-&gt;registry()-&gt;get_all_DBAdaptors();
<a name="l00622"></a>00622   my %dbcs;
<a name="l00623"></a>00623   <span class="keywordflow">foreach</span> my $dba ( values %{ $self-&gt;{dbas} } ) {
<a name="l00624"></a>00624     my $dbc = $dba-&gt;dbc();
<a name="l00625"></a>00625     $dbcs{ _dbc_to_locator($dbc) } = $dbc;
<a name="l00626"></a>00626   }
<a name="l00627"></a>00627   <span class="keywordflow">return</span> [ values %dbcs ];
<a name="l00628"></a>00628 }
<a name="l00629"></a>00629 
<a name="l00630"></a>00630 =head2 get_all_dbnames
<a name="l00631"></a>00631     Description : Return all database names used by the DBAs retrieved from the registry
<a name="l00632"></a>00632     Argument    : None
<a name="l00633"></a>00633     Exceptions  : None
<a name="l00634"></a>00634     Return type : Arrayref of strings
<a name="l00635"></a>00635 =cut
<a name="l00636"></a>00636 
<a name="l00637"></a>00637 sub get_all_dbnames {
<a name="l00638"></a>00638   my ($self) = @_;
<a name="l00639"></a>00639   my %dbnames;
<a name="l00640"></a>00640   <span class="keywordflow">foreach</span> my $dba ( values %{ $self-&gt;{dbas} } ) {
<a name="l00641"></a>00641     my $dbc = $dba-&gt;dbc();
<a name="l00642"></a>00642     $dbnames{ $dbc-&gt;dbname }++;
<a name="l00643"></a>00643   }
<a name="l00644"></a>00644   <span class="keywordflow">return</span> [ keys %dbnames ];
<a name="l00645"></a>00645 }
<a name="l00646"></a>00646 
<a name="l00647"></a>00647 =head2 get_all
<a name="l00648"></a>00648     Description : Return all database adaptors that have been retrieved from registry
<a name="l00649"></a>00649     Argument    : None
<a name="l00650"></a>00650     Exceptions  : None
<a name="l00651"></a>00651     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00652"></a>00652 =cut
<a name="l00653"></a>00653 
<a name="l00654"></a>00654 sub get_all {
<a name="l00655"></a>00655   my ($self) = @_;
<a name="l00656"></a>00656   <span class="keywordflow">return</span> [ values %{ $self-&gt;{dbas} } ];
<a name="l00657"></a>00657 }
<a name="l00658"></a>00658 
<a name="l00659"></a>00659 sub get_all_by_taxon_branch {
<a name="l00660"></a>00660   my ( $self, $root ) = @_;
<a name="l00661"></a>00661   <span class="keywordflow">if</span> ( ref($root) ne <span class="stringliteral">&#39;Bio::EnsEMBL::Taxonomy::TaxonomyNode&#39;</span> ) {
<a name="l00662"></a>00662     <span class="keywordflow">if</span> ( looks_like_number($root) ) {
<a name="l00663"></a>00663       $root = $self-&gt;taxonomy_adaptor()-&gt;fetch_by_taxon_id($root);
<a name="l00664"></a>00664     }
<a name="l00665"></a>00665     <span class="keywordflow">else</span> {
<a name="l00666"></a>00666       ($root) =
<a name="l00667"></a>00667         @{ $self-&gt;taxonomy_adaptor()-&gt;fetch_all_by_name($root) };
<a name="l00668"></a>00668     }
<a name="l00669"></a>00669   }
<a name="l00670"></a>00670   <span class="keywordflow">if</span> ( !defined $root ) {
<a name="l00671"></a>00671     <span class="keywordflow">return</span> [];
<a name="l00672"></a>00672   }
<a name="l00673"></a>00673   my @genomes = @{ $self-&gt;get_all_by_taxon_id( $root-&gt;taxon_id() ) };
<a name="l00674"></a>00674   <span class="keywordflow">for</span> my $node ( @{ $root-&gt;adaptor()-&gt;fetch_descendants($root) } ) {
<a name="l00675"></a>00675     @genomes = ( @genomes,
<a name="l00676"></a>00676                  @{ $self-&gt;get_all_by_taxon_id( $node-&gt;taxon_id() ) } );
<a name="l00677"></a>00677   }
<a name="l00678"></a>00678   <span class="keywordflow">return</span> \@genomes;
<a name="l00679"></a>00679 }
<a name="l00680"></a>00680 
<a name="l00681"></a>00681 =head2 get_all_by_taxon_id
<a name="l00682"></a>00682     Description : Returns all database adaptors that have the supplied taxonomy ID
<a name="l00683"></a>00683     Argument    : Int
<a name="l00684"></a>00684     Exceptions  : None
<a name="l00685"></a>00685     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00686"></a>00686 =cut
<a name="l00687"></a>00687 
<a name="l00688"></a>00688 sub get_all_by_taxon_id {
<a name="l00689"></a>00689   my ( $self, $id ) = @_;
<a name="l00690"></a>00690   <span class="keywordflow">return</span> $self-&gt;{dbas_by_taxid}{$id} || [];
<a name="l00691"></a>00691 }
<a name="l00692"></a>00692 
<a name="l00693"></a>00693 =head2 get_by_name_exact
<a name="l00694"></a>00694     Description : Return all database adaptors that have the supplied <span class="keywordtype">string</span> as an alias/name
<a name="l00695"></a>00695     Argument    : String
<a name="l00696"></a>00696     Exceptions  : None
<a name="l00697"></a>00697     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00698"></a>00698 =cut
<a name="l00699"></a>00699 
<a name="l00700"></a>00700 sub get_by_name_exact {
<a name="l00701"></a>00701   my ( $self, $name ) = @_;
<a name="l00702"></a>00702   <span class="keywordflow">return</span> $self-&gt;{dbas_by_name}{$name};
<a name="l00703"></a>00703 }
<a name="l00704"></a>00704 
<a name="l00705"></a>00705 =head2 get_all_by_accession
<a name="l00706"></a>00706     Description : Returns the database adaptor(s) that contains a seq_region with the supplied INSDC accession (or other seq_region name)
<a name="l00707"></a>00707     Argument    : Int
<a name="l00708"></a>00708     Exceptions  : None
<a name="l00709"></a>00709     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00710"></a>00710 =cut    
<a name="l00711"></a>00711 
<a name="l00712"></a>00712 sub get_all_by_accession {
<a name="l00713"></a>00713   my ( $self, $acc ) = @_;
<a name="l00714"></a>00714   my $dba = $self-&gt;{dbas_by_vacc}{$acc};
<a name="l00715"></a>00715   <span class="keywordflow">if</span> ( !defined $dba ) {
<a name="l00716"></a>00716     $acc =~ s/\.[0-9]+$<span class="comment">//x;</span>
<a name="l00717"></a>00717     $dba = $self-&gt;{dbas_by_acc}{$acc};
<a name="l00718"></a>00718   }
<a name="l00719"></a>00719   <span class="keywordflow">return</span> $dba;
<a name="l00720"></a>00720 }
<a name="l00721"></a>00721 
<a name="l00722"></a>00722 =head2 get_by_assembly_accession
<a name="l00723"></a>00723     Description : Returns the database adaptor that contains the assembly with the supplied INSDC assembly accession
<a name="l00724"></a>00724     Argument    : Int
<a name="l00725"></a>00725     Exceptions  : None
<a name="l00726"></a>00726     Return type : Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00727"></a>00727 =cut
<a name="l00728"></a>00728 
<a name="l00729"></a>00729 sub get_by_assembly_accession {
<a name="l00730"></a>00730   my ( $self, $acc ) = @_;
<a name="l00731"></a>00731   my $dba = $self-&gt;{dbas_by_vgc}{$acc};
<a name="l00732"></a>00732   <span class="keywordflow">if</span> ( !defined $dba ) {
<a name="l00733"></a>00733     $acc =~ s/\.[0-9]+$<span class="comment">//x;</span>
<a name="l00734"></a>00734     $dba = $self-&gt;{dbas_by_gc}{$acc};
<a name="l00735"></a>00735   }
<a name="l00736"></a>00736   <span class="keywordflow">return</span> $dba;
<a name="l00737"></a>00737 }
<a name="l00738"></a>00738 
<a name="l00739"></a>00739 =head2 get_all_by_name_pattern
<a name="l00740"></a>00740     Description : Return all database adaptors that have an alias/name that match the supplied regexp
<a name="l00741"></a>00741     Argument    : String
<a name="l00742"></a>00742     Exceptions  : None
<a name="l00743"></a>00743     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00744"></a>00744 =cut    
<a name="l00745"></a>00745 
<a name="l00746"></a>00746 sub get_all_by_name_pattern {
<a name="l00747"></a>00747   my ( $self, $pattern ) = @_;
<a name="l00748"></a>00748   my %dbas;
<a name="l00749"></a>00749   <span class="keywordflow">for</span> my $name ( keys %{ $self-&gt;{dbas_by_name} } ) {
<a name="l00750"></a>00750     <span class="keywordflow">if</span> ( $name =~ m/$pattern/x ) {
<a name="l00751"></a>00751       <span class="keywordflow">for</span> my $dba ( @{ $self-&gt;{dbas_by_name}{$name} } ) {
<a name="l00752"></a>00752         my $key = _dba_to_locator($dba);
<a name="l00753"></a>00753         $dbas{$key} = $dba;
<a name="l00754"></a>00754       }
<a name="l00755"></a>00755     }
<a name="l00756"></a>00756   }
<a name="l00757"></a>00757   <span class="keywordflow">return</span> [ values(%dbas) ];
<a name="l00758"></a>00758 }
<a name="l00759"></a>00759 
<a name="l00760"></a>00760 =head2 get_all_by_dbname
<a name="l00761"></a>00761     Description : Returns all database adaptors that have the supplied dbname
<a name="l00762"></a>00762     Argument    : String
<a name="l00763"></a>00763     Exceptions  : None
<a name="l00764"></a>00764     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00765"></a>00765 =cut
<a name="l00766"></a>00766 
<a name="l00767"></a>00767 sub get_all_by_dbname {
<a name="l00768"></a>00768   my ( $self, $dbname ) = @_;
<a name="l00769"></a>00769 
<a name="l00770"></a>00770   my @filtered_dbas;
<a name="l00771"></a>00771 
<a name="l00772"></a>00772   my $all = $self-&gt;get_all();
<a name="l00773"></a>00773   <span class="keywordflow">foreach</span> my $dba ( values %{ $self-&gt;{dbas} } ) {
<a name="l00774"></a>00774     <span class="keywordflow">if</span> ( $dba-&gt;dbc-&gt;dbname eq $dbname ) {
<a name="l00775"></a>00775       push @filtered_dbas, $dba;
<a name="l00776"></a>00776     }
<a name="l00777"></a>00777   }
<a name="l00778"></a>00778 
<a name="l00779"></a>00779   <span class="keywordflow">return</span> \@filtered_dbas;
<a name="l00780"></a>00780 }
<a name="l00781"></a>00781 
<a name="l00782"></a>00782 =head2 get_all_taxon_ids
<a name="l00783"></a>00783     Description : Return list of all taxon IDs registered with the helper
<a name="l00784"></a>00784     Exceptions  : None
<a name="l00785"></a>00785     Return type : Arrayref of integers
<a name="l00786"></a>00786 =cut
<a name="l00787"></a>00787 
<a name="l00788"></a>00788 sub get_all_taxon_ids {
<a name="l00789"></a>00789   my ($self) = @_;
<a name="l00790"></a>00790   <span class="keywordflow">return</span> [ keys( %{ $self-&gt;{dbas_by_taxid} } ) ];
<a name="l00791"></a>00791 }
<a name="l00792"></a>00792 
<a name="l00793"></a>00793 =head2 get_all_names
<a name="l00794"></a>00794     Description : Return list of all species names registered with the helper
<a name="l00795"></a>00795     Exceptions  : None
<a name="l00796"></a>00796     Return type : Arrayref of strings
<a name="l00797"></a>00797 =cut
<a name="l00798"></a>00798 
<a name="l00799"></a>00799 sub get_all_names {
<a name="l00800"></a>00800   my ($self) = @_;
<a name="l00801"></a>00801   <span class="keywordflow">return</span> [ keys( %{ $self-&gt;{dbas_by_name} } ) ];
<a name="l00802"></a>00802 }
<a name="l00803"></a>00803 
<a name="l00804"></a>00804 =head2 get_all_accessions
<a name="l00805"></a>00805     Description : Return list of all INSDC sequence accessions (or other seq_region names) registered with the helper
<a name="l00806"></a>00806     Exceptions  : None
<a name="l00807"></a>00807     Return type : Arrayref of strings
<a name="l00808"></a>00808 =cut
<a name="l00809"></a>00809 
<a name="l00810"></a>00810 sub get_all_accessions {
<a name="l00811"></a>00811   my ($self) = @_;
<a name="l00812"></a>00812   <span class="keywordflow">return</span> [ keys( %{ $self-&gt;{dbas_by_acc} } ) ];
<a name="l00813"></a>00813 }
<a name="l00814"></a>00814 
<a name="l00815"></a>00815 =head2 get_all_versioned_accessions
<a name="l00816"></a>00816     Description : Return list of all versioned INSDC sequence accessions (or other seq_region names) registered with the helper
<a name="l00817"></a>00817     Exceptions  : None
<a name="l00818"></a>00818     Return type : Arrayref of strings
<a name="l00819"></a>00819 =cut
<a name="l00820"></a>00820 
<a name="l00821"></a>00821 sub get_all_versioned_accessions {
<a name="l00822"></a>00822   my ($self) = @_;
<a name="l00823"></a>00823   <span class="keywordflow">return</span> [ keys( %{ $self-&gt;{dbas_by_vacc} } ) ];
<a name="l00824"></a>00824 }
<a name="l00825"></a>00825 
<a name="l00826"></a>00826 =head2 get_all_assemblies
<a name="l00827"></a>00827     Description : Return list of all INSDC assembly accessions registered with the helper
<a name="l00828"></a>00828     Exceptions  : None
<a name="l00829"></a>00829     Return type : Arrayref of strings
<a name="l00830"></a>00830 =cut
<a name="l00831"></a>00831 
<a name="l00832"></a>00832 sub get_all_assemblies {
<a name="l00833"></a>00833   my ($self) = @_;
<a name="l00834"></a>00834   <span class="keywordflow">return</span> [ keys( %{ $self-&gt;{dbas_by_gc} } ) ];
<a name="l00835"></a>00835 }
<a name="l00836"></a>00836 
<a name="l00837"></a>00837 =head2 get_all_versioned_assemblies
<a name="l00838"></a>00838     Description : Return list of all versioned INSDC assembly accessions registered with the helper
<a name="l00839"></a>00839     Exceptions  : None
<a name="l00840"></a>00840     Return type : Arrayref of strings
<a name="l00841"></a>00841 =cut
<a name="l00842"></a>00842 
<a name="l00843"></a>00843 sub get_all_versioned_assemblies {
<a name="l00844"></a>00844   my ($self) = @_;
<a name="l00845"></a>00845   <span class="keywordflow">return</span> [ keys( %{ $self-&gt;{dbas_by_vgc} } ) ];
<a name="l00846"></a>00846 }
<a name="l00847"></a>00847 
<a name="l00848"></a>00848 =head2 register_all_dbs
<a name="l00849"></a>00849     Description : Helper method to load the registry with all multispecies core databases on the supplied server
<a name="l00850"></a>00850     Argument    : Host
<a name="l00851"></a>00851     Argument    : Port
<a name="l00852"></a>00852     Argument    : User
<a name="l00853"></a>00853     Argument    : Password
<a name="l00854"></a>00854     Argument    : (optional) String with database regexp (<span class="keywordflow">default</span> is _collection_core_[0-9]_eVersion_[0-9]+)
<a name="l00855"></a>00855     Exceptions  : None
<a name="l00856"></a>00856     Return type : None
<a name="l00857"></a>00857 =cut
<a name="l00858"></a>00858 
<a name="l00859"></a>00859 sub register_all_dbs {
<a name="l00860"></a>00860   my ( $class, $host, $port, $user, $pass, $regexp ) = @_;
<a name="l00861"></a>00861 
<a name="l00862"></a>00862   <span class="keywordflow">if</span> ( !$regexp ) {
<a name="l00863"></a>00863     $regexp = <span class="stringliteral">&#39;_core_[0-9]+_&#39;</span> . software_version() . <span class="stringliteral">&#39;_[0-9]+&#39;</span>;
<a name="l00864"></a>00864   }
<a name="l00865"></a>00865 
<a name="l00866"></a>00866   <span class="keywordflow">if</span> ( !defined $host ) {
<a name="l00867"></a>00867     croak <span class="stringliteral">&quot;Host must be supplied for registration of databases&quot;</span>;
<a name="l00868"></a>00868   }
<a name="l00869"></a>00869 
<a name="l00870"></a>00870   <span class="keywordflow">if</span> ( !defined $port ) {
<a name="l00871"></a>00871     croak
<a name="l00872"></a>00872 <span class="stringliteral">&quot;Port must be supplied for registration of databases on host $host&quot;</span>;
<a name="l00873"></a>00873   }
<a name="l00874"></a>00874 
<a name="l00875"></a>00875   my $str = <span class="stringliteral">&quot;DBI:mysql:host=$host;port=$port&quot;</span>;
<a name="l00876"></a>00876   my $dbh = DBI-&gt;connect( $str, $user, $pass ) ||
<a name="l00877"></a>00877     croak <span class="stringliteral">&quot;Could not connect to database $str (user=$user, pass=$pass)&quot;</span>;
<a name="l00878"></a>00878 
<a name="l00879"></a>00879   my @dbnames =
<a name="l00880"></a>00880     grep { m/$regexp/xi }
<a name="l00881"></a>00881     map  { $_-&gt;[0] } @{ $dbh-&gt;selectall_arrayref(<span class="stringliteral">&#39;SHOW DATABASES&#39;</span>) };
<a name="l00882"></a>00882 
<a name="l00883"></a>00883   <span class="keywordflow">for</span> my $db (@dbnames) {
<a name="l00884"></a>00884     <span class="keywordflow">if</span> ( $db =~ m/.*_collection_core_.*/ ) {
<a name="l00885"></a>00885       _register_multispecies_core( $host, $port, $user, $pass, $db );
<a name="l00886"></a>00886     }
<a name="l00887"></a>00887     <span class="keywordflow">else</span> {
<a name="l00888"></a>00888       _register_singlespecies_core( $host, $port, $user, $pass, $db );
<a name="l00889"></a>00889     }
<a name="l00890"></a>00890   }
<a name="l00891"></a>00891   <span class="keywordflow">return</span>;
<a name="l00892"></a>00892 } ## end sub register_all_dbs
<a name="l00893"></a>00893 
<a name="l00894"></a>00894 =head2 _register_singlespecies_core
<a name="l00895"></a>00895     Description : Register core dbas <span class="keywordflow">for</span> all species in the supplied database
<a name="l00896"></a>00896 =cut
<a name="l00897"></a>00897 
<a name="l00898"></a>00898 sub _register_singlespecies_core {
<a name="l00899"></a>00899   my ( $host, $port, $user, $pass, $db ) = @_;
<a name="l00900"></a>00900   my ( $species, $num ) = (
<a name="l00901"></a>00901     $db =~ /(^[a-z]+_[a-z0-9]+(?:_[a-z0-9]+)?)  # species name
<a name="l00902"></a>00902                      _
<a name="l00903"></a>00903                      core                   # type
<a name="l00904"></a>00904                      _
<a name="l00905"></a>00905                      (?:\d+_)?               # optional endbit <span class="keywordflow">for</span> ensembl genomes
<a name="l00906"></a>00906                      (\d+)                   <span class="preprocessor"># databases release</span>
<a name="l00907"></a>00907 <span class="preprocessor"></span>                     _
<a name="l00908"></a>00908                       /x );
<a name="l00909"></a>00909   <span class="keywordflow">return</span>
<a name="l00910"></a>00910     Bio::EnsEMBL::DBSQL::DBAdaptor-&gt;new( -HOST    =&gt; $host,
<a name="l00911"></a>00911                                          -PORT    =&gt; $port,
<a name="l00912"></a>00912                                          -USER    =&gt; $user,
<a name="l00913"></a>00913                                          -PASS    =&gt; $pass,
<a name="l00914"></a>00914                                          -DBNAME  =&gt; $db,
<a name="l00915"></a>00915                                          -SPECIES =&gt; $species );
<a name="l00916"></a>00916 }
<a name="l00917"></a>00917 
<a name="l00918"></a>00918 =head2 _register_multispecies_core
<a name="l00919"></a>00919     Description : Register core dbas <span class="keywordflow">for</span> all species in the supplied database
<a name="l00920"></a>00920 =cut
<a name="l00921"></a>00921 
<a name="l00922"></a>00922 sub _register_multispecies_core {
<a name="l00923"></a>00923   <span class="keywordflow">return</span> _register_multispecies_x(
<a name="l00924"></a>00924     <span class="stringliteral">&#39;Bio::EnsEMBL::DBSQL::DBAdaptor&#39;</span>,
<a name="l00925"></a>00925     sub {
<a name="l00926"></a>00926       my %args = @_;
<a name="l00927"></a>00927       <span class="keywordflow">return</span> Bio::EnsEMBL::DBSQL::DBAdaptor-&gt;new(%args);
<a name="l00928"></a>00928     },
<a name="l00929"></a>00929     @_ );
<a name="l00930"></a>00930 }
<a name="l00931"></a>00931 
<a name="l00932"></a>00932 =head2 _register_multispecies_x
<a name="l00933"></a>00933     Description : Register specified dba type <span class="keywordflow">for</span> all species in the supplied database
<a name="l00934"></a>00934 =cut
<a name="l00935"></a>00935 
<a name="l00936"></a>00936 sub _register_multispecies_x {
<a name="l00937"></a>00937 
<a name="l00938"></a>00938   my ( $adaptor, $closure, $host, $port, $user, $pass, $instance,
<a name="l00939"></a>00939        $species, $alias_arr_ref, $disconnect_when_idle )
<a name="l00940"></a>00940     = @_;
<a name="l00941"></a>00941   _runtime_include($adaptor);
<a name="l00942"></a>00942   _check_name($instance);
<a name="l00943"></a>00943   my %args =
<a name="l00944"></a>00944     ( _login_hash( $host, $port, $user, $pass ), <span class="stringliteral">&#39;-DBNAME&#39;</span>, $instance );
<a name="l00945"></a>00945   my @species_array = @{ _query_multispecies_db( $species, %args ) };
<a name="l00946"></a>00946   my @dbas;
<a name="l00947"></a>00947   <span class="keywordflow">foreach</span> my $species_args (@species_array) {
<a name="l00948"></a>00948     my %new_args = ( %args, %{$species_args}, <span class="stringliteral">&#39;-MULTISPECIES_DB&#39;</span> =&gt; 1 );
<a name="l00949"></a>00949     $new_args{-DISCONNECT_WHEN_INACTIVE} = 1 <span class="keywordflow">if</span> $disconnect_when_idle;
<a name="l00950"></a>00950     my $dba = $closure-&gt;(%new_args);
<a name="l00951"></a>00951     <span class="keywordflow">if</span> ( defined $species &amp;&amp; defined $alias_arr_ref ) {
<a name="l00952"></a>00952       _add_aliases( $species, $alias_arr_ref );
<a name="l00953"></a>00953     }
<a name="l00954"></a>00954     push( @dbas, $dba );
<a name="l00955"></a>00955   }
<a name="l00956"></a>00956   <span class="keywordflow">return</span> \@dbas;
<a name="l00957"></a>00957 }
<a name="l00958"></a>00958 
<a name="l00959"></a>00959 =head2 _query_multispecies_db
<a name="l00960"></a>00960     Description : Find all species in the multispecies database
<a name="l00961"></a>00961 =cut
<a name="l00962"></a>00962 
<a name="l00963"></a>00963 sub _query_multispecies_db {
<a name="l00964"></a>00964   my ( $species, %db_args ) = @_;
<a name="l00965"></a>00965   my @species_array;
<a name="l00966"></a>00966   eval {
<a name="l00967"></a>00967     my $dbh = Bio::EnsEMBL::DBSQL::DBConnection-&gt;new(%db_args);
<a name="l00968"></a>00968     my $sql;
<a name="l00969"></a>00969     my @params = (<span class="stringliteral">&#39;species.db_name&#39;</span>);
<a name="l00970"></a>00970 
<a name="l00971"></a>00971     <span class="keywordflow">if</span> ( defined $species ) {
<a name="l00972"></a>00972       $sql =
<a name="l00973"></a>00973 <span class="stringliteral">&#39;select meta_value, species_id from meta where meta_key =? and meta_value =?&#39;</span>;
<a name="l00974"></a>00974       push( @params, $species );
<a name="l00975"></a>00975     }
<a name="l00976"></a>00976     <span class="keywordflow">else</span> {
<a name="l00977"></a>00977       $sql =
<a name="l00978"></a>00978 <span class="stringliteral">&#39;select distinct meta_value, species_id from meta where meta_key =? and species_id is not null&#39;</span>;
<a name="l00979"></a>00979     }
<a name="l00980"></a>00980     my $sth = $dbh-&gt;prepare($sql);
<a name="l00981"></a>00981     $sth-&gt;execute(@params);
<a name="l00982"></a>00982     <span class="keywordflow">while</span> ( my $row = $sth-&gt;fetchrow_arrayref() ) {
<a name="l00983"></a>00983       my ( $species_name, $species_id ) = @{$row};
<a name="l00984"></a>00984       push( @species_array,
<a name="l00985"></a>00985             { <span class="stringliteral">&#39;-SPECIES&#39;</span>    =&gt; $species_name,
<a name="l00986"></a>00986               <span class="stringliteral">&#39;-SPECIES_ID&#39;</span> =&gt; $species_id } );
<a name="l00987"></a>00987     }
<a name="l00988"></a>00988     $sth-&gt;finish();
<a name="l00989"></a>00989   };
<a name="l00990"></a>00990   <span class="keywordflow">throw</span>(<span class="stringliteral">&quot;Error detected: $@&quot;</span>) <span class="keywordflow">if</span> $@;
<a name="l00991"></a>00991   <span class="keywordflow">return</span> \@species_array;
<a name="l00992"></a>00992 } ## end sub _query_multispecies_db
<a name="l00993"></a>00993 
<a name="l00994"></a>00994 =head2 _add_aliases
<a name="l00995"></a>00995     Description : Registry all aliases
<a name="l00996"></a>00996     Argument    : Species name
<a name="l00997"></a>00997     Argument    : Arrayref of alias strings
<a name="l00998"></a>00998 =cut
<a name="l00999"></a>00999 
<a name="l01000"></a>01000 sub _add_aliases {
<a name="l01001"></a>01001   my ( $species_name, $alias_ref ) = @_;
<a name="l01002"></a>01002   <span class="keywordflow">if</span> ( defined $alias_ref ) {
<a name="l01003"></a>01003     <span class="keywordflow">if</span> ( ref($alias_ref) ne <span class="stringliteral">&#39;ARRAY&#39;</span> ) {
<a name="l01004"></a>01004       $alias_ref = [$alias_ref];
<a name="l01005"></a>01005     }
<a name="l01006"></a>01006 
<a name="l01007"></a>01007     <span class="keywordflow">if</span> ( scalar( @{$alias_ref} ) &gt; 0 ) {
<a name="l01008"></a>01008       Bio::EnsEMBL::Utils::ConfigRegistry-&gt;add_alias(
<a name="l01009"></a>01009                                               -species =&gt; $species_name,
<a name="l01010"></a>01010                                               -alias   =&gt; $alias_ref );
<a name="l01011"></a>01011     }
<a name="l01012"></a>01012   }
<a name="l01013"></a>01013   <span class="keywordflow">return</span>;
<a name="l01014"></a>01014 }
<a name="l01015"></a>01015 
<a name="l01016"></a>01016 =head2 _runtime_include
<a name="l01017"></a>01017     Description : Load the specified module (usually a DatabaseAdaptor)
<a name="l01018"></a>01018     Argument    : Module name
<a name="l01019"></a>01019 =cut
<a name="l01020"></a>01020 
<a name="l01021"></a>01021 sub _runtime_include {
<a name="l01022"></a>01022   my ($full_module) = @_;
<a name="l01023"></a>01023 <span class="preprocessor">  ## no critic (ProhibitStringyEval)</span>
<a name="l01024"></a>01024 <span class="preprocessor"></span>  eval <span class="stringliteral">&quot;use ${full_module}&quot;</span> ||
<a name="l01025"></a>01025     <span class="keywordflow">throw</span>(<span class="stringliteral">&quot;Cannot import module &#39;${full_module}&#39;: $@&quot;</span>)
<a name="l01026"></a>01026     <span class="keywordflow">if</span> $@;
<a name="l01027"></a>01027 <span class="preprocessor">  ## use critic</span>
<a name="l01028"></a>01028 <span class="preprocessor"></span>  <span class="keywordflow">return</span>;
<a name="l01029"></a>01029 }
<a name="l01030"></a>01030 
<a name="l01031"></a>01031 =head2 _check_name
<a name="l01032"></a>01032     Description : Check that the name is not longer than 64 characters
<a name="l01033"></a>01033     Argument    : Name to check
<a name="l01034"></a>01034 =cut
<a name="l01035"></a>01035 
<a name="l01036"></a>01036 sub _check_name {
<a name="l01037"></a>01037 <span class="preprocessor">  #Throws a wobbly if the db name is longer than 64 characters</span>
<a name="l01038"></a>01038 <span class="preprocessor"></span>  my ($name) = @_;
<a name="l01039"></a>01039   my $length = length($name);
<a name="l01040"></a>01040   <span class="keywordflow">throw</span>(
<a name="l01041"></a>01041 <span class="stringliteral">&quot;Invalid length for database name used. Max is 64. The name ${name}- was ${length}&quot;</span>
<a name="l01042"></a>01042   ) <span class="keywordflow">if</span> $length &gt; 64;
<a name="l01043"></a>01043   <span class="keywordflow">return</span>;
<a name="l01044"></a>01044 }
<a name="l01045"></a>01045 
<a name="l01046"></a>01046 =head2 _login_hash
<a name="l01047"></a>01047     Description : Generate dbadaptor login hash from supplied arguments
<a name="l01048"></a>01048     Argument    : Host
<a name="l01049"></a>01049     Argument    : Port
<a name="l01050"></a>01050     Argument    : User
<a name="l01051"></a>01051     Argument    : Password
<a name="l01052"></a>01052 =cut
<a name="l01053"></a>01053 
<a name="l01054"></a>01054 sub _login_hash {
<a name="l01055"></a>01055   my ( $host, $port, $user, $pass ) = @_;
<a name="l01056"></a>01056   my %details = ( -HOST =&gt; $host, -PORT =&gt; $port, -USER =&gt; $user );
<a name="l01057"></a>01057   $details{-PASS} = $pass <span class="keywordflow">if</span> defined $pass;
<a name="l01058"></a>01058   <span class="keywordflow">return</span> %details;
<a name="l01059"></a>01059 }
<a name="l01060"></a>01060 
<a name="l01061"></a>01061 1;
<a name="l01062"></a>01062 
</pre></div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="LocalLookUp_8pm.html">LocalLookUp.pm</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
