<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Ensembl Metadata: Bio/EnsEMBL/MetaData/Pipeline/ProcessGenome.pm Source File</title>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('ProcessGenome_8pm.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">ProcessGenome.pm</div>  </div>
</div>
<div class="contents">
<a href="ProcessGenome_8pm.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 
<a name="l00002"></a>00002 =head1 LICENSE
<a name="l00003"></a>00003 
<a name="l00004"></a>00004 Copyright [2009-2016] EMBL-European Bioinformatics Institute
<a name="l00005"></a>00005 
<a name="l00006"></a>00006 Licensed under the Apache License, Version 2.0 (the <span class="stringliteral">&quot;License&quot;</span>);
<a name="l00007"></a>00007 you may not use <span class="keyword">this</span> file except in compliance with the License.
<a name="l00008"></a>00008 You may obtain a copy of the License at
<a name="l00009"></a>00009 
<a name="l00010"></a>00010      http:<span class="comment">//www.apache.org/licenses/LICENSE-2.0</span>
<a name="l00011"></a>00011 
<a name="l00012"></a>00012 Unless required by applicable law or agreed to in writing, software
<a name="l00013"></a>00013 distributed under the License is distributed on an <span class="stringliteral">&quot;AS IS&quot;</span> BASIS,
<a name="l00014"></a>00014 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
<a name="l00015"></a><a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1Pipeline_1_1ProcessGenome.html">00015</a> See the License <span class="keywordflow">for</span> the specific language governing permissions and
<a name="l00016"></a>00016 limitations under the License.
<a name="l00017"></a>00017 
<a name="l00018"></a>00018 =head1 CONTACT
<a name="l00019"></a>00019 
<a name="l00020"></a>00020   Please email comments or questions to the <span class="keyword">public</span> Ensembl
<a name="l00021"></a>00021   developers list at &lt;dev@ensembl.org&gt;.
<a name="l00022"></a>00022 
<a name="l00023"></a>00023   Questions may also be sent to the Ensembl help desk at
<a name="l00024"></a>00024   &lt;helpdesk@ensembl.org&gt;.
<a name="l00025"></a>00025 
<a name="l00026"></a>00026 =head1 NAME
<a name="l00027"></a>00027 
<a name="l00028"></a>00028 <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1Pipeline_1_1ProcessGenome.html">Bio::EnsEMBL::MetaData::Pipeline::ProcessGenome</a>
<a name="l00029"></a>00029 
<a name="l00030"></a>00030 =head1 DESCRIPTION
<a name="l00031"></a>00031 
<a name="l00032"></a>00032 Runnable to process a species
<a name="l00033"></a>00033 
<a name="l00034"></a>00034 =head1 AUTHOR
<a name="l00035"></a>00035 
<a name="l00036"></a>00036 Dan Staines
<a name="l00037"></a>00037 
<a name="l00038"></a>00038 =cut
<a name="l00039"></a>00039 
<a name="l00040"></a>00040 use warnings;
<a name="l00041"></a>00041 use strict;
<a name="l00042"></a>00042 
<a name="l00043"></a>00043 <span class="keyword">package </span>Bio::EnsEMBL::MetaData::Pipeline::ProcessGenome;
<a name="l00044"></a>00044 
<a name="l00045"></a>00045 use base qw/Bio::EnsEMBL::Production::Pipeline::Base/;
<a name="l00046"></a>00046 
<a name="l00047"></a>00047 use <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html">Bio::EnsEMBL::MetaData::MetaDataProcessor</a>;
<a name="l00048"></a>00048 use <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1MetaDataDBAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::MetaDataDBAdaptor</a>;
<a name="l00049"></a>00049 use <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html">Bio::EnsEMBL::MetaData::AnnotationAnalyzer</a>;
<a name="l00050"></a>00050 
<a name="l00051"></a>00051 use Carp;
<a name="l00052"></a>00052 use Log::Log4perl qw(:easy);
<a name="l00053"></a>00053 Log::Log4perl-&gt;easy_init($INFO);
<a name="l00054"></a>00054 my $log = get_logger();
<a name="l00055"></a>00055 
<a name="l00056"></a>00056 sub param_defaults {
<a name="l00057"></a>00057   my ($self) = @_;
<a name="l00058"></a>00058   <span class="keywordflow">return</span> {};
<a name="l00059"></a>00059 }
<a name="l00060"></a>00060 
<a name="l00061"></a>00061 sub fetch_input {
<a name="l00062"></a>00062   my ($self) = @_;
<a name="l00063"></a>00063   <span class="keywordflow">return</span>;
<a name="l00064"></a>00064 }
<a name="l00065"></a>00065 
<a name="l00066"></a>00066 sub run {
<a name="l00067"></a>00067   my ($self)  = @_;
<a name="l00068"></a>00068   my $dbas    = {};
<a name="l00069"></a>00069   my $species = $self-&gt;param_required(<span class="stringliteral">&#39;species&#39;</span>);
<a name="l00070"></a>00070 
<a name="l00071"></a>00071   <span class="keywordflow">return</span> <span class="keywordflow">if</span> $species eq <span class="stringliteral">&#39;Ancestral sequences&#39;</span>;
<a name="l00072"></a>00072 
<a name="l00073"></a>00073   $log-&gt;info(<span class="stringliteral">&quot;Finding DBAdaptors for $species&quot;</span>);
<a name="l00074"></a>00074   <span class="keywordflow">for</span> my $type (qw/core variation otherfeatures funcgen/) {
<a name="l00075"></a>00075     $dbas-&gt;{$type} = $self-&gt;get_DBAdaptor($type);
<a name="l00076"></a>00076   }
<a name="l00077"></a>00077   $log-&gt;info(<span class="stringliteral">&quot;Connecting to info database&quot;</span>);
<a name="l00078"></a>00078   my $dba =
<a name="l00079"></a>00079     <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1MetaDataDBAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::MetaDataDBAdaptor</a>-&gt;new(
<a name="l00080"></a>00080                                              -USER =&gt;,
<a name="l00081"></a>00081                                              $self-&gt;param(<span class="stringliteral">&#39;info_user&#39;</span>),
<a name="l00082"></a>00082                                              -PASS =&gt;,
<a name="l00083"></a>00083                                              $self-&gt;param(<span class="stringliteral">&#39;info_pass&#39;</span>),
<a name="l00084"></a>00084                                              -HOST =&gt;,
<a name="l00085"></a>00085                                              $self-&gt;param(<span class="stringliteral">&#39;info_host&#39;</span>),
<a name="l00086"></a>00086                                              -PORT =&gt;,
<a name="l00087"></a>00087                                              $self-&gt;param(<span class="stringliteral">&#39;info_port&#39;</span>),
<a name="l00088"></a>00088                                              -DBNAME =&gt;,
<a name="l00089"></a>00089                                              $self-&gt;param(<span class="stringliteral">&#39;info_dbname&#39;</span>)
<a name="l00090"></a>00090     );
<a name="l00091"></a>00091   my $gdba = $dba-&gt;get_GenomeInfoAdaptor();
<a name="l00092"></a>00092 
<a name="l00093"></a>00093 <span class="preprocessor">  # set the release to use when storing genomes</span>
<a name="l00094"></a>00094 <span class="preprocessor"></span>  my $rdba = $dba-&gt;get_DataReleaseInfoAdaptor();
<a name="l00095"></a>00095   my $release;
<a name="l00096"></a>00096   <span class="keywordflow">if</span> ( defined $self-&gt;param(<span class="stringliteral">&#39;eg_release&#39;</span>) ) {
<a name="l00097"></a>00097     $release =
<a name="l00098"></a>00098       $rdba-&gt;fetch_by_ensembl_genomes_release(
<a name="l00099"></a>00099                                            $self-&gt;param(<span class="stringliteral">&#39;eg_release&#39;</span>) );
<a name="l00100"></a>00100   }
<a name="l00101"></a>00101   <span class="keywordflow">else</span> {
<a name="l00102"></a>00102     $release =
<a name="l00103"></a>00103       $rdba-&gt;fetch_by_ensembl_release( $self-&gt;param(<span class="stringliteral">&#39;release&#39;</span>) );
<a name="l00104"></a>00104   }
<a name="l00105"></a>00105   $gdba-&gt;data_release($release);
<a name="l00106"></a>00106 
<a name="l00107"></a>00107   my $upd = $self-&gt;param(<span class="stringliteral">&#39;force_update&#39;</span>) || 0;
<a name="l00108"></a>00108 
<a name="l00109"></a>00109   my $opts = { -INFO_ADAPTOR =&gt; $gdba,
<a name="l00110"></a>00110                -ANNOTATION_ANALYZER =&gt;
<a name="l00111"></a>00111                  <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html">Bio::EnsEMBL::MetaData::AnnotationAnalyzer</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1AnnotationAnalyzer.html#a5f4168a9b1eaea0d3834ced9f9949cc8">new</a>(),
<a name="l00112"></a>00112                -COMPARA      =&gt; 0,
<a name="l00113"></a>00113                -CONTIGS      =&gt; $self-&gt;param(<span class="stringliteral">&#39;contigs&#39;</span>) || 1,
<a name="l00114"></a>00114                -FORCE_UPDATE =&gt; $upd,
<a name="l00115"></a>00115                -VARIATION    =&gt; $self-&gt;param(<span class="stringliteral">&#39;variation&#39;</span>) || 1 };
<a name="l00116"></a>00116 
<a name="l00117"></a>00117   my $processor =
<a name="l00118"></a>00118     <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html">Bio::EnsEMBL::MetaData::MetaDataProcessor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1MetaDataProcessor.html#a32c63ed55ff1ef0c84687294a4380ecd">new</a>(%$opts);
<a name="l00119"></a>00119 
<a name="l00120"></a>00120   $log-&gt;info(<span class="stringliteral">&quot;Processing $species&quot;</span>);
<a name="l00121"></a>00121 
<a name="l00122"></a>00122   my $md = $processor-&gt;process_genome($dbas);
<a name="l00123"></a>00123 
<a name="l00124"></a>00124   <span class="keywordflow">if</span> ( defined $md-&gt;dbID() &amp;&amp; $upd == 1 ) {
<a name="l00125"></a>00125     $log-&gt;info( <span class="stringliteral">&quot;Updating &quot;</span> . $md-&gt;name() );
<a name="l00126"></a>00126     $gdba-&gt;update($md);
<a name="l00127"></a>00127   }
<a name="l00128"></a>00128   elsif ( !defined $md-&gt;dbID() ) {
<a name="l00129"></a>00129     $log-&gt;info( <span class="stringliteral">&quot;Storing &quot;</span> . $md-&gt;name() );
<a name="l00130"></a>00130     $gdba-&gt;store($md);
<a name="l00131"></a>00131   }
<a name="l00132"></a>00132 
<a name="l00133"></a>00133   $log-&gt;info(<span class="stringliteral">&quot;Completed processing $species&quot;</span>);
<a name="l00134"></a>00134   <span class="keywordflow">return</span>;
<a name="l00135"></a>00135 } ## end sub run
<a name="l00136"></a>00136 
<a name="l00137"></a>00137 sub write_output {
<a name="l00138"></a>00138   my ($self) = @_;
<a name="l00139"></a>00139   <span class="keywordflow">return</span>;
<a name="l00140"></a>00140 }
<a name="l00141"></a>00141 
<a name="l00142"></a>00142 1;
<a name="l00143"></a>00143 
</pre></div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="ProcessGenome_8pm.html">ProcessGenome.pm</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
