<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<title>Ensembl Metadata: Bio/EnsEMBL/LookUp/RemoteLookUp.pm Source File</title>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
</head>
<body>
<div id="top"><!-- do not remove this div! -->

<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr>
  
  
  
  
  
 </tr>
 </tbody>
</table>
</div>

<!-- Generated by Doxygen 1.7.4 -->
<script type="text/javascript"><!--
var searchBox = new SearchBox("searchBox", "search",false,'Search');
--></script>
<script type="text/javascript">
function hasClass(ele,cls) {
  return ele.className.match(new RegExp('(\\s|^)'+cls+'(\\s|$)'));
}

function addClass(ele,cls) {
  if (!this.hasClass(ele,cls)) ele.className += " "+cls;
}

function removeClass(ele,cls) {
  if (hasClass(ele,cls)) {
    var reg = new RegExp('(\\s|^)'+cls+'(\\s|$)');
    ele.className=ele.className.replace(reg,' ');
  }
}

function toggleVisibility(linkObj) {
 var base = linkObj.getAttribute('id');
 var summary = document.getElementById(base + '-summary');
 var content = document.getElementById(base + '-content');
 var trigger = document.getElementById(base + '-trigger');
 if ( hasClass(linkObj,'closed') ) {
   summary.style.display = 'none';
   content.style.display = 'block';
   trigger.src = 'open.png';
   removeClass(linkObj,'closed');
   addClass(linkObj,'opened');
 } else if ( hasClass(linkObj,'opened') ) {
   summary.style.display = 'block';
   content.style.display = 'none';
   trigger.src = 'closed.png';
   removeClass(linkObj,'opened');
   addClass(linkObj,'closed');
 }
 return false;
}
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Overview</span></a></li>
      <li><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="namespaces.html"><span>Namespaces</span></a></li>
      <li class="current"><a href="files.html"><span>Files</span></a></li>
      <li><a href="dirs.html"><span>Directory&#160;Browser</span></a></li>
      <li id="searchli">
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="files.html"><span>File&#160;List</span></a></li>
    </ul>
  </div>
</div>
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
  initNavTree('RemoteLookUp_8pm.html','');
</script>
<div id="doc-content">
<div class="header">
  <div class="headertitle">
<div class="title">RemoteLookUp.pm</div>  </div>
</div>
<div class="contents">
<a href="RemoteLookUp_8pm.html">Go to the documentation of this file.</a><div class="fragment"><pre class="fragment"><a name="l00001"></a>00001 
<a name="l00002"></a>00002 =head1 LICENSE
<a name="l00003"></a>00003 
<a name="l00004"></a>00004 Copyright [2009-2014] EMBL-European Bioinformatics Institute
<a name="l00005"></a>00005 
<a name="l00006"></a>00006 Licensed under the Apache License, Version 2.0 (the <span class="stringliteral">&quot;License&quot;</span>);
<a name="l00007"></a>00007 you may not use <span class="keyword">this</span> file except in compliance with the License.
<a name="l00008"></a>00008 You may obtain a copy of the License at
<a name="l00009"></a>00009 
<a name="l00010"></a>00010      http:<span class="comment">//www.apache.org/licenses/LICENSE-2.0</span>
<a name="l00011"></a>00011 
<a name="l00012"></a>00012 Unless required by applicable law or agreed to in writing, software
<a name="l00013"></a>00013 distributed under the License is distributed on an <span class="stringliteral">&quot;AS IS&quot;</span> BASIS,
<a name="l00014"></a>00014 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
<a name="l00015"></a>00015 See the License <span class="keywordflow">for</span> the specific language governing permissions and
<a name="l00016"></a>00016 limitations under the License.
<a name="l00017"></a>00017 
<a name="l00018"></a>00018 =cut
<a name="l00019"></a>00019 
<a name="l00020"></a>00020 =pod
<a name="l00021"></a>00021 
<a name="l00022"></a>00022 =head1 CONTACT
<a name="l00023"></a>00023 
<a name="l00024"></a>00024   Please email comments or questions to the <span class="keyword">public</span> Ensembl
<a name="l00025"></a>00025   developers list at &lt;dev@ensembl.org&gt;.
<a name="l00026"></a>00026 
<a name="l00027"></a>00027   Questions may also be sent to the Ensembl help desk at
<a name="l00028"></a>00028   &lt;helpdesk@ensembl.org&gt;.
<a name="l00029"></a>00029 
<a name="l00030"></a>00030 =head1 NAME
<a name="l00031"></a>00031 
<a name="l00032"></a>00032 <a class="code" href="classBio_1_1EnsEMBL_1_1LookUp.html">Bio::EnsEMBL::LookUp</a>
<a name="l00033"></a>00033 
<a name="l00034"></a>00034 =head1 SYNOPSIS
<a name="l00035"></a>00035 
<a name="l00036"></a>00036 my $adaptor = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa892bb1501e0bd0bbea154ed46dde3bc">build_ensembl_genomes_adaptor</a>();
<a name="l00037"></a>00037 my $lookup = Bio::EnsEMBL::RemoteLookUp-&gt;new(-ADAPTOR=&gt;$adaptor);
<a name="l00038"></a>00038 my $dbas = $lookup-&gt;registry()-&gt;get_all();
<a name="l00039"></a><a class="code" href="classBio_1_1EnsEMBL_1_1LookUp_1_1RemoteLookUp.html">00039</a> $dbas = $lookup-&gt;get_all_by_taxon_id(388919);
<a name="l00040"></a>00040 $dbas = $lookup-&gt;get_by_name_pattern(<span class="stringliteral">&quot;Escherichia.*&quot;</span>);
<a name="l00041"></a>00041 
<a name="l00042"></a>00042 =head1 DESCRIPTION
<a name="l00043"></a>00043 
<a name="l00044"></a>00044 This module is an implementation of <a class="code" href="classBio_1_1EnsEMBL_1_1LookUp.html">Bio::EnsEMBL::LookUp</a> that uses <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a> to
<a name="l00045"></a>00045 access a MySQL database containing information about Ensembl and Ensembl Genomes contents and then instantiate DBAdaptors.
<a name="l00046"></a>00046 
<a name="l00047"></a>00047 To instantiate <span class="keyword">using</span> the <span class="keyword">public</span> Ensembl/EG servers <span class="keywordflow">for</span> creating DBAdaptors:
<a name="l00048"></a>00048 
<a name="l00049"></a>00049     my $lookup = Bio::EnsEMBL::LookUp::RemoteLookUp-&gt;new(-ADAPTOR=&gt;$adaptor);
<a name="l00050"></a>00050 
<a name="l00051"></a>00051 To instantiate to use a specific server on which core databases are located:
<a name="l00052"></a>00052 
<a name="l00053"></a>00053     my $lookup = Bio::EnsEMBL::LookUp::RemoteLookUp-&gt;new(-USER=&gt;$user, -HOST=&gt;$host, -PORT=&gt;$port, -ADAPTOR=&gt;$adaptor);
<a name="l00054"></a>00054 
<a name="l00055"></a>00055 Once constructed, the <a class="code" href="classBio_1_1EnsEMBL_1_1LookUp.html">LookUp</a> instance can be used as documented in <a class="code" href="classBio_1_1EnsEMBL_1_1LookUp.html">Bio::EnsEMBL::LookUp</a>.
<a name="l00056"></a>00056   
<a name="l00057"></a>00057 =head1 AUTHOR
<a name="l00058"></a>00058 
<a name="l00059"></a>00059 dstaines
<a name="l00060"></a>00060 
<a name="l00061"></a>00061 =head1 MAINTANER
<a name="l00062"></a>00062 
<a name="l00063"></a>00063 $Author$
<a name="l00064"></a>00064 
<a name="l00065"></a>00065 =head1 VERSION
<a name="l00066"></a>00066 
<a name="l00067"></a>00067 $Revision$
<a name="l00068"></a>00068 
<a name="l00069"></a>00069 =cut
<a name="l00070"></a>00070 
<a name="l00071"></a>00071 <span class="keyword">package </span>Bio::EnsEMBL::LookUp::RemoteLookUp;
<a name="l00072"></a>00072 
<a name="l00073"></a>00073 use warnings;
<a name="l00074"></a>00074 use strict;
<a name="l00075"></a>00075 use Bio::EnsEMBL::DBSQL::DBAdaptor;
<a name="l00076"></a>00076 use Bio::EnsEMBL::Compara::DBSQL::DBAdaptor;
<a name="l00077"></a>00077 use Bio::EnsEMBL::Utils::Argument qw(rearrange);
<a name="l00078"></a>00078 use Bio::EnsEMBL::Utils::Exception qw(<span class="keywordflow">throw</span> warning);
<a name="l00079"></a>00079 use Bio::EnsEMBL::Utils::Scalar qw(assert_ref check_ref);
<a name="l00080"></a>00080 use <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>;
<a name="l00081"></a>00081 use <a class="code" href="classBio_1_1EnsEMBL_1_1Utils_1_1PublicMySQLServer.html">Bio::EnsEMBL::Utils::PublicMySQLServer</a> qw/e_args eg_args/;
<a name="l00082"></a>00082 use Carp;
<a name="l00083"></a>00083 use List::MoreUtils qw(uniq);
<a name="l00084"></a>00084 
<a name="l00085"></a>00085 =head1 SUBROUTINES/METHODS
<a name="l00086"></a>00086 
<a name="l00087"></a>00087 =head2 <span class="keyword">new</span>
<a name="l00088"></a>00088 
<a name="l00089"></a>00089   Description       : Creates a <span class="keyword">new</span> instance of <span class="keyword">this</span> <span class="keywordtype">object</span>. 
<a name="l00090"></a>00090   Returntype        : Instance of lookup
<a name="l00091"></a>00091   Status            : Stable
<a name="l00092"></a>00092 
<a name="l00093"></a>00093   Example           : 
<a name="l00094"></a>00094   my $lookup = Bio::EnsEMBL::RemoteLookUp-&gt;new();
<a name="l00095"></a>00095 =cut
<a name="l00096"></a>00096 
<a name="l00097"></a>00097 sub <span class="keyword">new</span> {
<a name="l00098"></a>00098   my ( $class, @args ) = @_;
<a name="l00099"></a>00099   my $self = bless( {}, ref($class) || $class );
<a name="l00100"></a>00100   ( $self-&gt;{_adaptor}, $self-&gt;{registry}, $self-&gt;{user},
<a name="l00101"></a>00101     $self-&gt;{pass},     $self-&gt;{host},     $self-&gt;{port} )
<a name="l00102"></a>00102     = rearrange( [ <span class="stringliteral">&#39;ADAPTOR&#39;</span>, <span class="stringliteral">&#39;REGISTRY&#39;</span>, <span class="stringliteral">&#39;USER&#39;</span>, <span class="stringliteral">&#39;PASS&#39;</span>, <span class="stringliteral">&#39;HOST&#39;</span>, <span class="stringliteral">&#39;PORT&#39;</span> ],
<a name="l00103"></a>00103                  @args );
<a name="l00104"></a>00104   $self-&gt;{dba_cache} = {};
<a name="l00105"></a>00105   $self-&gt;{registry} ||= q/Bio::EnsEMBL::Registry/;
<a name="l00106"></a>00106   <span class="keywordflow">return</span> $self;
<a name="l00107"></a>00107 }
<a name="l00108"></a>00108 
<a name="l00109"></a>00109 sub _cache {
<a name="l00110"></a>00110   my ($self) = @_;
<a name="l00111"></a>00111   <span class="keywordflow">return</span> $self-&gt;{dba_cache};
<a name="l00112"></a>00112 }
<a name="l00113"></a>00113 
<a name="l00114"></a>00114 sub _adaptor {
<a name="l00115"></a>00115   my ($self) = @_;
<a name="l00116"></a>00116   <span class="keywordflow">if</span>(!defined $self-&gt;{_adaptor}) {
<a name="l00117"></a>00117 <span class="preprocessor">    # default to previous behaviour</span>
<a name="l00118"></a>00118 <span class="preprocessor"></span>    $self-&gt;{_adaptor} = <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html">Bio::EnsEMBL::MetaData::DBSQL::GenomeInfoAdaptor</a>-&gt;<a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1DBSQL_1_1GenomeInfoAdaptor.html#aa892bb1501e0bd0bbea154ed46dde3bc">build_ensembl_genomes_adaptor</a>();
<a name="l00119"></a>00119   }
<a name="l00120"></a>00120   <span class="keywordflow">return</span> $self-&gt;{_adaptor};
<a name="l00121"></a>00121 }
<a name="l00122"></a>00122 
<a name="l00123"></a>00123 sub _get_args {
<a name="l00124"></a>00124   my ( $self, $genome_info ) = @_;
<a name="l00125"></a>00125 <span class="preprocessor">  # find arguments applicable to this particular info object</span>
<a name="l00126"></a>00126 <span class="preprocessor"></span>  my $args = {};
<a name="l00127"></a>00127   <span class="keywordflow">if</span> ( !defined $self-&gt;{host} ) {
<a name="l00128"></a>00128 <span class="preprocessor">    # no host specified, so try and use public servers</span>
<a name="l00129"></a>00129 <span class="preprocessor"></span>    <span class="keywordflow">if</span> ( defined $genome_info-&gt;data_release()-&gt;ensembl_genomes_release() ) {
<a name="l00130"></a>00130 <span class="preprocessor">      # use EG</span>
<a name="l00131"></a>00131 <span class="preprocessor"></span>      $args = eg_args();
<a name="l00132"></a>00132     }
<a name="l00133"></a>00133     <span class="keywordflow">else</span> {
<a name="l00134"></a>00134 <span class="preprocessor">      # use Ensembl</span>
<a name="l00135"></a>00135 <span class="preprocessor"></span>      $args = e_args();
<a name="l00136"></a>00136     }
<a name="l00137"></a>00137   }
<a name="l00138"></a>00138   <span class="keywordflow">else</span> {
<a name="l00139"></a>00139 <span class="preprocessor">    # use host specified</span>
<a name="l00140"></a>00140 <span class="preprocessor"></span>    $args = { -USER =&gt; $self-&gt;{user},
<a name="l00141"></a>00141               -PORT =&gt; $self-&gt;{port},
<a name="l00142"></a>00142               -PASS =&gt; $self-&gt;{pass},
<a name="l00143"></a>00143               -HOST =&gt; $self-&gt;{host} };
<a name="l00144"></a>00144   }
<a name="l00145"></a>00145   <span class="keywordflow">return</span> $args;
<a name="l00146"></a>00146 }
<a name="l00147"></a>00147 
<a name="l00148"></a>00148 =head2 genome_to_dba
<a name="l00149"></a>00149     Description : Build a Bio::EnsEMBL::DBSQL::DBAdaptor instance with the supplied info <span class="keywordtype">object</span>
<a name="l00150"></a>00150     Argument    : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00151"></a>00151     Exceptions  : None
<a name="l00152"></a>00152     Return type : Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00153"></a>00153 =cut
<a name="l00154"></a>00154 
<a name="l00155"></a>00155 sub genome_to_dba {
<a name="l00156"></a>00156   my ( $self, $genome_info ) = @_;
<a name="l00157"></a>00157   my $dba;
<a name="l00158"></a>00158   <span class="keywordflow">if</span> ( defined $genome_info ) {
<a name="l00159"></a>00159     assert_ref( $genome_info, <span class="stringliteral">&#39;Bio::EnsEMBL::MetaData::GenomeInfo&#39;</span> );
<a name="l00160"></a>00160     $dba = $self-&gt;_cache()-&gt;{ $genome_info-&gt;dbID() };
<a name="l00161"></a>00161     <span class="keywordflow">if</span> ( !defined $dba ) {
<a name="l00162"></a>00162 
<a name="l00163"></a>00163       my $args = $self-&gt;_get_args($genome_info);
<a name="l00164"></a>00164 
<a name="l00165"></a>00165       $args-&gt;{-DBNAME}     = $genome_info-&gt;dbname();
<a name="l00166"></a>00166       $args-&gt;{-SPECIES}    = $genome_info-&gt;name();
<a name="l00167"></a>00167       $args-&gt;{-SPECIES_ID} = $genome_info-&gt;species_id();
<a name="l00168"></a>00168       $args-&gt;{-MULTISPECIES_DB} =
<a name="l00169"></a>00169         $genome_info-&gt;dbname() =~ m/_collection_/ ? 1 : 0;
<a name="l00170"></a>00170       $args-&gt;{-GROUP} = <span class="stringliteral">&#39;core&#39;</span>;
<a name="l00171"></a>00171 
<a name="l00172"></a>00172       $dba = Bio::EnsEMBL::DBSQL::DBAdaptor-&gt;new(%$args);
<a name="l00173"></a>00173 
<a name="l00174"></a>00174       $self-&gt;_cache()-&gt;{ $genome_info-&gt;dbID() } = $dba;
<a name="l00175"></a>00175 
<a name="l00176"></a>00176     }
<a name="l00177"></a>00177   }
<a name="l00178"></a>00178   <span class="keywordflow">return</span> $dba;
<a name="l00179"></a>00179 } ## end sub genome_to_dba
<a name="l00180"></a>00180 
<a name="l00181"></a>00181 =head2 genomes_to_dbas
<a name="l00182"></a>00182     Description : Build a <span class="keyword">set</span> of Bio::EnsEMBL::DBSQL::DBAdaptor instances with the supplied info objects
<a name="l00183"></a>00183     Argument    : array ref of <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeInfo.html">Bio::EnsEMBL::MetaData::GenomeInfo</a>
<a name="l00184"></a>00184     Exceptions  : None
<a name="l00185"></a>00185     Return type : array ref of Bio::EnsEMBL::DBSQL::DBAdaptor
<a name="l00186"></a>00186 =cut
<a name="l00187"></a>00187 
<a name="l00188"></a>00188 sub genomes_to_dbas {
<a name="l00189"></a>00189   my ( $self, $genomes ) = @_;
<a name="l00190"></a>00190   my $dbas = [];
<a name="l00191"></a>00191   <span class="keywordflow">if</span> ( defined $genomes ) {
<a name="l00192"></a>00192     <span class="keywordflow">for</span> my $genome ( @{$genomes} ) {
<a name="l00193"></a>00193       push @$dbas, $self-&gt;genome_to_dba($genome);
<a name="l00194"></a>00194     }
<a name="l00195"></a>00195   }
<a name="l00196"></a>00196   <span class="keywordflow">return</span> $dbas;
<a name="l00197"></a>00197 }
<a name="l00198"></a>00198 
<a name="l00199"></a>00199 =head2 compara_to_dba
<a name="l00200"></a>00200     Description : Build a Bio::EnsEMBL::Compara::DBSQL::DBAdaptor instance with the supplied info <span class="keywordtype">object</span>
<a name="l00201"></a>00201     Argument    : <a class="code" href="classBio_1_1EnsEMBL_1_1MetaData_1_1GenomeComparaInfo.html">Bio::EnsEMBL::MetaData::GenomeComparaInfo</a>
<a name="l00202"></a>00202     Exceptions  : None
<a name="l00203"></a>00203     Return type : Arrayref of strings
<a name="l00204"></a>00204 =cut
<a name="l00205"></a>00205 
<a name="l00206"></a>00206 sub compara_to_dba {
<a name="l00207"></a>00207   my ( $self, $genome_info ) = @_;
<a name="l00208"></a>00208   assert_ref( $genome_info, <span class="stringliteral">&#39;Bio::EnsEMBL::MetaData::GenomeComparaInfo&#39;</span> );
<a name="l00209"></a>00209   my $dba = $self-&gt;_cache()-&gt;{compara}-&gt;{ $genome_info-&gt;dbID() };
<a name="l00210"></a>00210   <span class="keywordflow">if</span> ( !defined $dba ) {
<a name="l00211"></a>00211     my $div = $genome_info-&gt;division();
<a name="l00212"></a>00212     <span class="keywordflow">if</span> ( !$div eq <span class="stringliteral">&#39;Ensembl&#39;</span> ) {
<a name="l00213"></a>00213       $div = <span class="stringliteral">&#39;multi&#39;</span>;
<a name="l00214"></a>00214     }
<a name="l00215"></a>00215 
<a name="l00216"></a>00216     my $args = $self-&gt;_get_args($genome_info);
<a name="l00217"></a>00217     $args-&gt;{-DBNAME}  = $genome_info-&gt;dbname();
<a name="l00218"></a>00218     $args-&gt;{-SPECIES} = $div;
<a name="l00219"></a>00219     $args-&gt;{-GROUP}   = <span class="stringliteral">&#39;compara&#39;</span>;
<a name="l00220"></a>00220     $dba              = Bio::EnsEMBL::Compara::DBSQL::DBAdaptor-&gt;new(%$args);
<a name="l00221"></a>00221 
<a name="l00222"></a>00222     $self-&gt;_cache()-&gt;{compara}-&gt;{ $genome_info-&gt;dbID() } = $dba;
<a name="l00223"></a>00223 
<a name="l00224"></a>00224   }
<a name="l00225"></a>00225   <span class="keywordflow">return</span> $dba;
<a name="l00226"></a>00226 }
<a name="l00227"></a>00227 
<a name="l00228"></a>00228 =head2 get_all_dbnames
<a name="l00229"></a>00229     Description : Return all database names used by the DBAs retrieved from the registry
<a name="l00230"></a>00230     Argument    : None
<a name="l00231"></a>00231     Exceptions  : None
<a name="l00232"></a>00232     Return type : Arrayref of strings
<a name="l00233"></a>00233 =cut
<a name="l00234"></a>00234 
<a name="l00235"></a>00235 sub get_all_dbnames {
<a name="l00236"></a>00236   my ($self) = @_;
<a name="l00237"></a>00237   <span class="keywordflow">return</span> [ uniq( map { $_-&gt;dbname() } @{ $self-&gt;{_adaptor} } ) ];
<a name="l00238"></a>00238 }
<a name="l00239"></a>00239 
<a name="l00240"></a>00240 =head2 get_all
<a name="l00241"></a>00241     Description : Return all database adaptors that have been retrieved from registry
<a name="l00242"></a>00242     Argument    : None
<a name="l00243"></a>00243     Exceptions  : None
<a name="l00244"></a>00244     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00245"></a>00245 =cut
<a name="l00246"></a>00246 
<a name="l00247"></a>00247 sub get_all {
<a name="l00248"></a>00248   my ($self) = @_;
<a name="l00249"></a>00249   <span class="keywordflow">return</span> $self-&gt;genomes_to_dbas( $self-&gt;{_adaptor}-&gt;fetch_all() );
<a name="l00250"></a>00250 }
<a name="l00251"></a>00251 
<a name="l00252"></a>00252 sub get_all_by_taxon_branch {
<a name="l00253"></a>00253   my ( $self, $taxid ) = @_;
<a name="l00254"></a>00254   <span class="keywordflow">return</span> $self-&gt;genomes_to_dbas(
<a name="l00255"></a>00255                       $self-&gt;{_adaptor}-&gt;fetch_all_by_taxonomy_branch($taxid) );
<a name="l00256"></a>00256 }
<a name="l00257"></a>00257 
<a name="l00258"></a>00258 =head2 get_all_by_taxon_id
<a name="l00259"></a>00259     Description : Returns all database adaptors that have the supplied taxonomy ID
<a name="l00260"></a>00260     Argument    : Int
<a name="l00261"></a>00261     Exceptions  : None
<a name="l00262"></a>00262     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00263"></a>00263 =cut
<a name="l00264"></a>00264 
<a name="l00265"></a>00265 sub get_all_by_taxon_id {
<a name="l00266"></a>00266   my ( $self, $taxid ) = @_;
<a name="l00267"></a>00267   <span class="keywordflow">return</span> $self-&gt;genomes_to_dbas(
<a name="l00268"></a>00268                           $self-&gt;{_adaptor}-&gt;fetch_all_by_taxonomy_id($taxid) );
<a name="l00269"></a>00269 }
<a name="l00270"></a>00270 
<a name="l00271"></a>00271 =head2 get_by_name_exact
<a name="l00272"></a>00272     Description : Return database adaptor that has the supplied <span class="keywordtype">string</span> as an alias/name
<a name="l00273"></a>00273     Argument    : String
<a name="l00274"></a>00274     Exceptions  : None
<a name="l00275"></a>00275     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00276"></a>00276 =cut
<a name="l00277"></a>00277 
<a name="l00278"></a>00278 sub get_by_name_exact {
<a name="l00279"></a>00279   my ( $self, $name ) = @_;
<a name="l00280"></a>00280   <span class="keywordflow">return</span> $self-&gt;genome_to_dba( $self-&gt;{_adaptor}-&gt;fetch_by_any_name($name) );
<a name="l00281"></a>00281 }
<a name="l00282"></a>00282 
<a name="l00283"></a>00283 =head2 get_all_by_accession
<a name="l00284"></a>00284     Description : Returns the database adaptor(s) that contains a seq_region with the supplied INSDC accession (or other seq_region name)
<a name="l00285"></a>00285     Argument    : Int
<a name="l00286"></a>00286     Exceptions  : None
<a name="l00287"></a>00287     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00288"></a>00288 =cut    
<a name="l00289"></a>00289 
<a name="l00290"></a>00290 sub get_all_by_accession {
<a name="l00291"></a>00291   my ( $self, $acc ) = @_;
<a name="l00292"></a>00292   my $genomes = $self-&gt;{_adaptor}-&gt;fetch_all_by_sequence_accession($acc);
<a name="l00293"></a>00293   <span class="keywordflow">if</span> ( !defined $genomes || scalar(@$genomes) == 0 ) {
<a name="l00294"></a>00294     $genomes =
<a name="l00295"></a>00295       $self-&gt;{_adaptor}-&gt;fetch_all_by_sequence_accession_unversioned($acc);
<a name="l00296"></a>00296   }
<a name="l00297"></a>00297   <span class="keywordflow">return</span> $self-&gt;genomes_to_dbas($genomes);
<a name="l00298"></a>00298 }
<a name="l00299"></a>00299 
<a name="l00300"></a>00300 =head2 get_by_assembly_accession
<a name="l00301"></a>00301     Description : Returns the database adaptor that contains the assembly with the supplied INSDC assembly accession
<a name="l00302"></a>00302     Argument    : Int
<a name="l00303"></a>00303     Exceptions  : None
<a name="l00304"></a>00304     Return type : Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00305"></a>00305 =cut
<a name="l00306"></a>00306 
<a name="l00307"></a>00307 sub get_by_assembly_accession {
<a name="l00308"></a>00308   my ( $self, $acc ) = @_;
<a name="l00309"></a>00309   my $genome = $self-&gt;{_adaptor}-&gt;fetch_by_assembly_id($acc);
<a name="l00310"></a>00310   <span class="keywordflow">if</span> ( !defined $genome ) {
<a name="l00311"></a>00311     $genome = $self-&gt;{_adaptor}-&gt;fetch_by_assembly_id_unversioned($acc);
<a name="l00312"></a>00312   }
<a name="l00313"></a>00313   <span class="keywordflow">return</span> $self-&gt;genome_to_dba($genome);
<a name="l00314"></a>00314 }
<a name="l00315"></a>00315 
<a name="l00316"></a>00316 =head2 get_all_by_name_pattern
<a name="l00317"></a>00317     Description : Return all database adaptors that have an alias/name that match the supplied regexp
<a name="l00318"></a>00318     Argument    : String
<a name="l00319"></a>00319     Exceptions  : None
<a name="l00320"></a>00320     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00321"></a>00321 =cut    
<a name="l00322"></a>00322 
<a name="l00323"></a>00323 sub get_all_by_name_pattern {
<a name="l00324"></a>00324   my ( $self, $name ) = @_;
<a name="l00325"></a>00325   <span class="keywordflow">return</span> $self-&gt;genomes_to_dbas(
<a name="l00326"></a>00326                           $self-&gt;{_adaptor}-&gt;fetch_all_by_name_pattern($name) );
<a name="l00327"></a>00327 }
<a name="l00328"></a>00328 
<a name="l00329"></a>00329 =head2 get_all_by_dbname
<a name="l00330"></a>00330     Description : Returns all database adaptors that have the supplied dbname
<a name="l00331"></a>00331     Argument    : String
<a name="l00332"></a>00332     Exceptions  : None
<a name="l00333"></a>00333     Return type : Arrayref of Bio::EnsEMBL::DBSQL::DatabaseAdaptor
<a name="l00334"></a>00334 =cut
<a name="l00335"></a>00335 
<a name="l00336"></a>00336 sub get_all_by_dbname {
<a name="l00337"></a>00337   my ( $self, $name ) = @_;
<a name="l00338"></a>00338   <span class="keywordflow">return</span> $self-&gt;genomes_to_dbas($self-&gt;{_adaptor}-&gt;fetch_all_by_dbname($name) );
<a name="l00339"></a>00339 }
<a name="l00340"></a>00340 
<a name="l00341"></a>00341 =head2 get_all_taxon_ids
<a name="l00342"></a>00342     Description : Return list of all taxon IDs registered with the helper
<a name="l00343"></a>00343     Exceptions  : None
<a name="l00344"></a>00344     Return type : Arrayref of integers
<a name="l00345"></a>00345 =cut
<a name="l00346"></a>00346 
<a name="l00347"></a>00347 sub get_all_taxon_ids {
<a name="l00348"></a>00348   my ($self) = @_;
<a name="l00349"></a>00349   <span class="keywordflow">return</span> [
<a name="l00350"></a>00350         uniq( map { $_-&gt;taxonomy_id() } @{ $self-&gt;{_adaptor}-&gt;fetch_all() } ) ];
<a name="l00351"></a>00351 }
<a name="l00352"></a>00352 
<a name="l00353"></a>00353 =head2 get_all_names
<a name="l00354"></a>00354     Description : Return list of all species names registered with the helper
<a name="l00355"></a>00355     Exceptions  : None
<a name="l00356"></a>00356     Return type : Arrayref of strings
<a name="l00357"></a>00357 =cut
<a name="l00358"></a>00358 
<a name="l00359"></a>00359 sub get_all_names {
<a name="l00360"></a>00360   my ($self) = @_;
<a name="l00361"></a>00361   <span class="keywordflow">return</span> [ map { $_-&gt;name() } @{ $self-&gt;{_adaptor}-&gt;fetch_all() } ];
<a name="l00362"></a>00362 }
<a name="l00363"></a>00363 
<a name="l00364"></a>00364 =head2 get_all_accessions
<a name="l00365"></a>00365     Description : Return list of all INSDC sequence accessions (or other seq_region names) registered with the helper
<a name="l00366"></a>00366     Exceptions  : None
<a name="l00367"></a>00367     Return type : Arrayref of strings
<a name="l00368"></a>00368 =cut
<a name="l00369"></a>00369 
<a name="l00370"></a>00370 sub get_all_accessions {
<a name="l00371"></a>00371   <span class="keywordflow">throw</span> <span class="stringliteral">&quot;Unimplemented method&quot;</span>;
<a name="l00372"></a>00372 }
<a name="l00373"></a>00373 
<a name="l00374"></a>00374 =head2 get_all_versioned_accessions
<a name="l00375"></a>00375     Description : Return list of all versioned INSDC sequence accessions (or other seq_region names) registered with the helper
<a name="l00376"></a>00376     Exceptions  : None
<a name="l00377"></a>00377     Return type : Arrayref of strings
<a name="l00378"></a>00378 =cut
<a name="l00379"></a>00379 
<a name="l00380"></a>00380 sub get_all_versioned_accessions {
<a name="l00381"></a>00381   <span class="keywordflow">throw</span> <span class="stringliteral">&quot;Unimplemented method&quot;</span>;
<a name="l00382"></a>00382 }
<a name="l00383"></a>00383 
<a name="l00384"></a>00384 =head2 get_all_assemblies
<a name="l00385"></a>00385     Description : Return list of all INSDC assembly accessions registered with the helper
<a name="l00386"></a>00386     Exceptions  : None
<a name="l00387"></a>00387     Return type : Arrayref of strings
<a name="l00388"></a>00388 =cut
<a name="l00389"></a>00389 
<a name="l00390"></a>00390 sub get_all_assemblies {
<a name="l00391"></a>00391   my ($self) = @_;
<a name="l00392"></a>00392   <span class="keywordflow">return</span> [ map { s/\.[0-9]+$<span class="comment">// } @{ $self-&gt;get_all_versioned_assemblies() } ];</span>
<a name="l00393"></a>00393 }
<a name="l00394"></a>00394 
<a name="l00395"></a>00395 =head2 get_all_versioned_assemblies
<a name="l00396"></a>00396     Description : Return list of all versioned INSDC assembly accessions registered with the helper
<a name="l00397"></a>00397     Exceptions  : None
<a name="l00398"></a>00398     Return type : Arrayref of strings
<a name="l00399"></a>00399 =cut
<a name="l00400"></a>00400 
<a name="l00401"></a>00401 sub get_all_versioned_assemblies {
<a name="l00402"></a>00402   my ($self) = @_;
<a name="l00403"></a>00403   <span class="keywordflow">return</span> [
<a name="l00404"></a>00404      uniq( map { $_-&gt;assembly_id() || <span class="stringliteral">&#39;&#39;</span> } @{ $self-&gt;{_adaptor}-&gt;fetch_all() } )
<a name="l00405"></a>00405   ];
<a name="l00406"></a>00406 }
<a name="l00407"></a>00407 
<a name="l00408"></a>00408 1;
<a name="l00409"></a>00409 
</pre></div></div>
</div>
  <div id="nav-path" class="navpath">
    <ul>
      <li class="navelem"><a class="el" href="RemoteLookUp_8pm.html">RemoteLookUp.pm</a>      </li>
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Namespaces</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Files</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Functions</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

  </div>
</body>
</html>
